<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1.0"><title>Changelog • MungeSumstats</title><!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous"><script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css"><script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous"><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous"><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet"><script src="../pkgdown.js"></script><meta property="og:title" content="Changelog"><!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--></head><body data-spy="scroll" data-target="#toc">
    

    <div class="container template-news">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">MungeSumstats</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">1.5.1</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav"><li>
  <a href="../articles/MungeSumstats.html">Get started</a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu"><li>
      <a href="../articles/OpenGWAS.html">Import GWAS summary statistics from Open GWAS</a>
    </li>
    <li>
      <a href="../articles/docker.html">Docker/Singularity Containers</a>
    </li>
  </ul></li>
<li>
  <a href="../news/index.html">Changelog</a>
</li>
      </ul><ul class="nav navbar-nav navbar-right"><li>
  <a href="https://github.com/neurogenomics/MungeSumstats/" class="external-link">
    <span class="fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul></div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
      <h1 data-toc-skip>Changelog <small></small></h1>
      <small>Source: <a href="https://github.com/neurogenomics/MungeSumstats/blob/HEAD/NEWS.md" class="external-link"><code>NEWS.md</code></a></small>
    </div>

    <div class="section level2">
<h2 class="page-header" data-toc-text="1.5.0" id="changes-in-version-150">CHANGES IN VERSION 1.5.0<a class="anchor" aria-label="anchor" href="#changes-in-version-150"></a></h2>
<div class="section level3">
<h3 id="new-features-1-5-0">New features<a class="anchor" aria-label="anchor" href="#new-features-1-5-0"></a></h3>
<ul><li>Added “query” column to <code>find_sumstats</code> output to keep track of search parameters.</li>
<li>
<code>import_sumstats</code>:
<ul><li>Check if formatted file (<code>save_path</code>) exists <em>before</em> downloading to save time.<br></li>
<li>Pass up <code>force_new</code> in additional to <code>force_new_vcf</code>.</li>
</ul></li>
<li>Updated <em>Description</em> tag in DESCRIPTION file to better reflect the scope of <code>MungeSumstats</code>.</li>
<li>Upgraded <code>read_vcf</code> to be more robust.</li>
<li>Edited Deps/Suggests
<ul><li>Elevate <code>IRanges</code> to Imports.</li>
<li>Remove <code>stringr</code> (no longer used)</li>
</ul></li>
<li>Add new internal function <code>is_tabix</code> to check whether a file is already tabix-indexed.</li>
<li>Bump <em>Depends: R(&gt;= 4.2)</em>.</li>
<li>
<code>read_sumstats</code>: now takes <code>samples</code> as an arg. By default, only uses first sample (if multiple are present in file).<br></li>
<li>Remove <code>INFO_filter=</code> from ALS VCF examples in vignettes (no longer necessary now that INFO parsing has been corrected).</li>
</ul></div>
<div class="section level3">
<h3 id="bug-fixes-1-5-0">Bug fixes<a class="anchor" aria-label="anchor" href="#bug-fixes-1-5-0"></a></h3>
<ul><li>AF (allele frequency) was accidentally being assigned as INFO column in VCFs where the INFO rows started with “AF”. This caused a large number of SNPs to be incorrectly dropped during the <code>check_info_score</code> step.</li>
<li>If INFO score is not available, INFO column is now dropped entirely (rather than assigning all 1s).
<ul><li>Adjusted <em>test-vcf_formatting</em> to reflect this. This avoids ambiguity about whether the INFO score is real or not.</li>
</ul></li>
<li>
<code>check_info_score</code>:
<ul><li>Added extra messages in various conditions where INFO is not used for filtering, and don’t add <code>log_files$info_filter</code> in these instances.<br></li>
<li>Added unit tests.<br></li>
</ul></li>
<li>
<code>check_empty_cols</code> was accidentally dropping more columns than it should have.</li>
<li>Fix GHA pkgdown building:
<ul><li>The newest version of <a href="https://github.com/actions/checkout/issues/760" class="external-link">git introduced bugs when building pkgdown sites</a> from within Docker containers (e.g. via my Linux GHA workflow). Adjusting GHA to fix this.</li>
</ul></li>
<li>Fix <code>write_sumstats</code> when indexing VCF.</li>
<li>Ensure <code>read_sumstats</code> can read in any VCF files (local/remote, indexed/non-indexed).</li>
<li>Fix <code>test-vcf_formatting.R</code>
<ul><li>line 51: had wrong AF value in string</li>
<li>line 109: encountering error? due to duplicate SNPs?</li>
</ul></li>
<li>Fix <code>test-check_impute_se_beta</code>
<ul><li>lines 51/52: <code>setkey</code> on SNP (now automatically renamed from ID by <code>read_vcf</code>).</li>
</ul></li>
<li>Fix <code>test-read_sumstats</code>:
<ul><li>standardising of headers is now handled internally by <code>read_sumstats</code>.</li>
<li>Ensure CHR is a character vector when being read in.</li>
<li>line 44: Ensure extra cols in <code>vcf_ss</code> are dropped.</li>
</ul></li>
<li>
<code>parse_logs</code>: Add lines to parsing subfunctions to allow handling of logs that don’t contain certain info (thus avoid warnings when creating the final data.table).<br></li>
<li>‘<em>Avoid the use of ’paste’ in condition signals</em>’ fixed:
<ul><li><code>check_pos_se</code></li>
<li><code>check_signed_col</code></li>
</ul></li>
<li>Used to rely on <em>gunzip</em> to read bgz files, but apparently this functionality is no longer supported (possibly due to changes to how <code><a href="https://rdrr.io/pkg/Rsamtools/man/zip.html" class="external-link">Rsamtools::bgzip</a></code> does compression in Bioc 3.15. Switched to using <code>fread + readLines</code> in:
<ul><li><code>read_header</code></li>
<li><code>read_sumstats</code></li>
</ul></li>
<li>
<code>read_header</code>: wasn’t reading in enough lines to get past the VCF header. Increase to <code>readLines(n=1000)</code>.<br></li>
<li>
<code>read_vcf</code>: Would sometimes induce duplicate rows. Now only unique rows are used (after sample and columns filtering).</li>
<li>Issue with mix of chr:bp:a1:a2 and chr:bp and rs id resolved</li>
</ul></div>
</div>
    <div class="section level2">
<h2 class="page-header" data-toc-text="1.3.19" id="changes-in-version-1319">CHANGES IN VERSION 1.3.19<a class="anchor" aria-label="anchor" href="#changes-in-version-1319"></a></h2>
<div class="section level3">
<h3 id="bug-fixes-1-3-19">Bug fixes<a class="anchor" aria-label="anchor" href="#bug-fixes-1-3-19"></a></h3>
<ul><li>
<code>format_sumstats</code> can now import remote files (other than OpenGWAS).</li>
</ul></div>
<div class="section level3">
<h3 id="new-features-1-3-19">New features<a class="anchor" aria-label="anchor" href="#new-features-1-3-19"></a></h3>
<ul><li>New <code>sumstatsColHeaders</code> entries:
<ul><li>“PosGRCh37” –&gt; “BP”</li>
<li>“testedAllele” –&gt; “A1”</li>
</ul></li>
</ul></div>
</div>
    <div class="section level2">
<h2 class="page-header" data-toc-text="1.3.18" id="changes-in-version-1318">CHANGES IN VERSION 1.3.18<a class="anchor" aria-label="anchor" href="#changes-in-version-1318"></a></h2>
<div class="section level3">
<h3 id="new-features-1-3-18">New features<a class="anchor" aria-label="anchor" href="#new-features-1-3-18"></a></h3>
<ul><li>Can now handle general remote sumstats not just IEU GWAS</li>
<li>More column header mappings</li>
</ul></div>
</div>
    <div class="section level2">
<h2 class="page-header" data-toc-text="1.3.17" id="changes-in-version-1317">CHANGES IN VERSION 1.3.17<a class="anchor" aria-label="anchor" href="#changes-in-version-1317"></a></h2>
<div class="section level3">
<h3 id="new-features-1-3-17">New features<a class="anchor" aria-label="anchor" href="#new-features-1-3-17"></a></h3>
<ul><li>Clean up of column header mapping file, including FREQUENCY given priority over MAF and addition of new CHR mappings.</li>
</ul></div>
</div>
    <div class="section level2">
<h2 class="page-header" data-toc-text="1.3.15" id="changes-in-version-1315">CHANGES IN VERSION 1.3.15<a class="anchor" aria-label="anchor" href="#changes-in-version-1315"></a></h2>
<div class="section level3">
<h3 id="bug-fixes-1-3-15">Bug fixes<a class="anchor" aria-label="anchor" href="#bug-fixes-1-3-15"></a></h3>
<ul><li>Handle cases for multi-trait GWAS when P columns exists separate to the trait specific P value so that when renaming occurs there isn’t two P columns. Inputted P column will be renamed to ‘P_input’</li>
<li>Issue where ‘check allele flip’ wasn’t running when the sumstats had all SNP IDs missing and incorrect direction of A1/A2 and effect columns has now been fixed.</li>
</ul></div>
</div>
    <div class="section level2">
<h2 class="page-header" data-toc-text="1.3.14" id="changes-in-version-1314">CHANGES IN VERSION 1.3.14<a class="anchor" aria-label="anchor" href="#changes-in-version-1314"></a></h2>
<div class="section level3">
<h3 id="new-features-1-3-14">New features<a class="anchor" aria-label="anchor" href="#new-features-1-3-14"></a></h3>
<ul><li>
<code>liftover</code>
<ul><li>Now exported function.</li>
<li>Added args for more user flexibility.</li>
<li>Uses <code><a href="https://rdrr.io/pkg/GenomeInfoDb/man/mapGenomeBuilds.html" class="external-link">GenomeInfoDb::mapGenomeBuilds</a></code> to standardise build names.</li>
<li>Warns users when mapped builds do not match one of the conversion options.</li>
<li>Choice to output as data.table or GRanges.</li>
<li>Added units tests for exported version.</li>
</ul></li>
<li>
<code>standardise_sumstats_column_headers_crossplatform</code>
<ul><li>Exported as <code>standardise_header</code> while keeping the original function name as an internal function (they call the same code).</li>
<li>Added unit tests for exported version.</li>
</ul></li>
<li>Added chunks to *Getting started<code>vignette     -</code>liftover` tutorial
<ul><li>“Quick formatting” of headers and file formats.</li>
</ul></li>
</ul></div>
<div class="section level3">
<h3 id="bug-fixes-1-3-14">Bug fixes<a class="anchor" aria-label="anchor" href="#bug-fixes-1-3-14"></a></h3>
<ul><li>
<code>check_pos_se</code>: Remove extra <code><a href="https://rdrr.io/r/base/message.html" class="external-link">message()</a></code> call around string.</li>
<li>
<code>check_signed_col</code>: Remove extra <code><a href="https://rdrr.io/r/base/message.html" class="external-link">message()</a></code> call around string.</li>
<li>
<code>write_sumstats</code>
<ul><li>Added extra round of sorting when <code>tabix_index=TRUE</code> because this is required for tabix.</li>
</ul></li>
</ul></div>
</div>
    <div class="section level2">
<h2 class="page-header" data-toc-text="1.3.13" id="changes-in-version-1313">CHANGES IN VERSION 1.3.13<a class="anchor" aria-label="anchor" href="#changes-in-version-1313"></a></h2>
<div class="section level3">
<h3 id="new-features-1-3-13">New Features<a class="anchor" aria-label="anchor" href="#new-features-1-3-13"></a></h3>
<ul><li>Additional mappings for CHR</li>
<li>Make A1, A2 upper-case</li>
</ul></div>
<div class="section level3">
<h3 id="bug-fixes-1-3-13">Bug fixes<a class="anchor" aria-label="anchor" href="#bug-fixes-1-3-13"></a></h3>
<ul><li>Bug fix for dealing with imputing SNP ID when there are indels</li>
</ul></div>
</div>
    <div class="section level2">
<h2 class="page-header" data-toc-text="1.3.11" id="changes-in-version-1311">CHANGES IN VERSION 1.3.11<a class="anchor" aria-label="anchor" href="#changes-in-version-1311"></a></h2>
<div class="section level3">
<h3 id="new-features-1-3-11">New Features<a class="anchor" aria-label="anchor" href="#new-features-1-3-11"></a></h3>
<ul><li>MungeSumstats can now handle Indels better. It will:
<ul><li>Not impute the RS ID of a SNP for an Indel</li>
<li>Not remove the Indel based on the RS ID not being present in the SNP ref dataset.</li>
<li>Not remove the Indel if it has the same base-pair location as a SNP in the sumstats.</li>
</ul></li>
<li>Can now handle vcfs with extensions .vcf.tsv, .vcf.tsv.gz and .vcf.tsv.bgz</li>
</ul></div>
<div class="section level3">
<h3 id="bug-fixes-1-3-11">Bug fixes<a class="anchor" aria-label="anchor" href="#bug-fixes-1-3-11"></a></h3>
<ul><li>For non-bi-allelic SNP runs, no longer remove duplicated SNPs based on their base-pair position or their RS ID.</li>
</ul></div>
</div>
    <div class="section level2">
<h2 class="page-header" data-toc-text="1.3.9" id="changes-in-version-139">CHANGES IN VERSION 1.3.9<a class="anchor" aria-label="anchor" href="#changes-in-version-139"></a></h2>
<div class="section level3">
<h3 id="new-features-1-3-9">New Features<a class="anchor" aria-label="anchor" href="#new-features-1-3-9"></a></h3>
<ul><li>Exported functions. Added examples and unit tests:
<ul><li><code>compute_nsize</code></li>
<li><code>standardise_sumstats_column_headers_crossplatform</code></li>
<li><code>formatted_example</code></li>
</ul></li>
<li>New arguments:
<ul><li>
<code>standardise_sumstats_column_headers_crossplatform</code>: Added arg <code>uppercase_unmapped</code> to to allow users to specify whether they want make the columns that could not be mapped to a standard name uppercase (default=<code>TRUE</code> for backcompatibility). Added arg <code>return_list</code> to specify whether to return a named list (default) or just the <code>data.table</code>.</li>
<li>
<code>formatted_example</code>: Added args <code>formatted</code> to specify whether the file should have its colnames standardised. Added args <code>sorted</code> to specify whether the file should sort the data by coordinates. Added arg <code>return_list</code> to specify whether to return a named list (default) or just the <code>data.table</code>.</li>
</ul></li>
<li>Removed <em>codecode.yml</em> and *_pkgdown.yml* files (no longer necessary).</li>
<li>Added Issues templates for Bugs and Feature requests.</li>
<li>Added <code>.datatable.aware=TRUE</code> to <em>.zzz</em> as <a href="https://rdatatable.gitlab.io/data.table/articles/datatable-importing.html#optionally-import-data-table-suggests" class="external-link">extra precaution</a>.<br></li>
<li>
<code>vcf2df</code>: Documented arguments.</li>
<li>Made v2 of hex sticker: <em>inst/hex/hex.png</em>
</li>
</ul></div>
<div class="section level3">
<h3 id="bug-fixes-1-3-9">Bug fixes<a class="anchor" aria-label="anchor" href="#bug-fixes-1-3-9"></a></h3>
<ul><li>Regenerated the <em>gh-pages</em> branch after it accidentally got deleted.</li>
<li>Remove temporary <em>docs/</em> folder.<br></li>
<li>Updated GitHub Actions.</li>
<li>Updated <em>Dockerfile</em> so it doesn’t run checks (this is now take care of by the GHA workflow).</li>
<li>Added Windows-specific folders to <em>.Rbuildignore</em>.</li>
<li>Made <em>to_GRanges.R</em> and <em>to_VRanges.R</em> file names lowercase to be congruent with function names.</li>
</ul></div>
</div>
    <div class="section level2">
<h2 class="page-header" data-toc-text="1.3.7" id="changes-in-version-137">CHANGES IN VERSION 1.3.7<a class="anchor" aria-label="anchor" href="#changes-in-version-137"></a></h2>
<div class="section level3">
<h3 id="bug-fixes-1-3-7">Bug fixes<a class="anchor" aria-label="anchor" href="#bug-fixes-1-3-7"></a></h3>
<ul><li>Bug in checking for bad characters in RSID fixed</li>
</ul></div>
</div>
    <div class="section level2">
<h2 class="page-header" data-toc-text="1.3.6" id="changes-in-version-136">CHANGES IN VERSION 1.3.6<a class="anchor" aria-label="anchor" href="#changes-in-version-136"></a></h2>
<div class="section level3">
<h3 id="new-features-1-3-6">New Features<a class="anchor" aria-label="anchor" href="#new-features-1-3-6"></a></h3>
<ul><li>Columns Beta and Standard Error can now be imputed. However note that this imputation is an approximation so could have an effect on downstream analysis. Use with caution.</li>
</ul></div>
</div>
    <div class="section level2">
<h2 class="page-header" data-toc-text="1.3.5" id="changes-in-version-135">CHANGES IN VERSION 1.3.5<a class="anchor" aria-label="anchor" href="#changes-in-version-135"></a></h2>
<div class="section level3">
<h3 id="bug-fixes-1-3-5">Bug fixes<a class="anchor" aria-label="anchor" href="#bug-fixes-1-3-5"></a></h3>
<ul><li>Flipping of Odds Ratio corrected (1/OR rather than -1*OR)</li>
</ul></div>
</div>
    <div class="section level2">
<h2 class="page-header" data-toc-text="1.3.4" id="changes-in-version-134">CHANGES IN VERSION 1.3.4<a class="anchor" aria-label="anchor" href="#changes-in-version-134"></a></h2>
<div class="section level3">
<h3 id="bug-fixes-1-3-4">Bug fixes<a class="anchor" aria-label="anchor" href="#bug-fixes-1-3-4"></a></h3>
<ul><li>Issue downloading chain file resolved</li>
</ul></div>
</div>
    <div class="section level2">
<h2 class="page-header" data-toc-text="1.3.3" id="changes-in-version-133">CHANGES IN VERSION 1.3.3<a class="anchor" aria-label="anchor" href="#changes-in-version-133"></a></h2>
<div class="section level3">
<h3 id="new-features-1-3-3">New Features<a class="anchor" aria-label="anchor" href="#new-features-1-3-3"></a></h3>
<ul><li>More mappings added to default mapping file.</li>
</ul></div>
</div>
    <div class="section level2">
<h2 class="page-header" data-toc-text="1.3.2" id="changes-in-version-132">CHANGES IN VERSION 1.3.2<a class="anchor" aria-label="anchor" href="#changes-in-version-132"></a></h2>
<div class="section level3">
<h3 id="bug-fixes-1-3-2">Bug fixes<a class="anchor" aria-label="anchor" href="#bug-fixes-1-3-2"></a></h3>
<ul><li>Previously rsids with characters added (e.g. rs1234567w) would cause an error when checking for the rsid on the reference genome. This has been fixed and the correct rsid will now be imputed from the reference genome for these cases.</li>
</ul></div>
</div>
    <div class="section level2">
<h2 class="page-header" data-toc-text="1.3.1" id="changes-in-version-131">CHANGES IN VERSION 1.3.1<a class="anchor" aria-label="anchor" href="#changes-in-version-131"></a></h2>
<div class="section level3">
<h3 id="new-features-1-3-1">New Features<a class="anchor" aria-label="anchor" href="#new-features-1-3-1"></a></h3>
<ul><li>
<code>import_sumstats</code>: Create individual folders for each GWAS dataset, with a respective <code>logs</code> subfolder to avoid overwriting log files when processing multiple GWAS.<br></li>
<li>
<code>parse_logs</code>: <strong>New function</strong> to convert logs from one or more munged GWAS into a <code>data.table</code>.<br></li>
<li>
<code>list_sumstats</code>: <strong>New function</strong> to recursively search for local summary stats files previously munged with <code>MungeSumstats</code>.<br></li>
<li>Added new dataset <code>inst/extdata/MungeSumstats_log_msg.txt</code> to test logs files.<br></li>
<li>Added unit tests for <code>list_sumstats</code> and <code>parse_logs</code>.</li>
<li>Added new Docker vignette.<br></li>
<li>Updated GHA workflows using <a href="https://github.com/neurogenomics/r_workflows" class="external-link">r_workflows</a>.<br></li>
<li>Remove <em>docs/</em> folder as the website will now be pushed to the <code>gh-pages</code> branch automatically by new GHA workflow.<br></li>
<li>Made documentation in README more clear and concise.<br></li>
<li>Added checks for p-values &gt;1 or &lt;0 via args <code>convert_large_p</code> and <code>convert_neg_p</code>, respectively. These are both handled by the new internal function <code>check_range_p_val</code>, which also reports the number of SNPs found meeting these criteria to the console/logs.<br></li>
<li>
<code>check_small_p_val</code> records which SNPs were imputed in a more robust way, by recording which SNPs met the criteria before making the changes (as opposed to inferred this info from which columns are 0 after making the changes). This function now only handles non-negative p-values, so that rows with negative p-values can be recorded/reported separately in the <code>check_range_p_val</code> step.<br></li>
<li>
<code>check_small_p_val</code> now reports the number of SNPs &lt;= 5e-324 to console/logs.</li>
<li>Unit tests have been added for both <code>check_range_p_val</code> and <code>check_small_p_val</code>.</li>
<li>
<code>parse_logs</code> can now extract information reported by <code>check_range_p_val</code> and <code>check_small_p_val</code>.<br></li>
<li>New internal function <code>logs_example</code> provides easy access to log file stored in <em>inst/extdata</em>, and includes documentation on how it was created.<br></li>
<li>Both <code>check_range_p_val</code> and <code>check_small_p_val</code> now use <code>#' <a href="https://github.com/inheritParams" class="external-link">@inheritParams</a> format_sumstats</code> to improve consistency of documentation.</li>
</ul></div>
<div class="section level3">
<h3 id="bug-fixes-1-3-1">Bug fixes<a class="anchor" aria-label="anchor" href="#bug-fixes-1-3-1"></a></h3>
<ul><li>Reduced vignette sizes.</li>
<li>Removed usage of <code>suppressWarnings</code> where possible.<br></li>
<li>Deleted old <em>.Rproj</em> file and hidden folder (contained large files).</li>
<li>Configured <em>.Rproj</em> so it doesn’t store large data files.</li>
<li>Fix badger issues: <a href="https://github.com/GuangchuangYu/badger/issues/34" class="external-link uri">https://github.com/GuangchuangYu/badger/issues/34</a>
</li>
<li>Prevent <em>test-index_tabix.R</em> from running due to errors (for now).</li>
</ul></div>
</div>
    <div class="section level2">
<h2 class="page-header" data-toc-text="1.3.0" id="changes-in-version-130">CHANGES IN VERSION 1.3.0<a class="anchor" aria-label="anchor" href="#changes-in-version-130"></a></h2>
<div class="section level3">
<h3 id="new-features-1-3-0">New Features<a class="anchor" aria-label="anchor" href="#new-features-1-3-0"></a></h3>
<ul><li>Version bump to align with Bioconductor release 3.14.</li>
</ul></div>
</div>
    <div class="section level2">
<h2 class="page-header" data-toc-text="1.1.27" id="changes-in-version-1127">CHANGES IN VERSION 1.1.27<a class="anchor" aria-label="anchor" href="#changes-in-version-1127"></a></h2>
<div class="section level3">
<h3 id="bug-fixes-1-1-27">Bug fixes<a class="anchor" aria-label="anchor" href="#bug-fixes-1-1-27"></a></h3>
<ul><li>
<code>validate_parameters</code> can now handle <code>ref_genome=NULL</code><br></li>
<li>
<em>.tsv.gz</em> no longer assigned suffix <em>.tsv</em>.<br></li>
<li>Made code width &lt;80 characters.<br></li>
<li>Changed <code>to_GRanges</code>/<code>to_GRanges</code> functions to all-lowercase functions (for consistency with other functions).</li>
<li>Set <code>nThread=1</code> in <code>data.table</code> test functions.</li>
</ul></div>
<div class="section level3">
<h3 id="new-features-1-1-27">New Features<a class="anchor" aria-label="anchor" href="#new-features-1-1-27"></a></h3>
<ul><li>Added tests for <code>get_genome_builds</code><br></li>
<li>Added early check for making sure the directory <code>save_path</code> is in was actually created (as opposed to finding out at the very end of the pipeline).</li>
<li>Tabix-indexing now available for tabular output data.</li>
<li>
<code>read_header</code> and <code>read_sumstats</code> now both work with .bgz files.</li>
</ul></div>
</div>
    <div class="section level2">
<h2 class="page-header" data-toc-text="1.1.26" id="changes-in-version-1126">CHANGES IN VERSION 1.1.26<a class="anchor" aria-label="anchor" href="#changes-in-version-1126"></a></h2>
<div class="section level3">
<h3 id="new-features-1-1-26">New Features<a class="anchor" aria-label="anchor" href="#new-features-1-1-26"></a></h3>
<ul><li>Extra mappings for FRQ column, see <code>data("sumstatsColHeaders")</code> for details</li>
</ul></div>
</div>
    <div class="section level2">
<h2 class="page-header" data-toc-text="1.1.23" id="changes-in-version-1123">CHANGES IN VERSION 1.1.23<a class="anchor" aria-label="anchor" href="#changes-in-version-1123"></a></h2>
<div class="section level3">
<h3 id="new-features-1-1-23">New Features<a class="anchor" aria-label="anchor" href="#new-features-1-1-23"></a></h3>
<ul><li>
<code>format_sumstats(FRQ_filter)</code> added so SNPs can now be filtered by allele frequency</li>
<li>Mapping file now has mappings for allele frequency (AF) to FRQ</li>
<li>VCF files with AF in INFO column e.g. ‘AF=…’ now converted to AF column</li>
<li>
<code>format_sumstats(frq_is_maf)</code> check added to infer if FRQ column values are minor/effect allele frequencies or not. frq_is_maf allows users to rename the FRQ column as MAJOR_ALLELE_FRQ if some values appear to be major allele frequencies</li>
</ul></div>
</div>
    <div class="section level2">
<h2 class="page-header" data-toc-text="1.1.19" id="changes-in-version-1119">CHANGES IN VERSION 1.1.19<a class="anchor" aria-label="anchor" href="#changes-in-version-1119"></a></h2>
<div class="section level3">
<h3 id="new-features-1-1-19">New Features<a class="anchor" aria-label="anchor" href="#new-features-1-1-19"></a></h3>
<ul><li>
<code><a href="../reference/get_genome_builds.html">get_genome_builds()</a></code> can now be called to quickly get the genome build without running the whole reformatting.</li>
<li>
<code>format_sumstats(compute_n)</code> now has more methods to compute the effective sample size with “ldsc”, “sum”, “giant” or “metal”.</li>
<li>
<code>format_sumstats(convert_ref_genome)</code> now implemented which can perform liftover to GRCh38 from GRCh37 and vice-versa enabling better cohesion between different study’s summary statistics.</li>
</ul></div>
</div>
    <div class="section level2">
<h2 class="page-header" data-toc-text="1.1.11" id="changes-in-version-1111">CHANGES IN VERSION 1.1.11<a class="anchor" aria-label="anchor" href="#changes-in-version-1111"></a></h2>
<div class="section level3">
<h3 id="bug-fixes-1-1-11">Bug fixes<a class="anchor" aria-label="anchor" href="#bug-fixes-1-1-11"></a></h3>
<ul><li>
<code>check_no_rs_snp</code> can now handle extra information after an RS ID. So if you have <code>rs1234:A:G</code> that will be separated into two columns.</li>
<li>
<code>check_two_step_col</code> and <code>check_four_step_col</code>, the two checks for when multiple columns are in one, have been updated so if not all SNPs have multiple columns or some have more than the expected number, this can now be handled.</li>
<li>Extra mappings for the <code>FRQ</code> column have been added to the mapping file</li>
</ul></div>
</div>
    <div class="section level2">
<h2 class="page-header" data-toc-text="1.1.11" id="changes-in-version-1111-1">CHANGES IN VERSION 1.1.11<a class="anchor" aria-label="anchor" href="#changes-in-version-1111-1"></a></h2>
<div class="section level3">
<h3 id="new-features-1-1-11">New Features<a class="anchor" aria-label="anchor" href="#new-features-1-1-11"></a></h3>
<ul><li>
<code>check_multi_rs_snp</code> can now handle all punctuation with/without spaces. So if a row contains <code>rs1234,rs5678</code> or <code>rs1234, rs5678</code> or any other punctuation character other than <code>,</code> these can be handled.</li>
<li>
<code>format_sumstats(path)</code> can now be passed a dataframe/datatable of the summary statistics directly as well as a path to their saved location.</li>
<li>Input summary statistics with <code>A0/A1</code> corresponding to ref/alt can now be handled by the mappign file as well as <code>A1/A2</code> corresponding to ref/alt.</li>
</ul></div>
</div>
    <div class="section level2">
<h2 class="page-header" data-toc-text="1.1.2" id="changes-in-version-112">CHANGES IN VERSION 1.1.2<a class="anchor" aria-label="anchor" href="#changes-in-version-112"></a></h2>
<div class="section level3">
<h3 id="new-features-1-1-2">New Features<a class="anchor" aria-label="anchor" href="#new-features-1-1-2"></a></h3>
<ul><li>
<code>import_sumstats</code> reads GWAS sum stats directly from Open GWAS. Now parallelised and reports how long each dataset took to import/format in total.</li>
<li>
<code>find_sumstats</code> searches Open GWAS for datasets.</li>
<li>
<code>compute_z</code> computes Z-score from P.</li>
<li>
<code>compute_n</code> computes N for all SNPs from user defined smaple size.</li>
<li>
<code>format_sumstats(ldsc_format=TRUE)</code> ensures sum stats can be fed directly into <a href="https://github.com/bulik/ldsc" class="external-link">LDSC</a> without any additional munging.</li>
<li>
<code>read_sumstats</code>, <code>write_sumstas</code>, and <code>download_vcf</code> functions now exported.<br></li>
<li>
<code>format_sumstats(sort_coordinates=TRUE)</code> sorts results by their genomic coordinates.</li>
<li>
<code>format_sumstats(return_data=TRUE)</code> returns data directly to user. Can be returned in either <code>data.table</code> (default), <code>GRanges</code> or <code>VRanges</code> format using <code>format_sumstats(return_format="granges")</code>.<br></li>
<li>
<code>format_sumstats(N_dropNA=TRUE)</code> (default) drops rows where N is missing.</li>
<li>
<code>format_sumstats(snp_ids_are_rs_ids=TRUE)</code> (default) Should the SNP IDs inputted be inferred as RS IDs or some arbitrary ID.</li>
<li>
<code>format_sumstats(write_vcf=TRUE)</code> writes a tabix-indexed VCF file instead of tabular format.</li>
<li>
<code>format_sumstats(save_path=...)</code> lets users decide where their results are saved and what they’re named.</li>
<li>When the <code>save_path</code> indicates it’s in <code><a href="https://rdrr.io/r/base/tempfile.html" class="external-link">tempdir()</a></code>, message warns users that these files will be deleted when R session ends.<br></li>
<li>Summary of data is given at the beginning and the end of <code>format_sumstats</code> via <code><a href="../reference/report_summary.html">report_summary()</a></code>.<br></li>
<li>Readability of <code><a href="../reference/preview_sumstats.html">preview_sumstats()</a></code> messages improved.<br></li>
<li>New checks standard error (SE) must &gt;0 and BETA (and other effect columns) must not equal 0: <code>format_sumstats(pos_se=TRUE,effect_columns_nonzero=TRUE)</code>
</li>
<li>Log directory containing all removed SNPs is now available and can be changed to a different directory by setting: <code>format_sumstats(log_folder_ind=TRUE,log_folder=tempdir())</code>
</li>
<li>All imputed data can now be identified with a column in the output using: <code>format_sumstats(imputation_ind=TRUE)</code>
</li>
<li>Users can now input their own mapping file to be used for the column header mapping in place of <code>data(sumstatsColHeaders)</code>. See <code>format_sumstats(mapping_file = mapping_file)</code>.</li>
</ul></div>
<div class="section level3">
<h3 id="bug-fixes-1-1-2">Bug fixes<a class="anchor" aria-label="anchor" href="#bug-fixes-1-1-2"></a></h3>
<ul><li>CHR column now standardised (X and Y caps, no “chr” prefix).</li>
<li>Allele flipping done on a per-SNP basis (instead of whole-column).</li>
<li>Allele flipping now includes FRQ column as well as effect columns.</li>
<li>The effect allele is now interpreted as the A2 allele consistent with <a href="https://www.ncbi.nlm.nih.gov/pmc/articles/PMC7805039/" class="external-link">IEU GWAS VCF approach</a>. A1 will always be the reference allele.</li>
<li>
<code>read_vcf</code> upgraded to account for more VCF formats.</li>
<li>
<code>check_n_num</code> now accounts for situations where N is a character vector and converts to numeric.</li>
</ul></div>
</div>
    <div class="section level2">
<h2 class="page-header" data-toc-text="1.1.1" id="changes-in-version-111">CHANGES IN VERSION 1.1.1<a class="anchor" aria-label="anchor" href="#changes-in-version-111"></a></h2>
<div class="section level3">
<h3 id="bug-fixes-1-1-1">Bug fixes<a class="anchor" aria-label="anchor" href="#bug-fixes-1-1-1"></a></h3>
<ul><li>Preprint publication citation added.</li>
</ul></div>
</div>
    <div class="section level2">
<h2 class="page-header" data-toc-text="1.0.0" id="changes-in-version-100">CHANGES IN VERSION 1.0.0<a class="anchor" aria-label="anchor" href="#changes-in-version-100"></a></h2>
<div class="section level3">
<h3 id="new-features-1-0-0">New Features<a class="anchor" aria-label="anchor" href="#new-features-1-0-0"></a></h3>
<ul><li>MungeSumstats released to Bioconductor.</li>
</ul></div>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <nav id="toc" data-toggle="toc" class="sticky-top"><h2 data-toc-skip>Contents</h2>
    </nav></div>

</div>


      <footer><div class="copyright">
  <p></p><p>Developed by Alan Murphy, Brian Schilder, Nathan Skene.</p>
</div>

<div class="pkgdown">
  <p></p><p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.3.9000.</p>
</div>

      </footer></div>

  


  

  </body></html>

