<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1.0"><title>Ensure A1 &amp; A2 are correctly named, if GWAS SNP constructed as
Alternative/Reference or Risk/Nonrisk alleles these SNPs will need to be
converted to Reference/Alternative or Nonrisk/Risk. Here non-risk is defined
as what's on the reference genome (this may not always be the case). — check_allele_flip • MungeSumstats</title><!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous"><script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css"><script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous"><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous"><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet"><script src="../pkgdown.js"></script><meta property="og:title" content="Ensure A1 &amp; A2 are correctly named, if GWAS SNP constructed as
Alternative/Reference or Risk/Nonrisk alleles these SNPs will need to be
converted to Reference/Alternative or Nonrisk/Risk. Here non-risk is defined
as what's on the reference genome (this may not always be the case). — check_allele_flip"><meta property="og:description" content="Ensure A1 &amp;amp; A2 are correctly named, if GWAS SNP constructed as
Alternative/Reference or Risk/Nonrisk alleles these SNPs will need to be
converted to Reference/Alternative or Nonrisk/Risk. Here non-risk is defined
as what's on the reference genome (this may not always be the case)."><!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--></head><body data-spy="scroll" data-target="#toc">
    

    <div class="container template-reference-topic">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">MungeSumstats</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">1.7.10</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav"><li>
  <a href="../articles/MungeSumstats.html">Get started</a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu"><li>
      <a href="../articles/OpenGWAS.html">Import GWAS summary statistics from Open GWAS</a>
    </li>
    <li>
      <a href="../articles/docker.html">Docker/Singularity Containers</a>
    </li>
  </ul></li>
<li>
  <a href="../news/index.html">Changelog</a>
</li>
      </ul><ul class="nav navbar-nav navbar-right"><li>
  <a href="https://github.com/neurogenomics/MungeSumstats/" class="external-link">
    <span class="fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul></div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>Ensure A1 &amp; A2 are correctly named, if GWAS SNP constructed as
Alternative/Reference or Risk/Nonrisk alleles these SNPs will need to be
converted to Reference/Alternative or Nonrisk/Risk. Here non-risk is defined
as what's on the reference genome (this may not always be the case).</h1>
    <small class="dont-index">Source: <a href="https://github.com/neurogenomics/MungeSumstats/blob/HEAD/R/check_allele_flip.R" class="external-link"><code>R/check_allele_flip.R</code></a></small>
    <div class="hidden name"><code>check_allele_flip.Rd</code></div>
    </div>

    <div class="ref-description">
    <p>Ensure A1 &amp; A2 are correctly named, if GWAS SNP constructed as
Alternative/Reference or Risk/Nonrisk alleles these SNPs will need to be
converted to Reference/Alternative or Nonrisk/Risk. Here non-risk is defined
as what's on the reference genome (this may not always be the case).</p>
    </div>

    <div id="ref-usage">
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">check_allele_flip</span><span class="op">(</span></span>
<span>  <span class="va">sumstats_dt</span>,</span>
<span>  <span class="va">path</span>,</span>
<span>  <span class="va">ref_genome</span>,</span>
<span>  <span class="va">rsids</span>,</span>
<span>  <span class="va">allele_flip_check</span>,</span>
<span>  <span class="va">allele_flip_drop</span>,</span>
<span>  <span class="va">allele_flip_z</span>,</span>
<span>  <span class="va">allele_flip_frq</span>,</span>
<span>  <span class="va">bi_allelic_filter</span>,</span>
<span>  <span class="va">imputation_ind</span>,</span>
<span>  <span class="va">log_folder_ind</span>,</span>
<span>  <span class="va">check_save_out</span>,</span>
<span>  <span class="va">tabix_index</span>,</span>
<span>  <span class="va">nThread</span>,</span>
<span>  <span class="va">log_files</span>,</span>
<span>  standardise_headers <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  <span class="va">mapping_file</span>,</span>
<span>  <span class="va">dbSNP</span></span>
<span><span class="op">)</span></span></code></pre></div>
    </div>

    <div id="arguments">
    <h2>Arguments</h2>
    <dl><dt>path</dt>
<dd><p>Filepath for the summary statistics file to be formatted. A
dataframe or datatable of the summary statistics file can also be passed
directly to MungeSumstats using the path parameter.</p></dd>


<dt>ref_genome</dt>
<dd><p>name of the reference genome used for the GWAS ("GRCh37" or
"GRCh38"). Argument is case-insensitive. Default is NULL which infers the
reference genome from the data.</p></dd>


<dt>allele_flip_check</dt>
<dd><p>Binary Should the allele columns be checked against
reference genome to infer if flipping is necessary. Default is TRUE.</p></dd>


<dt>allele_flip_drop</dt>
<dd><p>Binary Should the SNPs for which neither their A1 or
A2 base pair values match a reference genome be dropped. Default is TRUE.</p></dd>


<dt>allele_flip_z</dt>
<dd><p>Binary should the Z-score be flipped along with effect
and FRQ columns like Beta? It is assumed to be calculated off the effect size
not the P-value and so will be flipped i.e. default TRUE.</p></dd>


<dt>allele_flip_frq</dt>
<dd><p>Binary should the frequency (FRQ) column be flipped
along with effect and z-score columns like Beta? Default TRUE.</p></dd>


<dt>bi_allelic_filter</dt>
<dd><p>Binary Should non-biallelic SNPs be removed. Default
is TRUE.</p></dd>


<dt>imputation_ind</dt>
<dd><p>Binary Should a column be added for each imputation
step to show what SNPs have imputed values for differing fields. This
includes a field denoting SNP allele flipping (flipped). On the flipped
value, this denoted whether the alelles where switched based on
MungeSumstats initial choice of A1, A2 from the input column headers and thus
may not align with what the creator intended.<strong>Note</strong> these columns will be
in the formatted summary statistics returned. Default is FALSE.</p></dd>


<dt>log_folder_ind</dt>
<dd><p>Binary Should log files be stored containing all
filtered out SNPs (separate file per filter). The data is outputted in the
same format specified for the resulting sumstats file. The only exception to
this rule is if output is vcf, then log file saved as .tsv.gz. Default is
FALSE.</p></dd>


<dt>tabix_index</dt>
<dd><p>Index the formatted summary statistics with
<a href="http://www.htslib.org/doc/tabix.html" class="external-link">tabix</a> for fast querying.</p></dd>


<dt>nThread</dt>
<dd><p>Number of threads to use for parallel processes.</p></dd>


<dt>log_files</dt>
<dd><p>list of log file locations</p></dd>


<dt>standardise_headers</dt>
<dd><p>Run
<code>standardise_sumstats_column_headers_crossplatform</code> first.</p></dd>


<dt>mapping_file</dt>
<dd><p>MungeSumstats has a pre-defined column-name mapping file
which should cover the most common column headers and their interpretations.
However, if a column header that is in youf file is missing of the mapping we
give is incorrect you can supply your own mapping file. Must be a 2 column
dataframe with column names "Uncorrected" and "Corrected". See
data(sumstatsColHeaders) for default mapping and necessary format.</p></dd>


<dt>dbSNP</dt>
<dd><p>version of dbSNP to be used for imputation (144 or 155).</p></dd>

</dl></div>
    <div id="value">
    <h2>Value</h2>
    

<p>A list containing two data tables:</p><ul><li><p><code>sumstats_dt</code>: the modified summary statistics
<code>data.table</code> object.</p></li>
<li><p><code>rsids</code>: snpsById, filtered to SNPs of interest if
loaded already. Or else NULL.</p></li>
<li><p><code>log_files</code>: log file list</p></li>
</ul></div>

  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <nav id="toc" data-toggle="toc" class="sticky-top"><h2 data-toc-skip>Contents</h2>
    </nav></div>
</div>


      <footer><div class="copyright">
  <p></p><p>Developed by Alan Murphy, Brian Schilder, Nathan Skene.</p>
</div>

<div class="pkgdown">
  <p></p><p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.6.9000.</p>
</div>

      </footer></div>

  


  

  </body></html>

