<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1.0"><title>Infer genome builds — get_genome_builds • MungeSumstats</title><!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous"><script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css"><script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous"><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous"><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet"><script src="../pkgdown.js"></script><meta property="og:title" content="Infer genome builds — get_genome_builds"><meta property="og:description" content="Infers the genome build of  summary statistics files (GRCh37 or GRCh38)
from the data. Uses SNP (RSID) &amp;amp; CHR &amp;amp; BP to get genome build."><!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--></head><body data-spy="scroll" data-target="#toc">
    

    <div class="container template-reference-topic">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">MungeSumstats</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">1.3.12</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav"><li>
  <a href="../articles/MungeSumstats.html">Get started</a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu"><li>
      <a href="../articles/OpenGWAS.html">Import GWAS summary statistics from Open GWAS</a>
    </li>
    <li>
      <a href="../articles/docker.html">Docker</a>
    </li>
  </ul></li>
<li>
  <a href="../news/index.html">Changelog</a>
</li>
      </ul><ul class="nav navbar-nav navbar-right"><li>
  <a href="https://github.com/neurogenomics/MungeSumstats/" class="external-link">
    <span class="fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul></div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>Infer genome builds</h1>
    <small class="dont-index">Source: <a href="https://github.com/neurogenomics/MungeSumstats/blob/HEAD/R/get_genome_builds.R" class="external-link"><code>R/get_genome_builds.R</code></a></small>
    <div class="hidden name"><code>get_genome_builds.Rd</code></div>
    </div>

    <div class="ref-description">
    <p>Infers the genome build of  summary statistics files (GRCh37 or GRCh38)
from the data. Uses SNP (RSID) &amp; CHR &amp; BP to get genome build.</p>
    </div>

    <div id="ref-usage">
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="fu">get_genome_builds</span><span class="op">(</span>
  <span class="va">sumstats_list</span>,
  header_only <span class="op">=</span> <span class="cn">TRUE</span>,
  sampled_snps <span class="op">=</span> <span class="fl">10000</span>,
  names_from_paths <span class="op">=</span> <span class="cn">FALSE</span>,
  nThread <span class="op">=</span> <span class="fl">1</span>
<span class="op">)</span></code></pre></div>
    </div>

    <div id="arguments">
    <h2>Arguments</h2>
    <dl><dt>sumstats_list</dt>
<dd><p>A named list of paths to summary statistics,
or a named list of <code>data.table</code> objects.</p></dd>
<dt>header_only</dt>
<dd><p>Instead of reading in the entire <code>sumstats</code> file,
only read in the first N rows where N=<code>sampled_snps</code>.
This should help speed up cases where you have to read in <code>sumstats</code>
from disk each time.</p></dd>
<dt>sampled_snps</dt>
<dd><p>Downsample the number of SNPs used when inferring genome
build to save time.</p></dd>
<dt>names_from_paths</dt>
<dd><p>Infer the name of each item in <code>sumstats_list</code>
from its respective file path.
Only works if <code>sumstats_list</code> is a list of paths.</p></dd>
<dt>nThread</dt>
<dd><p>Number of threads to use for parallel processes.</p></dd>
</dl></div>
    <div id="value">
    <h2>Value</h2>
    <p>ref_genome the genome build of the data</p>
    </div>
    <div id="details">
    <h2>Details</h2>
    <p>Iterative version of <code>get_genome_build</code>.</p>
    </div>

    <div id="ref-examples">
    <h2>Examples</h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="r-in"><span class="co"># Pass path to Educational Attainment Okbay sumstat file to a temp directory</span></span>
<span class="r-in"></span>
<span class="r-in"><span class="va">eduAttainOkbayPth</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"extdata"</span>, <span class="st">"eduAttainOkbay.txt"</span>,</span>
<span class="r-in">    package <span class="op">=</span> <span class="st">"MungeSumstats"</span></span>
<span class="r-in"><span class="op">)</span></span>
<span class="r-in"><span class="va">sumstats_list</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span>ss1 <span class="op">=</span> <span class="va">eduAttainOkbayPth</span>, ss2 <span class="op">=</span> <span class="va">eduAttainOkbayPth</span><span class="op">)</span></span>
<span class="r-in"></span>
<span class="r-in"><span class="co">## Call uses reference genome as default with more than 2GB of memory,</span></span>
<span class="r-in"><span class="co">## which is more than what 32-bit Windows can handle so remove certain checks</span></span>
<span class="r-in"><span class="va">is_32bit_windows</span> <span class="op">&lt;-</span></span>
<span class="r-in">    <span class="va">.Platform</span><span class="op">$</span><span class="va">OS.type</span> <span class="op">==</span> <span class="st">"windows"</span> <span class="op">&amp;&amp;</span> <span class="va">.Platform</span><span class="op">$</span><span class="va">r_arch</span> <span class="op">==</span> <span class="st">"i386"</span></span>
<span class="r-in"><span class="kw">if</span> <span class="op">(</span><span class="op">!</span><span class="va">is_32bit_windows</span><span class="op">)</span> <span class="op">{</span></span>
<span class="r-in">    </span>
<span class="r-in">    <span class="co">#multiple sumstats can be passed at once to get all their genome builds:</span></span>
<span class="r-in">    <span class="co">#ref_genomes &lt;- get_genome_builds(sumstats_list = sumstats_list)</span></span>
<span class="r-in">    <span class="co">#just passing first here for speed</span></span>
<span class="r-in">    <span class="va">sumstats_list_quick</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span>ss1 <span class="op">=</span> <span class="va">eduAttainOkbayPth</span><span class="op">)</span></span>
<span class="r-in">    <span class="va">ref_genomes</span> <span class="op">&lt;-</span> <span class="fu">get_genome_builds</span><span class="op">(</span>sumstats_list <span class="op">=</span> <span class="va">sumstats_list_quick</span><span class="op">)</span></span>
<span class="r-in"><span class="op">}</span></span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Inferring genome build of 1 sumstats file(s).</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Inferring genome build.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Reading in only the first 10000 rows of sumstats.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Reading header.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Tabular format detected.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Importing tabular file: /__w/_temp/Library/MungeSumstats/extdata/eduAttainOkbay.txt</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Standardising column headers.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> First line of summary statistics file: </span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> MarkerName	CHR	POS	A1	A2	EAF	Beta	SE	Pval	</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Loading reference genome data.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Loading reference genome data.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Inferred genome build: GRCH37</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Time difference of 51.1068 secs</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> GRCH37: 1 file(s)</span>
</code></pre></div>
    </div>
  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <nav id="toc" data-toggle="toc" class="sticky-top"><h2 data-toc-skip>Contents</h2>
    </nav></div>
</div>


      <footer><div class="copyright">
  <p></p><p>Developed by Alan Murphy, Brian Schilder, Nathan Skene.</p>
</div>

<div class="pkgdown">
  <p></p><p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a>
2.0.2.9000.</p>
</div>

      </footer></div>

  


  

  </body></html>

