<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1.0"><title>Genome build liftover — liftover • MungeSumstats</title><!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous"><script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css"><script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous"><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous"><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet"><script src="../pkgdown.js"></script><meta property="og:title" content="Genome build liftover — liftover"><meta property="og:description" content="Transfer genomic coordinates from one genome build to another."><!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--></head><body data-spy="scroll" data-target="#toc">
    

    <div class="container template-reference-topic">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">MungeSumstats</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">1.5.2</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav"><li>
  <a href="../articles/MungeSumstats.html">Get started</a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu"><li>
      <a href="../articles/OpenGWAS.html">Import GWAS summary statistics from Open GWAS</a>
    </li>
    <li>
      <a href="../articles/docker.html">Docker/Singularity Containers</a>
    </li>
  </ul></li>
<li>
  <a href="../news/index.html">Changelog</a>
</li>
      </ul><ul class="nav navbar-nav navbar-right"><li>
  <a href="https://github.com/neurogenomics/MungeSumstats/" class="external-link">
    <span class="fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul></div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>Genome build liftover</h1>
    <small class="dont-index">Source: <a href="https://github.com/neurogenomics/MungeSumstats/blob/HEAD/R/liftover.R" class="external-link"><code>R/liftover.R</code></a></small>
    <div class="hidden name"><code>liftover.Rd</code></div>
    </div>

    <div class="ref-description">
    <p>Transfer genomic coordinates from one genome build to another.</p>
    </div>

    <div id="ref-usage">
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="fu">liftover</span><span class="op">(</span>
  <span class="va">sumstats_dt</span>,
  <span class="va">convert_ref_genome</span>,
  <span class="va">ref_genome</span>,
  imputation_ind <span class="op">=</span> <span class="cn">TRUE</span>,
  chrom_col <span class="op">=</span> <span class="st">"CHR"</span>,
  start_col <span class="op">=</span> <span class="st">"BP"</span>,
  end_col <span class="op">=</span> <span class="va">start_col</span>,
  as_granges <span class="op">=</span> <span class="cn">FALSE</span>,
  style <span class="op">=</span> <span class="st">"NCBI"</span>,
  verbose <span class="op">=</span> <span class="cn">TRUE</span>
<span class="op">)</span></code></pre></div>
    </div>

    <div id="source">
    <h2>Source</h2>
    <p><a href="https://doi.org/doi:10.18129/B9.bioc.liftOver" class="external-link">liftOver</a></p>
<p><a href="https://hgdownload.cse.ucsc.edu/goldenpath/hg19/liftOver/" class="external-link">
UCSC chain files</a></p>
    </div>
    <div id="arguments">
    <h2>Arguments</h2>
    <dl><dt>sumstats_dt</dt>
<dd><p>data table obj of the summary statistics
file for the GWAS.</p></dd>
<dt>convert_ref_genome</dt>
<dd><p>name of the reference genome to convert to
("GRCh37" or "GRCh38"). This will only occur if the current genome build does
not match. Default is not to convert the genome build (NULL).</p></dd>
<dt>ref_genome</dt>
<dd><p>name of the reference genome used for the GWAS ("GRCh37" or
"GRCh38"). Argument is case-insensitive. Default is NULL which infers the
reference genome from the data.</p></dd>
<dt>imputation_ind</dt>
<dd><p>Binary Should a column be added for each imputation
step to show what SNPs have imputed values for differing fields. This
includes a field denoting SNP allele flipping (flipped). On the flipped
value, this denoted whether the alelles where switched based on
MungeSumstats initial choice of A1, A2 from the input column headers and thus
may not align with what the creator intended.<strong>Note</strong> these columns will be
in the formatted summary statistics returned. Default is FALSE.</p></dd>
<dt>chrom_col</dt>
<dd><p>Name of the chromosome column in
<code>sumstats_dt</code> (e.g. "CHR").</p></dd>
<dt>start_col</dt>
<dd><p>Name of the starting genomic position
column in <code>sumstats_dt</code> (e.g. "POS","start").</p></dd>
<dt>end_col</dt>
<dd><p>Name of the ending genomic position
column in <code>sumstats_dt</code> (e.g. "POS","end").
Can be the same as <code>start_col</code> when <code>sumstats_dt</code>
only contains SNPs that span 1 base pair (bp) each.</p></dd>
<dt>as_granges</dt>
<dd><p>Return results as <a href="https://rdrr.io/pkg/GenomicRanges/man/GRanges-class.html" class="external-link">GRanges</a>
instead of a <a href="https://Rdatatable.gitlab.io/data.table/reference/data.table.html" class="external-link">data.table</a> (default: <code>FALSE</code>).</p></dd>
<dt>style</dt>
<dd><p>Style to return <a href="https://rdrr.io/pkg/GenomicRanges/man/GRanges-class.html" class="external-link">GRanges</a> object in
(e.g.  "NCBI" = 4; "UCSC" = "chr4";) (default: <code>"NCBI"</code>).</p></dd>
<dt>verbose</dt>
<dd><p>Print messages.</p></dd>
</dl></div>
    <div id="value">
    <h2>Value</h2>
    <p>Lifted summary stats in <code>data.table</code>or <a href="https://rdrr.io/pkg/GenomicRanges/man/GRanges-class.html" class="external-link">GRanges</a> format.</p>
    </div>

    <div id="ref-examples">
    <h2>Examples</h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="r-in"><span class="va">sumstats_dt</span> <span class="op">&lt;-</span> <span class="fu">MungeSumstats</span><span class="fu">::</span><span class="fu"><a href="formatted_example.html">formatted_example</a></span><span class="op">(</span><span class="op">)</span></span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Standardising column headers.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> First line of summary statistics file: </span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> MarkerName	CHR	POS	A1	A2	EAF	Beta	SE	Pval	</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Sorting coordinates.</span>
<span class="r-in"></span>
<span class="r-in"><span class="va">sumstats_dt_hg38</span> <span class="op">&lt;-</span> <span class="fu">liftover</span><span class="op">(</span>sumstats_dt<span class="op">=</span><span class="va">sumstats_dt</span>, </span>
<span class="r-in">                             ref_genome <span class="op">=</span> <span class="st">"hg19"</span>,</span>
<span class="r-in">                             convert_ref_genome<span class="op">=</span><span class="st">"hg38"</span><span class="op">)</span></span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Performing data liftover from hg19 to hg38.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Converting summary statistics to Genomic Ranges.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Downloading chain file from UCSC Genome Browser.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> /tmp/RtmpsSC7az/hg19ToHg38.over.chain.gz</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Reordering so first three column headers are SNP, CHR and BP in this order.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Reordering so the fourth and fifth columns are A1 and A2.</span>
</code></pre></div>
    </div>
  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <nav id="toc" data-toggle="toc" class="sticky-top"><h2 data-toc-skip>Contents</h2>
    </nav></div>
</div>


      <footer><div class="copyright">
  <p></p><p>Developed by Alan Murphy, Brian Schilder, Nathan Skene.</p>
</div>

<div class="pkgdown">
  <p></p><p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.3.9000.</p>
</div>

      </footer></div>

  


  

  </body></html>

