<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1.0"><title>Read VCF: parallel — read_vcf_parallel • MungeSumstats</title><!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous"><script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css"><script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous"><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous"><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet"><script src="../pkgdown.js"></script><meta property="og:title" content="Read VCF: parallel — read_vcf_parallel"><meta property="og:description" content="Read a VCF file across 1 or more threads in parallel.
If tilewidth is not specified, the size of each chunk will be
determined by total genome size divided by ntile.
By default, ntile is equal to the number of threads, nThread.
For further discussion on how this function was optimised,
see
here
and
here."><!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--></head><body data-spy="scroll" data-target="#toc">
    

    <div class="container template-reference-topic">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">MungeSumstats</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">1.5.2</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav"><li>
  <a href="../articles/MungeSumstats.html">Get started</a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu"><li>
      <a href="../articles/OpenGWAS.html">Import GWAS summary statistics from Open GWAS</a>
    </li>
    <li>
      <a href="../articles/docker.html">Docker/Singularity Containers</a>
    </li>
  </ul></li>
<li>
  <a href="../news/index.html">Changelog</a>
</li>
      </ul><ul class="nav navbar-nav navbar-right"><li>
  <a href="https://github.com/neurogenomics/MungeSumstats/" class="external-link">
    <span class="fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul></div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>Read VCF: parallel</h1>
    <small class="dont-index">Source: <a href="https://github.com/neurogenomics/MungeSumstats/blob/HEAD/R/read_vcf_parallel.R" class="external-link"><code>R/read_vcf_parallel.R</code></a></small>
    <div class="hidden name"><code>read_vcf_parallel.Rd</code></div>
    </div>

    <div class="ref-description">
    <p>Read a VCF file across 1 or more threads in parallel.
If <code>tilewidth</code> is not specified, the size of each chunk will be
determined by total genome size divided by <code>ntile</code>.
By default, <code>ntile</code> is equal to the number of threads, <code>nThread</code>.
For further discussion on how this function was optimised,
see
<a href="https://github.com/Bioconductor/VariantAnnotation/issues/59" class="external-link">here</a>
and
<a href="https://github.com/Bioconductor/VariantAnnotation/issues/57" class="external-link">here</a>.</p>
    </div>

    <div id="ref-usage">
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="fu">read_vcf_parallel</span><span class="op">(</span>
  <span class="va">path</span>,
  samples <span class="op">=</span> <span class="fl">1</span>,
  which <span class="op">=</span> <span class="cn">NULL</span>,
  use_params <span class="op">=</span> <span class="cn">TRUE</span>,
  as_datatable <span class="op">=</span> <span class="cn">TRUE</span>,
  sampled_rows <span class="op">=</span> <span class="fl">10000L</span>,
  include_xy <span class="op">=</span> <span class="cn">FALSE</span>,
  download <span class="op">=</span> <span class="cn">TRUE</span>,
  vcf_dir <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/tempfile.html" class="external-link">tempdir</a></span><span class="op">(</span><span class="op">)</span>,
  download_method <span class="op">=</span> <span class="st">"download.file"</span>,
  force_new <span class="op">=</span> <span class="cn">FALSE</span>,
  tilewidth <span class="op">=</span> <span class="cn">NULL</span>,
  mt_thresh <span class="op">=</span> <span class="fl">100000L</span>,
  nThread <span class="op">=</span> <span class="fl">1</span>,
  ntile <span class="op">=</span> <span class="va">nThread</span>,
  verbose <span class="op">=</span> <span class="cn">TRUE</span>
<span class="op">)</span></code></pre></div>
    </div>

    <div id="source">
    <h2>Source</h2>
    <p><code>
path &lt;- "https://gwas.mrcieu.ac.uk/files/ieu-a-298/ieu-a-298.vcf.gz"#### Single-threaded ####
vcf &lt;- MungeSumstats:::read_vcf_parallel(path = path)
#### Parallel ####
vcf2 &lt;- MungeSumstats:::read_vcf_parallel(path = path, nThread=11)
</code></p>
    </div>
    <div id="arguments">
    <h2>Arguments</h2>
    <dl><dt>path</dt>
<dd><p>Path to local or remote VCF file.</p></dd>
<dt>samples</dt>
<dd><p>Which samples to use:</p><ul><li><p>1 : Only the first sample will be used (<em>DEFAULT</em>).</p></li>
<li><p>NULL : All samples will be used.</p></li>
<li><p>c("&lt;sample_id1&gt;","&lt;sample_id2&gt;",...) : 
Only user-selected samples will be used (case-insensitive).</p></li>
</ul></dd>
<dt>which</dt>
<dd><p>A <code>GRanges</code> describing the sequences and
    ranges to be queried. Variants whose <code>POS</code> lies in the interval(s)
    <code>[start, end]</code> are returned. If <code>which</code> is not specified all
    ranges are returned.</p></dd>
<dt>use_params</dt>
<dd><p>When <code>TRUE</code> (default), increases the speed of reading in the VCF by
omitting columns that are empty based on the head of the VCF (NAs only).
NOTE that that this requires the VCF to be sorted, bgzip-compressed,
tabix-indexed, which <a href="read_vcf.html">read_vcf</a> will attempt to do.</p></dd>
<dt>as_datatable</dt>
<dd><p>Return the data as a
<a href="https://Rdatatable.gitlab.io/data.table/reference/data.table.html" class="external-link">data.table</a> (default: <code>TRUE</code>)
or a <a href="https://rdrr.io/pkg/VariantAnnotation/man/VCF-class.html" class="external-link">VCF</a> (<code>FALSE</code>).</p></dd>
<dt>sampled_rows</dt>
<dd><p>First N rows to sample.
Set <code>NULL</code> to use full <code>sumstats_file</code>.
when determining whether cols are empty.</p></dd>
<dt>download</dt>
<dd><p>Download the VCF (and its index file)
to a temp folder before reading it into R.
This is important to keep <code>TRUE</code> when <code>nThread&gt;1</code> to avoid
making too many queries to remote file.</p></dd>
<dt>vcf_dir</dt>
<dd><p>Where to download the original VCF from Open GWAS.
<em>WARNING:</em> This is set to <code><a href="https://rdrr.io/r/base/tempfile.html" class="external-link">tempdir()</a></code> by default.
This means the raw (pre-formatted) VCFs be deleted upon ending the R session.
Change this to keep the raw VCF file on disk
(e.g. <code>vcf_dir="./raw_vcf"</code>).</p></dd>
<dt>download_method</dt>
<dd><p><code>"axel"</code> (multi-threaded) or
<code>"download.file"</code> (single-threaded) .</p></dd>
<dt>force_new</dt>
<dd><p>If a formatted file of the same names as <code>save_path</code>
exists, formatting will be skipped and this file will be imported instead
(default). Set <code>force_new=TRUE</code> to override this.</p></dd>
<dt>tilewidth</dt>
<dd><p>The desired tile width. The effective tile width might be slightly
    different but is guaranteed to never be more than the desired width.</p></dd>
<dt>mt_thresh</dt>
<dd><p>When the number of rows (variants) in the VCF is
<code>&lt; mt_thresh</code>, only use single-threading for reading in the VCF.
This is because the overhead of parallelisation outweighs the speed benefits
when VCFs are small.</p></dd>
<dt>nThread</dt>
<dd><p>Number of threads to use for parallel processes.</p></dd>
<dt>ntile</dt>
<dd><p>The number of tiles to generate.</p></dd>
<dt>verbose</dt>
<dd><p>Print messages.</p></dd>
</dl></div>
    <div id="value">
    <h2>Value</h2>
    <p>VCF file.</p>
    </div>

  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <nav id="toc" data-toggle="toc" class="sticky-top"><h2 data-toc-skip>Contents</h2>
    </nav></div>
</div>


      <footer><div class="copyright">
  <p></p><p>Developed by Alan Murphy, Brian Schilder, Nathan Skene.</p>
</div>

<div class="pkgdown">
  <p></p><p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.3.9000.</p>
</div>

      </footer></div>

  


  

  </body></html>

