<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Standardise the format of summary statistics from GWAS with *MungeSumstats* • MungeSumstats</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link href="https://cdnjs.cloudflare.com/ajax/libs/bootswatch/3.4.0/flatly/bootstrap.min.css" rel="stylesheet" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="Standardise the format of summary statistics from GWAS with *MungeSumstats*">
<meta property="og:description" content="MungeSumstats">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--><!-- Global site tag (gtag.js) - Google Analytics --><script async src="https://www.googletagmanager.com/gtag/js?id=UA-175929032-1"></script><script>
  window.dataLayer = window.dataLayer || [];
  function gtag(){dataLayer.push(arguments);}
  gtag('js', new Date());

  gtag('config', 'UA-175929032-1');
</script>
</head>
<body data-spy="scroll" data-target="#toc">
    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">MungeSumstats</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">0.99.31</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../index.html">
    <span class="fas fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../articles/MungeSumstats.html">Get started</a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li>
  <a href="../news/index.html">Changelog</a>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/neurogenomics/MungeSumstats/">
    <span class="fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><script src="MungeSumstats_files/header-attrs-2.7/header-attrs.js"></script><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>Standardise the format of summary statistics from GWAS with <em>MungeSumstats</em>
</h1>
                        <h4 class="author">Alan Murphy and Nathan Skene</h4>
            
            <h4 class="date">2021-04-30</h4>
      
      <small class="dont-index">Source: <a href="https://github.com/neurogenomics/MungeSumstats/blob/master/vignettes/MungeSumstats.Rmd"><code>vignettes/MungeSumstats.Rmd</code></a></small>
      <div class="hidden name"><code>MungeSumstats.Rmd</code></div>

    </div>

    
    
<script type="text/javascript">
document.addEventListener("DOMContentLoaded", function() {
  document.querySelector("h1").className = "title";
});
</script><script type="text/javascript">
document.addEventListener("DOMContentLoaded", function() {
  var links = document.links;  
  for (var i = 0, linksLength = links.length; i < linksLength; i++)
    if (links[i].hostname != window.location.hostname)
      links[i].target = '_blank';
});
</script><div id="citation" class="section level1">
<h1 class="hasAnchor">
<a href="#citation" class="anchor"></a>Citation</h1>
<p>If you use the MungeSumstats package, please cite</p>
<p><a href="https://www.nature.com/articles/s41588-018-0129-5">Skene, et al. Genetic identification of brain cell types underlying schizophrenia, 2018.</a></p>
</div>
<div id="overview" class="section level1">
<h1 class="hasAnchor">
<a href="#overview" class="anchor"></a>Overview</h1>
<p>The <em>MungeSumstats</em> package is designed to facilitate the standardisation of GWAS summary statistics as utilised in our Nature Genetics paper.<span class="citation"><sup>1</sup></span></p>
<p>The package is designed to handle the lack of standardisation of output files by the GWAS community. There is a group who have now manually standardised many GWAS: <a href="https://mrcieu.github.io/ieugwasr/">R interface to the IEU GWAS database API • ieugwasr</a> and <a href="https://github.com/MRCIEU/gwasvcf">gwasvcf</a> but because a lot of GWAS remain closed access, these repositories are not all encompassing.</p>
<p>The <a href="https://github.com/mikegloudemans/gwas-download">GWAS-Download project</a> has collated summary statistics from 200+ GWAS. This repository has been utilsed to identify the most common formats, all of which can be standardised with <em>MungeSumstats</em>.</p>
<p>Moreover, there is an emerging standard of VCF format for summary statistics files with multiple, useful, associated R packages such as <em>vcfR</em>. However, there is currently no method to convert VCF formats to a standardised format that matches older approaches.</p>
<p>The <em>MungeSumstats</em> package standardises both VCF and the most common summary statistic file formats to enable downstream integration and analysis.</p>
<p><em>MungeSumstats</em> also offers a number of Quality Control (QC) steps which are important prerequisites for downstream analysis like Linkage disequilibrium score regression (LDSC) and MAGMA.</p>
</div>
<div id="aim" class="section level1">
<h1 class="hasAnchor">
<a href="#aim" class="anchor"></a>Aim</h1>
<p><em>MungeSumstats</em> will ensure that the all essential columns for analysis are present and syntactically correct. Generally, summary statistic files include (but are not limited to) the columns:</p>
<ul>
<li>SNP : SNP ID (rs IDs)</li>
<li>CHR : Chromosome number</li>
<li>BP : Base pair positions</li>
<li>A1 : Effect allele</li>
<li>A2 : Non-effect allele</li>
<li>Z : Z-score</li>
<li>BETA : Effect size estimate relative to the alternative allele</li>
<li>P : Unadjusted p-value for SNP</li>
<li>N : Sample size</li>
<li>INFO: The imputation information score</li>
<li>FRQ: The minor allele frequency (MAF) of the SNP</li>
</ul>
<p>Tests run by <em>MungeSumstats</em> include:</p>
<ul>
<li>Check VCF format</li>
<li>Check tab, space or comma delimited</li>
<li>Check for header name synonyms</li>
<li>Check for multiple models or traits in GWAS</li>
<li>Check for uniformity in SNP ID - no mix of rs/missing rs/chr:bp</li>
<li>Check for CHR:BP:A2:A1 all in one column</li>
<li>Check for CHR:BP in one column</li>
<li>Check for A1/A2 in one column</li>
<li>Check if CHR and/or BP is missing (infer from reference genome)</li>
<li>Check if SNP ID is missing (infer from reference genome)</li>
<li>Check if A1 and/or A2 are missing (infer from reference genome)</li>
<li>Check that vital columns are present (SNP,CHR,BP,P,A1,A2)</li>
<li>Check for one signed/effect column (Z,OR,BETA,LOG_ODDS,SIGNED_SUMSTAT)</li>
<li>Check for missing data</li>
<li>Check for duplicated columns</li>
<li>Check for small p-values (lower than 5e-324)</li>
<li>Check N column is an integer</li>
<li>Check for SNPs with N greater than 5 times standard dev. plus the mean<br>
</li>
<li>Check SNPs are RS ID’s</li>
<li>Check for duplicated rows, based on SNP ID</li>
<li>Check for duplicated rows, based on base-pair position</li>
<li>Check for SNPs on reference genome. Correct not found SNP IDs using CHR and BP (infer from reference genome)</li>
<li>Check INFO score</li>
<li>Check for strand-ambiguous SNPs</li>
<li>Check for non-biallelic SNPs (infer from reference genome)</li>
<li>Check for allele flipping</li>
<li>Check for SNPs on chromosome X, Y, and mitochondrial SNPs (MT)</li>
</ul>
<p>If a test is failed, the user will be notified and if possible, the input will be corrected. The QC steps from the checks above can also be adjusted to suit the user’s analysis, see <code><a href="../reference/format_sumstats.html">MungeSumstats::format_sumstats</a></code>.</p>
</div>
<div id="data" class="section level1">
<h1 class="hasAnchor">
<a href="#data" class="anchor"></a>Data</h1>
<p>The <em>MungeSumstats</em> package contains small subsets of GWAS summary statistics files. Firstly, on Educational Attainment by Okbay et al 2016: PMID: 27898078 PMCID: PMC5509058 DOI: 10.1038/ng1216-1587b.</p>
<p>Secondly, a VCF file (VCFv4.2) relating to the GWAS Amyotrophic lateral sclerosis from ieu open GWAS project. Dataset: ebi-a-GCST005647: <a href="https://gwas.mrcieu.ac.uk/datasets/ebi-a-GCST005647/" class="uri">https://gwas.mrcieu.ac.uk/datasets/ebi-a-GCST005647/</a></p>
<p>These datasets will be used to showcase MungeSumstats functionality.</p>
</div>
<div id="running-mungesumstats" class="section level1">
<h1 class="hasAnchor">
<a href="#running-mungesumstats" class="anchor"></a>Running MungeSumstats</h1>
<p><em>MungeSumstats</em> is available on Bioconductor. To install the package on Bioconductor run the following lines of code:</p>
<pre><code>if (!require("BiocManager"))
    install.packages("BiocManager")

BiocManager::install("MungeSumstats")</code></pre>
<p>Once installed, load the package:</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/neurogenomics/MungeSumstats">MungeSumstats</a></span><span class="op">)</span></code></pre></div>
<p>To standardise the summary statistics’ file format, simply call <code><a href="../reference/format_sumstats.html">format_sumstats()</a></code> passing in the path to your summary statistics file. You must also specify which genome build was used in the GWAS(GRCh37 or GRCh38). The reference genome is used for multiple checks like deriving missing data such SNP/BP/CHR/A1/A2 and for QC steps like removing non-biallelic SNPs or strand-ambiguous SNPs. The path to the reformatted summary statistics file is returned by the function call.</p>
<p>The Educational Attainment by Okbay GWAS summary statistics file is saved as a text document in the package’s external data folder so we can just pass the file path to it straight to <em>MungeSumstats</em>.</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">eduAttainOkbayPth</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html">system.file</a></span><span class="op">(</span><span class="st">"extdata"</span>,<span class="st">"eduAttainOkbay.txt"</span>,
                                  package<span class="op">=</span><span class="st">"MungeSumstats"</span><span class="op">)</span>
<span class="va">reformatted</span> <span class="op">&lt;-</span> 
  <span class="fu">MungeSumstats</span><span class="fu">::</span><span class="fu"><a href="../reference/format_sumstats.html">format_sumstats</a></span><span class="op">(</span>path<span class="op">=</span><span class="va">eduAttainOkbayPth</span>,ref_genome<span class="op">=</span><span class="st">"GRCh37"</span><span class="op">)</span></code></pre></div>
<pre><code>## First line of summary statistics file:</code></pre>
<pre><code>## MarkerName   CHR POS A1  A2  EAF Beta    SE  Pval    </code></pre>
<pre><code>## Succesfully finished preparing sumstats file, preview:</code></pre>
<pre><code>## SNP  CHR BP  A1  A2  FRQ BETA    SE  P</code></pre>
<pre><code>## rs12646808   4   3249828 T   C   0.6418  0.016   0.003   4.002e-08</code></pre>
<pre><code>## rs301800 1   8490603 T   C   0.1791  0.019   0.003   1.794e-08</code></pre>
<p>The hyperparameters <code>format_sumstats</code> in that control the level of QC conducted by <em>MungeSumstats</em> are:</p>
<ul>
<li>
<em>convert_small_p</em> Binary, should p-values &lt; 5e-324 be converted to 0? Small p-values pass the R limit and can cause errors with LDSC/MAGMA and should be converted. Default is TRUE.</li>
<li>
<em>convert_n_int</em> Binary, if N (the number of samples) is not an integer, should this be rounded? Default is TRUE. analysis_trait If multiple traits were studied, name of the trait for analysis from the GWAS. Default is NULL</li>
<li>
<em>INFO_filter</em> 0-1 The minimum value permissible of the imputation information score (if present in sumstatsfile). Default 0.9</li>
<li>
<em>N_std</em> Numeric, the number of standard deviations above the mean a SNP’s N is needed to be removed. Default is 5.</li>
<li>
<em>rmv_chr</em> vector or character The chromosomes on which the SNPs should be removed. Use NULL if no filtering necessary. Default is X, Y and mitochondrial.</li>
<li>
<em>on_ref_genome</em> Binary, should a check take place that all SNPs are on the reference genome by SNP ID. Any that aren’t on SNPs not on the reference genome, will be corrected from the reference genome (if possible) from the chromosome and base pair postion data. Default is TRUE</li>
<li>
<em>strand_ambig_filter</em> Binary, should SNPs with strand-ambiguous alleles be removed. Default is FALSE</li>
<li>
<em>allele_flip_check</em> Binary, should the allele columns be chacked against reference genome to infer if flipping is necessary. Default is TRUE</li>
<li>
<em>bi_allelic_filter</em> Binary, should non-biallelic SNPs be removed. Default is TRUE</li>
</ul>
<p>VCF files can also be standardised to the same format as other summary statistic files. A subset of the Amyotrophic lateral sclerosis GWAS from the ieu open GWAS project (a .vcf file) has been added to <em>MungeSumstats</em> to demonstrate this functionality.Simply pass the path to the file in the same manner you would for other summary statistic files:</p>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co">#save ALS GWAS from the ieu open GWAS project to a temp directory</span>
<span class="va">ALSvcfPth</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html">system.file</a></span><span class="op">(</span><span class="st">"extdata"</span>,<span class="st">"ALSvcf.vcf"</span>,package<span class="op">=</span><span class="st">"MungeSumstats"</span><span class="op">)</span>
<span class="co">#set a low INFO filter so we get a return</span>
<span class="va">reformatted_vcf</span> <span class="op">&lt;-</span> 
  <span class="fu">MungeSumstats</span><span class="fu">::</span><span class="fu"><a href="../reference/format_sumstats.html">format_sumstats</a></span><span class="op">(</span>path<span class="op">=</span><span class="va">ALSvcfPth</span>, ref_genome<span class="op">=</span><span class="st">"GRCh37"</span>,INFO_filter<span class="op">=</span><span class="fl">0.1</span><span class="op">)</span></code></pre></div>
<pre><code>## VCF format detected, this will be converted to a standard summary statistics file format.</code></pre>
<pre><code>## Inputted VCF format has -log10 P-values, these will be converted to unadjusted p-values in the 'P' column.</code></pre>
<pre><code>## First line of summary statistics file:</code></pre>
<pre><code>## CHROM    POS REF ALT INFO    ES  SE  LP  AF  ID  P   </code></pre>
<pre><code>## 12 SNPs are not on the reference genome.  These will be corrected from the reference genome.</code></pre>
<pre><code>## Reordering so first three column headers are SNP, CHR and BP in this order.</code></pre>
<pre><code>## 67 SNPs are below the INFO threshold of 0.1 and will be removed</code></pre>
<pre><code>## Succesfully finished preparing sumstats file, preview:</code></pre>
<pre><code>## SNP  CHR BP  A1  A2  INFO    BETA    SE  LP  AF  P</code></pre>
<pre><code>## rs58108140   1   10583   G   A   0.1589  0.0312  0.0393  0.369267    0.1589  0.427300105456596</code></pre>
<pre><code>## rs806731 1   30923   G   T   0.7843  -0.0114 0.0353  0.126854    0.7843  0.746699739815279</code></pre>
</div>
<div id="future-enhancements" class="section level1">
<h1 class="hasAnchor">
<a href="#future-enhancements" class="anchor"></a>Future Enhancements</h1>
<p>The <em>MungeSumstats</em> package aims to be able to handle the most common summary statistic file formats including VCF. If your file can not be formatted by <em>MungeSumstats</em> feel free to report the bug on github: <a href="https://github.com/neurogenomics/MungeSumstats" class="uri">https://github.com/neurogenomics/MungeSumstats</a> along with your summary statistic file header.</p>
<p>We also encourage people to edit the code to resolve their particular issues too and are happy to incorporate these through pull requests on github. If your summary statistic file headers are not recognised by <em>MungeSumstats</em> but correspond to one of SNP, BP, CHR, A1, A2, P, Z, OR, BETA, LOG_ODDS, SIGNED_SUMSTAT, N, N_CAS, N_CON, NSTUDY, INFO or FRQ, feel free to update the <code><a href="../reference/sumstatsColHeaders.html">MungeSumstats::sumstatsColHeaders</a></code> following the approach in the data.R file and add your mapping. Then use a pull request on github and we will incorporate this change into the package.</p>
</div>
<div id="session-information" class="section level1">
<h1 class="hasAnchor">
<a href="#session-information" class="anchor"></a>Session Information</h1>
<pre><code>## R Under development (unstable) (2021-02-19 r80028)
## Platform: x86_64-apple-darwin17.0 (64-bit)
## Running under: macOS Big Sur 10.16
## 
## Matrix products: default
## BLAS:   /Library/Frameworks/R.framework/Versions/4.1/Resources/lib/libRblas.dylib
## LAPACK: /Library/Frameworks/R.framework/Versions/4.1/Resources/lib/libRlapack.dylib
## 
## locale:
## [1] en_IE.UTF-8/en_IE.UTF-8/en_IE.UTF-8/C/en_IE.UTF-8/en_IE.UTF-8
## 
## attached base packages:
## [1] stats     graphics  grDevices utils     datasets  methods   base     
## 
## other attached packages:
## [1] MungeSumstats_0.99.31 BiocStyle_2.19.2     
## 
## loaded via a namespace (and not attached):
##  [1] SummarizedExperiment_1.21.3                
##  [2] xfun_0.22                                  
##  [3] bslib_0.2.4                                
##  [4] lattice_0.20-41                            
##  [5] htmltools_0.5.1.1                          
##  [6] stats4_4.1.0                               
##  [7] rtracklayer_1.51.5                         
##  [8] yaml_2.2.1                                 
##  [9] XML_3.99-0.6                               
## [10] rlang_0.4.10                               
## [11] pkgdown_1.6.1                              
## [12] jquerylib_0.1.4                            
## [13] BiocParallel_1.25.5                        
## [14] BiocGenerics_0.37.4                        
## [15] matrixStats_0.58.0                         
## [16] GenomeInfoDbData_1.2.4                     
## [17] stringr_1.4.0                              
## [18] zlibbioc_1.37.0                            
## [19] MatrixGenerics_1.3.1                       
## [20] Biostrings_2.59.2                          
## [21] ragg_1.1.2                                 
## [22] memoise_2.0.0                              
## [23] evaluate_0.14                              
## [24] restfulr_0.0.13                            
## [25] Biobase_2.51.0                             
## [26] knitr_1.33                                 
## [27] IRanges_2.25.10                            
## [28] fastmap_1.1.0                              
## [29] GenomeInfoDb_1.27.11                       
## [30] parallel_4.1.0                             
## [31] BSgenome.Hsapiens.1000genomes.hs37d5_0.99.1
## [32] BSgenome_1.59.2                            
## [33] BiocManager_1.30.12                        
## [34] SNPlocs.Hsapiens.dbSNP144.GRCh37_0.99.20   
## [35] cachem_1.0.4                               
## [36] DelayedArray_0.17.10                       
## [37] S4Vectors_0.29.17                          
## [38] desc_1.3.0                                 
## [39] jsonlite_1.7.2                             
## [40] XVector_0.31.1                             
## [41] systemfonts_1.0.1                          
## [42] fs_1.5.0                                   
## [43] Rsamtools_2.7.2                            
## [44] textshaping_0.3.3                          
## [45] rjson_0.2.20                               
## [46] digest_0.6.27                              
## [47] stringi_1.5.3                              
## [48] bookdown_0.22                              
## [49] BiocIO_1.1.2                               
## [50] GenomicRanges_1.43.4                       
## [51] rprojroot_2.0.2                            
## [52] grid_4.1.0                                 
## [53] tools_4.1.0                                
## [54] bitops_1.0-7                               
## [55] magrittr_2.0.1                             
## [56] sass_0.3.1                                 
## [57] RCurl_1.98-1.3                             
## [58] crayon_1.4.1                               
## [59] Matrix_1.3-2                               
## [60] data.table_1.14.0                          
## [61] rmarkdown_2.7                              
## [62] R6_2.5.0                                   
## [63] GenomicAlignments_1.27.2                   
## [64] compiler_4.1.0</code></pre>
</div>
<div id="references" class="section level1 unnumbered">
<h1 class="hasAnchor">
<a href="#references" class="anchor"></a>References</h1>
<div id="refs" class="references csl-bib-body" line-spacing="2">
<div id="ref-Skene2018" class="csl-entry">
<div class="csl-left-margin">1. </div>
<div class="csl-right-inline">Nathan G. Skene, T. E. B., Julien Bryois. Genetic identification of brain cell types underlying schizophrenia. <em>Nature Genetics</em> (2018). doi:<a href="https://doi.org/10.1038/s41588-018-0129-5">10.1038/s41588-018-0129-5</a>
</div>
</div>
</div>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

        <nav id="toc" data-toggle="toc"><h2 data-toc-skip>Contents</h2>
    </nav>
</div>

</div>



      <footer><div class="copyright">
  <p>Developed by Alan Murphy, Nathan Skene.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.6.1.</p>
</div>

      </footer>
</div>

  


  </body>
</html>
