<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Standardise GWAS summary statistics with *MungeSumstats* • MungeSumstats</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link href="https://cdnjs.cloudflare.com/ajax/libs/bootswatch/3.4.0/flatly/bootstrap.min.css" rel="stylesheet" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="Standardise GWAS summary statistics with *MungeSumstats*">
<meta property="og:description" content="MungeSumstats">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--><!-- Global site tag (gtag.js) - Google Analytics --><script async src="https://www.googletagmanager.com/gtag/js?id=UA-175929032-1"></script><script>
  window.dataLayer = window.dataLayer || [];
  function gtag(){dataLayer.push(arguments);}
  gtag('js', new Date());

  gtag('config', 'UA-175929032-1');
</script>
</head>
<body data-spy="scroll" data-target="#toc">
    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">MungeSumstats</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">1.1.29</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../index.html">
    <span class="fas fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../articles/MungeSumstats.html">Get started</a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/OpenGWAS.html">Import GWAS summary statistics from Open GWAS</a>
    </li>
  </ul>
</li>
<li>
  <a href="../news/index.html">Changelog</a>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/neurogenomics/MungeSumstats/">
    <span class="fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><script src="MungeSumstats_files/header-attrs-2.11/header-attrs.js"></script><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>Standardise GWAS summary statistics with <em>MungeSumstats</em>
</h1>
                        <h4 class="author">Alan Murphy, Brian Schilder, and Nathan Skene</h4>
            
            <h4 class="date">2021-10-09</h4>
      
      <small class="dont-index">Source: <a href="https://github.com/neurogenomics/MungeSumstats/blob/master/vignettes/MungeSumstats.Rmd"><code>vignettes/MungeSumstats.Rmd</code></a></small>
      <div class="hidden name"><code>MungeSumstats.Rmd</code></div>

    </div>

    
    
<script type="text/javascript">
document.addEventListener("DOMContentLoaded", function() {
  document.querySelector("h1").className = "title";
});
</script><script type="text/javascript">
document.addEventListener("DOMContentLoaded", function() {
  var links = document.links;  
  for (var i = 0, linksLength = links.length; i < linksLength; i++)
    if (links[i].hostname != window.location.hostname)
      links[i].target = '_blank';
});
</script><div id="citation" class="section level1">
<h1 class="hasAnchor">
<a href="#citation" class="anchor"></a>Citation</h1>
<p>If you use the MungeSumstats package, please cite</p>
<p><a href="https://www.biorxiv.org/content/10.1101/2021.06.21.449239v1">Murphy and Skene. MungeSumstats: A Bioconductor package for the standardisation and quality control of many GWAS summary statistics</a>.</p>
</div>
<div id="overview" class="section level1">
<h1 class="hasAnchor">
<a href="#overview" class="anchor"></a>Overview</h1>
<p>The <em>MungeSumstats</em> package is designed to facilitate the standardisation of GWAS summary statistics as utilised in our Nature Genetics paper.<span class="citation"><sup>1</sup></span></p>
<p>The package is designed to handle the lack of standardisation of output files by the GWAS community. There is a group who have now manually standardised many GWAS: <a href="https://mrcieu.github.io/ieugwasr/">R interface to the IEU GWAS database API • ieugwasr</a> and <a href="https://github.com/MRCIEU/gwasvcf">gwasvcf</a> but because a lot of GWAS remain closed access, these repositories are not all encompassing.</p>
<p>The <a href="https://github.com/mikegloudemans/gwas-download">GWAS-Download project</a> has collated summary statistics from 200+ GWAS. This repository has been utilsed to identify the most common formats, all of which can be standardised with <em>MungeSumstats</em>.</p>
<p>Moreover, there is an emerging standard of VCF format for summary statistics files with multiple, useful, associated R packages such as <em>vcfR</em>. However, there is currently no method to convert VCF formats to a standardised format that matches older approaches.</p>
<p>The <em>MungeSumstats</em> package standardises both VCF and the most common summary statistic file formats to enable downstream integration and analysis.</p>
<p><em>MungeSumstats</em> also offers comprehensive Quality Control (QC) steps which are important prerequisites for downstream analysis like Linkage disequilibrium score regression (LDSC) and MAGMA.</p>
<p>Moreover, <em>MungeSumstats</em> is efficiently written resulting in all reformatting and quality control checks completing in minutes for GWAS summary statistics with 500k SNPs on a standard desktop machine. This speed can be increased further by increasing the number of threads (nThread) for <code>data.table</code> to use.</p>
<p>Currently <em>MungeSumstats</em> only works on data from humans, as it uses human-based genome references.</p>
</div>
<div id="aim" class="section level1">
<h1 class="hasAnchor">
<a href="#aim" class="anchor"></a>Aim</h1>
<p><em>MungeSumstats</em> will ensure that the all essential columns for analysis are present and syntactically correct. Generally, summary statistic files include (but are not limited to) the columns:</p>
<ul>
<li>SNP : SNP ID (rs IDs)</li>
<li>CHR : Chromosome number</li>
<li>BP : Base pair positions</li>
<li>A1 : reference allele</li>
<li>A2 : alternative allele</li>
<li>Z : Z-score</li>
<li>BETA : Effect size estimate relative to the alternative allele</li>
<li>P : Unadjusted p-value for SNP</li>
<li>SE : The standard error</li>
<li>N : Sample size</li>
<li>INFO: The imputation information score</li>
<li>FRQ: The minor/effect allele frequency (MAF/EAF) of the SNP</li>
</ul>
<p><em>MungeSumstats</em> uses a mapping file to infer the inputted column names (run <code><a href="https://rdrr.io/r/utils/data.html">data("sumstatsColHeaders")</a></code> to view these). This mapping file is far more comprehensive than any other publicly available munging tool containing more than 200 unique mappings at the time of writing this vignette. However, if your column headers are missing or if you want to change the mapping, you can do so by passing your own mapping file (see <code><a href="../reference/format_sumstats.html">format_sumstats(mapping_file)</a></code>).</p>
<p><em>MungeSumstats</em> offers unmatched levels of quality control to ensure, for example, consistency of allele assignment and direction of effects. Tests run by <em>MungeSumstats</em> include:</p>
<ul>
<li>Check VCF format</li>
<li>Check tab, space or comma delimited, zipped, csv or tsv file</li>
<li>Check for header name synonyms</li>
<li>Check for multiple models or traits in GWAS</li>
<li>Check for uniformity in SNP ID - no mix of rs/missing rs/chr:bp</li>
<li>Check for CHR:BP:A2:A1 all in one column</li>
<li>Check for CHR:BP in one column</li>
<li>Check for A1/A2 in one column</li>
<li>Check if CHR and/or BP is missing (infer from reference genome)</li>
<li>Check if SNP ID is missing (infer from reference genome)</li>
<li>Check if A1 and/or A2 are missing (infer from reference genome)</li>
<li>Check that vital columns are present (SNP,CHR,BP,P,A1,A2)</li>
<li>Check for one signed/effect column (Z,OR,BETA,LOG_ODDS,SIGNED_SUMSTAT)</li>
<li>Check for missing data</li>
<li>Check for duplicated columns</li>
<li>Check for small p-values (lower than 5e-324)</li>
<li>Check N column is an integer</li>
<li>Check for SNPs with N greater than 5 times standard dev. plus the mean</li>
<li>Check SNPs are RS ID’s</li>
<li>Check for uniformity of SNP ID format</li>
<li>Check for duplicated rows, based on SNP ID</li>
<li>Check for duplicated rows, based on base-pair position</li>
<li>Check for SNPs on reference genome. Correct not found SNP IDs using CHR and BP (infer from reference genome)</li>
<li>Check INFO score</li>
<li>Check FRQ value</li>
<li>Check FRQ is minor allele frequency (MAF)</li>
<li>Check that the SNPs’ standard error (SE) is positive</li>
<li>Check that SNPs’ effect columns (like BETA) aren’t equal to 0</li>
<li>Check for strand-ambiguous SNPs</li>
<li>Check for non-biallelic SNPs (infer from reference genome)</li>
<li>Check for allele flipping</li>
<li>Check for SNPs on chromosome X, Y, and mitochondrial SNPs (MT)</li>
<li>Check for z-score (Z) and impute if missing</li>
<li>Check for N and impute if missing</li>
<li>Check output format is LDSC ready</li>
<li>Check and perform liftover to desired reference genome if necessary</li>
</ul>
<p>Users can specify which checks to run on their data. A <strong>note</strong> on the allele flipping check: <strong>MungeSumstats</strong> infers the effect allele will always be the A2 allele, this is the approach done for <a href="https://www.ncbi.nlm.nih.gov/pmc/articles/PMC7805039/">IEU GWAS VCF</a> and has such also been adopted here. This inference is first from the inputted file’s column headers however, the allele flipping check ensures this by comparing A1, what should be the reference allele, to the reference genome. If a SNP’s A1 DNA base doesn’t match the reference genome but it’s A2 (what should be the alternative allele) does, the alleles will be flipped along with the effect information (e.g. Beta, Odds Ratio, signed summary statistics, FRQ, Z-score*).</p>
<p>*-by default the Z-score is assumed to be calculated off the effect size not the P-value and so will be flipped. This can be changed by a user.</p>
<p>If a test is failed, the user will be notified and if possible, the input will be corrected. The QC steps from the checks above can also be adjusted to suit the user’s analysis, see <code><a href="../reference/format_sumstats.html">MungeSumstats::format_sumstats</a></code>.</p>
<p><em>MungeSumstats</em> can handle VCF, txt, tsv, csv file types or .gz/.bgz versions of these file types. The package also gives the user the flexibility to export the reformatted file as tab-delimited, VCF or R native objects such as data.table, GRanges or VRanges objects. The output can also be outputted in an <strong>LDSC ready</strong> format which means the file can be fed directly into LDSC without the need for additional munging.</p>
</div>
<div id="data" class="section level1">
<h1 class="hasAnchor">
<a href="#data" class="anchor"></a>Data</h1>
<p>The <em>MungeSumstats</em> package contains small subsets of GWAS summary statistics files. Firstly, on Educational Attainment by Okbay et al 2016: PMID: 27898078 PMCID: PMC5509058 DOI: 10.1038/ng1216-1587b.</p>
<p>Secondly, a VCF file (VCFv4.2) relating to the GWAS Amyotrophic lateral sclerosis from ieu open GWAS project. Dataset: ebi-a-GCST005647: <a href="https://gwas.mrcieu.ac.uk/datasets/ebi-a-GCST005647/" class="uri">https://gwas.mrcieu.ac.uk/datasets/ebi-a-GCST005647/</a></p>
<p>These datasets will be used to showcase <em>MungeSumstats</em> functionality.</p>
</div>
<div id="running-mungesumstats" class="section level1">
<h1 class="hasAnchor">
<a href="#running-mungesumstats" class="anchor"></a>Running <em>MungeSumstats</em>
</h1>
<p><em>MungeSumstats</em> is available on Bioconductor. To install the package on Bioconductor run the following lines of code:</p>
<pre><code>if (!require("BiocManager"))
    install.packages("BiocManager")

BiocManager::install("MungeSumstats")</code></pre>
<p>Once installed, load the package:</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/neurogenomics/MungeSumstats">MungeSumstats</a></span><span class="op">)</span></code></pre></div>
<p>To standardise the summary statistics’ file format, simply call <code><a href="../reference/format_sumstats.html">format_sumstats()</a></code> passing in the path to your summary statistics file or directly pass the summary statistics as a dataframe or datatable. You can specify which genome build was used in the GWAS(GRCh37 or GRCh38) or, as default, infer the genome build from the data.The reference genome is used for multiple checks like deriving missing data such SNP/BP/CHR/A1/A2 and for QC steps like removing non-biallelic SNPs, strand-ambiguous SNPs or ensuring correct allele and direction of SNP effects. The path to the reformatted summary statistics file can be returned by the function call, the user can specify a location to save the file or the user can return an R native object for the data: data.table, VRanges or GRanges object.</p>
<p>Note that for a number of the checks implored by <em>MungeSumstats</em> a reference genome is used. If your GWAS summary statistics file of interest relates to <em>GRCh38</em>, you will need to install <code>SNPlocs.Hsapiens.dbSNP144.GRCh38</code> and <code>BSgenome.Hsapiens.NCBI.GRCh38</code> from Bioconductor as follows:</p>
<pre><code><a href="https://rdrr.io/pkg/BiocManager/man/install.html">BiocManager::install("SNPlocs.Hsapiens.dbSNP144.GRCh38")
BiocManager::install("BSgenome.Hsapiens.NCBI.GRCh38")</a></code></pre>
<p>If your GWAS summary statistics file of interest relates to <em>GRCh37</em>, you will need to install <code>SNPlocs.Hsapiens.dbSNP144.GRCh37</code> and <code>BSgenome.Hsapiens.1000genomes.hs37d5</code> from Bioconductor as follows:</p>
<pre><code><a href="https://rdrr.io/pkg/BiocManager/man/install.html">BiocManager::install("SNPlocs.Hsapiens.dbSNP144.GRCh37")
BiocManager::install("BSgenome.Hsapiens.1000genomes.hs37d5")</a></code></pre>
<p>These may take some time to install and are not included in the package as some users may only need one of <em>GRCh37</em>/<em>GRCh38</em>.</p>
<p>The Educational Attainment by Okbay GWAS summary statistics file is saved as a text document in the package’s external data folder so we can just pass the file path to it straight to <em>MungeSumstats</em>.</p>
<p><strong>NOTE</strong> - By default, Formatted results will be saved to <code><a href="https://rdrr.io/r/base/tempfile.html">tempdir()</a></code>. This means all formatted summary stats will be deleted upon ending the R session if not copied to a local file path. Otherwise, to keep formatted summary stats, change <code>save_path</code> ( e.g.<code><a href="https://rdrr.io/r/base/file.path.html">file.path('./formatted',basename(path))</a></code>), or make sure to copy files elsewhere after processing ( e.g.<code><a href="https://rdrr.io/r/base/files.html">file.copy(save_path, './formatted/' )</a></code>.</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">eduAttainOkbayPth</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html">system.file</a></span><span class="op">(</span><span class="st">"extdata"</span>,<span class="st">"eduAttainOkbay.txt"</span>,
                                  package<span class="op">=</span><span class="st">"MungeSumstats"</span><span class="op">)</span>
<span class="va">reformatted</span> <span class="op">&lt;-</span> 
  <span class="fu">MungeSumstats</span><span class="fu">::</span><span class="fu"><a href="../reference/format_sumstats.html">format_sumstats</a></span><span class="op">(</span>path<span class="op">=</span><span class="va">eduAttainOkbayPth</span>,
                                 ref_genome<span class="op">=</span><span class="st">"GRCh37"</span><span class="op">)</span></code></pre></div>
<pre><code>## 
## 
## ******::NOTE::******
## - Formatted results will be saved to `tempdir()` by default.
## - This means all formatted summary stats will be deletedupon ending the R session.
## - To keep formatted summary stats, change `save_path` ( e.g. `save_path=file.path('./formatted',basename(path))` ),  or make sure to copy files elsewhere after processing ( e.g. `file.copy(save_path, './formatted/' )`.
## ********************</code></pre>
<pre><code>## Formatted summary statistics will be saved to ==&gt; /var/folders/hd/jm8lzp7s4dl_wlkykzhz66x80000gn/T//RtmpG1GXnw/file68271e8bbad2.tsv.gz</code></pre>
<pre><code>## Reading header.</code></pre>
<pre><code>## Tabular format detected.</code></pre>
<pre><code>## Importing tabular file: /private/var/folders/hd/jm8lzp7s4dl_wlkykzhz66x80000gn/T/RtmpQvPXso/temp_libpath5f95645cb8/MungeSumstats/extdata/eduAttainOkbay.txt</code></pre>
<pre><code>## Standardising column headers.</code></pre>
<pre><code>## First line of summary statistics file:</code></pre>
<pre><code>## MarkerName   CHR POS A1  A2  EAF Beta    SE  Pval    </code></pre>
<pre><code>## Summary statistics report:
##    - 93 rows
##    - 93 unique variants
##    - 70 genome-wide significant variants (P&lt;5e-8)
##    - 20 chromosomes</code></pre>
<pre><code>## Checking for multi-GWAS.</code></pre>
<pre><code>## Checking for multiple RSIDs on one row.</code></pre>
<pre><code>## Checking SNP RSIDs.</code></pre>
<pre><code>## Checking for merged allele column.</code></pre>
<pre><code>## Ensuring all SNPs are on the reference genome.</code></pre>
<pre><code>## Loading reference genome data.</code></pre>
<pre><code>## Checking for correct direction of A1 (reference) and A2 (alternative allele).</code></pre>
<pre><code>## There are 46 SNPs where A1 doesn't match the reference genome.
## These will be flipped with their effect columns.</code></pre>
<pre><code>## Checking for missing data.</code></pre>
<pre><code>## Checking for duplicate columns.</code></pre>
<pre><code>## Checking for duplicate SNPs from SNP ID.</code></pre>
<pre><code>## Checking for SNPs with duplicated base-pair positions.</code></pre>
<pre><code>## Filtering SNPs, ensuring SE&gt;0.</code></pre>
<pre><code>## Ensuring all SNPs have N&lt;5 std dev above mean.</code></pre>
<pre><code>## Removing 'chr' prefix from CHR.</code></pre>
<pre><code>## Making X/Y CHR uppercase.</code></pre>
<pre><code>## Checking for bi-allelic SNPs.</code></pre>
<pre><code>## Warning: When method is an integer, must be &gt;0.</code></pre>
<pre><code>## 67 SNPs (72%) have FRQ values &gt; 0.5. Conventionally the FRQ column is intended to show the minor/effect allele frequency.
## The FRQ column was mapped from one of the following from the inputted  summary statistics file:
## FRQ, EAF, MAF, FRQ_U, F_U, FREQUENCY, FREQ, FREQ_TESTED_ALLELE, FRQ_TESTED_ALLELE, FREQ_EFFECT_ALLELE, FRQ_EFFECT_ALLELE, EFFECT_ALLELE_FREQUENCY, EFFECT_ALLELE_FREQ, EFFECT_ALLELE_FRQ, A1FREQ, A1FRQ, A2FREQ, A2FRQ, ALLELE_FREQUENCY, ALLELE_FREQ, ALLELE_FRQ, AF, MINOR_AF, EFFECT_AF, A2_AF, EFF_AF, ALT_AF, ALTERNATIVE_AF, INC_AF, A_2_AF, TESTED_AF</code></pre>
<pre><code>## As frq_is_maf=TRUE, the FRQ column will not be renamed. If the FRQ values were intended to represent major allele frequency,
## set frq_is_maf=FALSE to rename the column as MAJOR_ALLELE_FRQ and differentiate it from minor/effect allele frequency.</code></pre>
<pre><code>## Sorting coordinates.</code></pre>
<pre><code>## Writing in tabular format ==&gt; /var/folders/hd/jm8lzp7s4dl_wlkykzhz66x80000gn/T//RtmpG1GXnw/file68271e8bbad2.tsv.gz</code></pre>
<pre><code>## Summary statistics report:
##    - 93 rows (100% of original 93 rows)
##    - 93 unique variants
##    - 70 genome-wide significant variants (P&lt;5e-8)
##    - 20 chromosomes</code></pre>
<pre><code>## Successfully finished preparing sumstats file, preview:</code></pre>
<pre><code>## Reading header.</code></pre>
<pre><code>##           SNP CHR       BP A1 A2     FRQ   BETA    SE         P
## 1:   rs301800   1  8490603  T  C 0.17910  0.019 0.003 1.794e-08
## 2: rs11210860   1 43982527  G  A 0.63060 -0.017 0.003 2.359e-10
## 3: rs34305371   1 72733610  G  A 0.91231 -0.035 0.005 3.762e-14
## 4:  rs2568955   1 72762169  T  C 0.23690 -0.017 0.003 1.797e-08</code></pre>
<pre><code>## Returning path to saved data.</code></pre>
<p>The arguments <code>format_sumstats</code> in that control the level of QC conducted by <em>MungeSumstats</em> are:</p>
<ul>
<li>
<strong>convert_small_p</strong> Binary, should <code>p-values &lt; 5e-324</code> be converted to 0? Small p-values pass the R limit and can cause errors with LDSC/MAGMA and should be converted. Default is TRUE.</li>
<li>
<strong>convert_n_int</strong> Binary, if N (the number of samples) is not an integer, should this be rounded? Default is TRUE. analysis_trait If multiple traits were studied, name of the trait for analysis from the GWAS. Default is NULL</li>
<li>
<strong>INFO_filter</strong> 0-1 The minimum value permissible of the imputation information score (if present in sumstatsfile). Default 0.9</li>
<li>
<strong>FRQ_filter</strong> 0-1 The minimum value permissible of the frequency(FRQ) of the SNP (i.e. Allele Frequency (AF)) (if present in sumstats file). By default no filtering is done, i.e. value of 0.<br>
</li>
<li>
<strong>pos_se</strong> Binary Should the standard Error (SE) column be checked to ensure it is greater than 0? Those that are, are removed (if present in sumstats file). Default TRUE.</li>
<li>
<strong>effect_columns_nonzero</strong> Binary should the effect columns in the data BETA,OR (odds ratio),LOG_ODDS,SIGNED_SUMSTAT be checked to ensure no SNP=0. Those that do are removed(if present in sumstats file). Default TRUE.<br>
</li>
<li>
<strong>N_std</strong> Numeric, the number of standard deviations above the mean a SNP’s N is needed to be removed. Default is 5. <strong>N_dropNA</strong> controls whether the SNPs with a missing N value are dropped or not (Default is TRUE).</li>
<li>
<strong>rmv_chr</strong> vector or character The chromosomes on which the SNPs should be removed. Use NULL if no filtering necessary. Default is X, Y and mitochondrial.<strong>rmv_chrPrefix</strong> controls whether “chr”/“CHR” is removed from chromosome names (Default is TRUE).</li>
<li>
<strong>on_ref_genome</strong> Binary, should a check take place that all SNPs are on the reference genome by SNP ID. Any SNPs not on the reference genome, will be corrected from the reference genome (if possible) using the chromosome and base pair position data. Default is TRUE</li>
<li>
<strong>strand_ambig_filter</strong> Binary, should SNPs with strand-ambiguous alleles be removed. Default is FALSE</li>
<li>
<strong>allele_flip_check</strong> Binary, should the allele columns be checked against reference genome to infer if flipping is necessary. Default is TRUE. <strong>allele_flip_drop</strong> controls whether the SNPs for which neither their A1 or A2 base pair values match a reference genome be dropped. Default is TRUE. <strong>allele_flip_z</strong> controls whether the Z-score value should be flipped along with effect and FRQ columns (e.g. Beta). Default is TRUE. <strong>allele_flip_frq</strong> controls whether the frequency (FRQ) value should be flipped along with effect and Z-score columns (e.g. Beta). Default is TRUE.</li>
<li>
<strong>bi_allelic_filter</strong> Binary, should non-biallelic SNPs be removed. Default is TRUE</li>
<li>
<strong>snp_ids_are_rs_ids</strong> Binary, should the SNP IDs inputted be inferred as RS IDs or some arbitrary ID. Default is TRUE.<br>
</li>
<li>
<strong>frq_is_maf</strong> Binary, conventionally the FRQ column is intended to show the minor/effect allele frequency (MAF) but sometimes the major allele frequency can be inferred as the FRQ column. This logical variable indicates that the FRQ column should be renamed to MAJOR_ALLELE_FRQ if the frequency values appear to relate to the major allele i.e. &gt;0.5. By default mapping won’t occur i.e. is TRUE.</li>
<li>
<strong>sort_coordinates</strong> Whether to sort by coordinates of resulting sumstats.<br>
</li>
<li>
<strong>nThread</strong> Number of threads to use for parallel processes.</li>
<li>
<strong>write_vcf</strong> Whether to write as VCF (TRUE) or tabular file (FALSE). While <strong>tabix_index</strong> is a binary input for whether to index the formatted summary statistics with <a href="http://www.htslib.org/doc/tabix.html">tabix</a> for fast querying.</li>
<li>
<strong>return_data</strong> Return <code>data.table</code>, <code>GRanges</code> or <code>VRanges</code>directly to user. Otherwise, return the path to the save data. Default is FALSE.</li>
<li>
<strong>return_format</strong> If return_data is TRUE. Object type to be returned (“data.table”,“vranges”,“granges”).</li>
<li>
<strong>ldsc_format</strong> Ensure that output format meets all requirements to be passed directly into LDSC without the need for additional munging.</li>
<li>
<strong>log_folder_ind</strong> Should log files be stored containing all filtered out SNPs (separate file per filter). The data is outputted in the same format specified for the resulting sumstats file.</li>
<li>
<strong>log_mungesumstats_msgs</strong> Binary Should a log be stored containing all messages and errors printed by MungeSumstats in a run.</li>
<li>
<strong>imputation_ind</strong> Binary Should a column be added for each imputation step to show what SNPs have imputed values for differing fields. This includes a field denoting SNP allele flipping (flipped). On the flipped value, this denoted whether the alelles where switched based on MungeSumstats initial choice of A1, A2 from the input column headers and thus may not align with what the creator intended.<strong>Note</strong> these columns will be in the formatted summary statistics returned.</li>
<li>
<strong>log_folder</strong> File path to the directory for the log files and the log of MungeSumstats messages to be stored. Default is a temporary directory.</li>
<li>
<strong>force_new</strong> If a formatted file of the same names as  exists, formatting will be skipped and this file will be imported instead (default). Set  to override this.</li>
<li>
<strong>mapping_file</strong> MungeSumstats has a pre-defined column-name mapping file which should cover the most common column headers and their interpretations. However, if a column header that is in youf file is missing of the mapping we give is incorrect you can supply your own mapping file. Must be a 2 column dataframe with column names “Uncorrected” and “Corrected”. See <code><a href="https://rdrr.io/r/utils/data.html">data(sumstatsColHeaders)</a></code> for default mapping and necessary format.</li>
</ul>
<p>See <code>?MungeSumstats::format_sumstats()</code> for the full list of parameters to control MungeSumstats QC and standardisation steps.</p>
<p>VCF files can also be standardised to the same format as other summary statistic files. A subset of the Amyotrophic lateral sclerosis GWAS from the ieu open GWAS project (a .vcf file) has been added to <em>MungeSumstats</em> to demonstrate this functionality.Simply pass the path to the file in the same manner you would for other summary statistic files:</p>
<div class="sourceCode" id="cb42"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co">#save ALS GWAS from the ieu open GWAS project to a temp directory</span>
<span class="va">ALSvcfPth</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html">system.file</a></span><span class="op">(</span><span class="st">"extdata"</span>,<span class="st">"ALSvcf.vcf"</span>, package<span class="op">=</span><span class="st">"MungeSumstats"</span><span class="op">)</span>
<span class="co">#set a low INFO filter so we get a return</span>
<span class="va">reformatted_vcf</span> <span class="op">&lt;-</span> 
  <span class="fu">MungeSumstats</span><span class="fu">::</span><span class="fu"><a href="../reference/format_sumstats.html">format_sumstats</a></span><span class="op">(</span>path<span class="op">=</span><span class="va">ALSvcfPth</span>, 
                                 ref_genome<span class="op">=</span><span class="st">"GRCh37"</span>,
                                 INFO_filter<span class="op">=</span><span class="fl">0.1</span><span class="op">)</span></code></pre></div>
<pre><code>## 
## 
## ******::NOTE::******
## - Formatted results will be saved to `tempdir()` by default.
## - This means all formatted summary stats will be deletedupon ending the R session.
## - To keep formatted summary stats, change `save_path` ( e.g. `save_path=file.path('./formatted',basename(path))` ),  or make sure to copy files elsewhere after processing ( e.g. `file.copy(save_path, './formatted/' )`.
## ********************</code></pre>
<pre><code>## Formatted summary statistics will be saved to ==&gt; /var/folders/hd/jm8lzp7s4dl_wlkykzhz66x80000gn/T//RtmpG1GXnw/file6827482b0398.tsv.gz</code></pre>
<pre><code>## Reading header.</code></pre>
<pre><code>## Importing VCF file: /private/var/folders/hd/jm8lzp7s4dl_wlkykzhz66x80000gn/T/RtmpQvPXso/temp_libpath5f95645cb8/MungeSumstats/extdata/ALSvcf.vcf</code></pre>
<pre><code>## Reading VCF file.</code></pre>
<pre><code>## Standardising column headers.</code></pre>
<pre><code>## First line of summary statistics file:</code></pre>
<pre><code>## CHROM    POS ID  REF ALT QUAL    FILTER  INFO    FORMAT  EBI-a-GCST005647    </code></pre>
<pre><code>## Removing non-standard columns: QUAL, FILTER, FORMAT</code></pre>
<pre><code>## Parsing 'EBI-A-GCST005647' data column.</code></pre>
<pre><code>## 0 empty column(s) detected.</code></pre>
<pre><code>## VCF file has -log10 P-values, these will be converted to unadjusted p-values in the 'P' column.</code></pre>
<pre><code>## Formatting INFO column.</code></pre>
<pre><code>## INFO column is actually AF, it will be converted.</code></pre>
<pre><code>## Standardising column headers.</code></pre>
<pre><code>## First line of summary statistics file:</code></pre>
<pre><code>## CHR  BP  SNP A1  A2  INFO    ES  SE  LP  AF  ID  P   </code></pre>
<pre><code>## Summary statistics report:
##    - 101 rows
##    - 101 unique variants
##    - 0 genome-wide significant variants (P&lt;5e-8)
##    - 1 chromosomes</code></pre>
<pre><code>## Checking for multi-GWAS.</code></pre>
<pre><code>## Checking for multiple RSIDs on one row.</code></pre>
<pre><code>## Checking SNP RSIDs.</code></pre>
<pre><code>## Checking for merged allele column.</code></pre>
<pre><code>## Ensuring all SNPs are on the reference genome.</code></pre>
<pre><code>## Loading reference genome data.</code></pre>
<pre><code>## 14 SNPs are not on the reference genome. These will be corrected from the reference genome.</code></pre>
<pre><code>## Checking for correct direction of A1 (reference) and A2 (alternative allele).</code></pre>
<pre><code>## Reordering so first three column headers are SNP, CHR and BP in this order.</code></pre>
<pre><code>## Reordering so the fourth and fifth columns are A1 and A2.</code></pre>
<pre><code>## Checking for missing data.</code></pre>
<pre><code>## Checking for duplicate columns.</code></pre>
<pre><code>## Checking for duplicate SNPs from SNP ID.</code></pre>
<pre><code>## Checking for SNPs with duplicated base-pair positions.</code></pre>
<pre><code>## Filtering SNPs based on INFO score.</code></pre>
<pre><code>## 67 SNPs are below the INFO threshold of 0.1 and will be removed.</code></pre>
<pre><code>## Filtering SNPs, ensuring SE&gt;0.</code></pre>
<pre><code>## Ensuring all SNPs have N&lt;5 std dev above mean.</code></pre>
<pre><code>## Removing 'chr' prefix from CHR.</code></pre>
<pre><code>## Making X/Y CHR uppercase.</code></pre>
<pre><code>## Checking for bi-allelic SNPs.</code></pre>
<pre><code>## Warning: When method is an integer, must be &gt;0.</code></pre>
<pre><code>## 2 SNPs (7.7%) have FRQ values &gt; 0.5. Conventionally the FRQ column is intended to show the minor/effect allele frequency.
## The FRQ column was mapped from one of the following from the inputted  summary statistics file:
## FRQ, EAF, MAF, FRQ_U, F_U, FREQUENCY, FREQ, FREQ_TESTED_ALLELE, FRQ_TESTED_ALLELE, FREQ_EFFECT_ALLELE, FRQ_EFFECT_ALLELE, EFFECT_ALLELE_FREQUENCY, EFFECT_ALLELE_FREQ, EFFECT_ALLELE_FRQ, A1FREQ, A1FRQ, A2FREQ, A2FRQ, ALLELE_FREQUENCY, ALLELE_FREQ, ALLELE_FRQ, AF, MINOR_AF, EFFECT_AF, A2_AF, EFF_AF, ALT_AF, ALTERNATIVE_AF, INC_AF, A_2_AF, TESTED_AF</code></pre>
<pre><code>## As frq_is_maf=TRUE, the FRQ column will not be renamed. If the FRQ values were intended to represent major allele frequency,
## set frq_is_maf=FALSE to rename the column as MAJOR_ALLELE_FRQ and differentiate it from minor/effect allele frequency.</code></pre>
<pre><code>## Sorting coordinates.</code></pre>
<pre><code>## Writing in tabular format ==&gt; /var/folders/hd/jm8lzp7s4dl_wlkykzhz66x80000gn/T//RtmpG1GXnw/file6827482b0398.tsv.gz</code></pre>
<pre><code>## Summary statistics report:
##    - 26 rows (25.7% of original 101 rows)
##    - 26 unique variants
##    - 0 genome-wide significant variants (P&lt;5e-8)
##    - 1 chromosomes</code></pre>
<pre><code>## Successfully finished preparing sumstats file, preview:</code></pre>
<pre><code>## Reading header.</code></pre>
<pre><code>##            SNP CHR    BP A1 A2   INFO    BETA     SE       LP    FRQ
## 1:  rs58108140   1 10583  G  A 0.1589  0.0312 0.0393 0.369267 0.1589
## 2:    rs806731   1 30923  G  T 0.7843 -0.0114 0.0353 0.126854 0.7843
## 3: rs116400033   1 51479  T  A 0.1829  0.0711 0.0370 1.262410 0.1829
## 4: rs141149254   1 54490  G  A 0.1353  0.0963 0.0423 1.641870 0.1353
##             ID          P
## 1:  rs58108140 0.42730011
## 2:    rs806731 0.74669974
## 3: rs116400033 0.05464998
## 4: rs141149254 0.02281025</code></pre>
<pre><code>## Returning path to saved data.</code></pre>
<p>You can also get more information on the SNPs which have had data imputed or have been filtered out by <em>MungeSumstats</em> by using the <code>imputation_ind</code> and <code>log_folder_ind</code> parameters respectively. For example:</p>
<div class="sourceCode" id="cb92"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co">#set</span>
<span class="va">reformatted_vcf_2</span> <span class="op">&lt;-</span> 
  <span class="fu">MungeSumstats</span><span class="fu">::</span><span class="fu"><a href="../reference/format_sumstats.html">format_sumstats</a></span><span class="op">(</span>path<span class="op">=</span><span class="va">ALSvcfPth</span>,
                                 ref_genome<span class="op">=</span><span class="st">"GRCh37"</span>,
                                 INFO_filter <span class="op">=</span> <span class="fl">0.1</span>,
                                 log_folder_ind<span class="op">=</span><span class="cn">TRUE</span>,
                                 imputation_ind<span class="op">=</span><span class="cn">TRUE</span>,
                                 log_mungesumstats_msgs<span class="op">=</span><span class="cn">TRUE</span><span class="op">)</span></code></pre></div>
<pre><code>## 
## 
## ******::NOTE::******
## - Formatted results will be saved to `tempdir()` by default.
## - This means all formatted summary stats will be deletedupon ending the R session.
## - To keep formatted summary stats, change `save_path` ( e.g. `save_path=file.path('./formatted',basename(path))` ),  or make sure to copy files elsewhere after processing ( e.g. `file.copy(save_path, './formatted/' )`.
## ********************</code></pre>
<pre><code>## ******::NOTE::******
## - Log results will be saved to `tempdir()` by default.
## - This means all log data from the run will be deleted upon ending the R session.
## - To keep it, change `log_folder` to an actual directory (e.g. log_folder='./').
## ********************</code></pre>
<pre><code>## Formatted summary statistics will be saved to ==&gt; /var/folders/hd/jm8lzp7s4dl_wlkykzhz66x80000gn/T//RtmpG1GXnw/file68275b0ffd7d.tsv.gz</code></pre>
<pre><code>## Log data to be saved to ==&gt; /var/folders/hd/jm8lzp7s4dl_wlkykzhz66x80000gn/T//RtmpG1GXnw</code></pre>
<pre><code>## Saving output messages to:
## /var/folders/hd/jm8lzp7s4dl_wlkykzhz66x80000gn/T//RtmpG1GXnw/MungeSumstats_log_msg.txt
## Any runtime errors will be saved to:
## /var/folders/hd/jm8lzp7s4dl_wlkykzhz66x80000gn/T//RtmpG1GXnw/MungeSumstats_log_output.txt
## Messages will not be printed to terminal.</code></pre>
<pre><code>## Reading header.</code></pre>
<pre><code>## Importing VCF file: /private/var/folders/hd/jm8lzp7s4dl_wlkykzhz66x80000gn/T/RtmpQvPXso/temp_libpath5f95645cb8/MungeSumstats/extdata/ALSvcf.vcf</code></pre>
<pre><code>## Reading VCF file.</code></pre>
<pre><code>## Standardising column headers.</code></pre>
<pre><code>## First line of summary statistics file:</code></pre>
<pre><code>## CHROM    POS ID  REF ALT QUAL    FILTER  INFO    FORMAT  EBI-a-GCST005647    </code></pre>
<pre><code>## Removing non-standard columns: QUAL, FILTER, FORMAT</code></pre>
<pre><code>## Parsing 'EBI-A-GCST005647' data column.</code></pre>
<pre><code>## 0 empty column(s) detected.</code></pre>
<pre><code>## VCF file has -log10 P-values, these will be converted to unadjusted p-values in the 'P' column.</code></pre>
<pre><code>## Formatting INFO column.</code></pre>
<pre><code>## INFO column is actually AF, it will be converted.</code></pre>
<pre><code>## Standardising column headers.</code></pre>
<pre><code>## First line of summary statistics file:</code></pre>
<pre><code>## CHR  BP  SNP A1  A2  INFO    ES  SE  LP  AF  ID  P   </code></pre>
<pre><code>## Summary statistics report:
##    - 101 rows
##    - 101 unique variants
##    - 0 genome-wide significant variants (P&lt;5e-8)
##    - 1 chromosomes</code></pre>
<pre><code>## Checking for multi-GWAS.</code></pre>
<pre><code>## Checking for multiple RSIDs on one row.</code></pre>
<pre><code>## Checking SNP RSIDs.</code></pre>
<pre><code>## Checking for merged allele column.</code></pre>
<pre><code>## Ensuring all SNPs are on the reference genome.</code></pre>
<pre><code>## Loading reference genome data.</code></pre>
<pre><code>## 14 SNPs are not on the reference genome. These will be corrected from the reference genome.</code></pre>
<pre><code>## Writing in tabular format ==&gt; /var/folders/hd/jm8lzp7s4dl_wlkykzhz66x80000gn/T//RtmpG1GXnw/snp_not_found_from_chr_bp.tsv.gz</code></pre>
<pre><code>## Checking for correct direction of A1 (reference) and A2 (alternative allele).</code></pre>
<pre><code>## Reordering so first three column headers are SNP, CHR and BP in this order.</code></pre>
<pre><code>## Reordering so the fourth and fifth columns are A1 and A2.</code></pre>
<pre><code>## Checking for missing data.</code></pre>
<pre><code>## Checking for duplicate columns.</code></pre>
<pre><code>## Checking for duplicate SNPs from SNP ID.</code></pre>
<pre><code>## Checking for SNPs with duplicated base-pair positions.</code></pre>
<pre><code>## Filtering SNPs based on INFO score.</code></pre>
<pre><code>## 67 SNPs are below the INFO threshold of 0.1 and will be removed.</code></pre>
<pre><code>## Writing in tabular format ==&gt; /var/folders/hd/jm8lzp7s4dl_wlkykzhz66x80000gn/T//RtmpG1GXnw/info_filter.tsv.gz</code></pre>
<pre><code>## Filtering SNPs, ensuring SE&gt;0.</code></pre>
<pre><code>## Ensuring all SNPs have N&lt;5 std dev above mean.</code></pre>
<pre><code>## Removing 'chr' prefix from CHR.</code></pre>
<pre><code>## Making X/Y CHR uppercase.</code></pre>
<pre><code>## Checking for bi-allelic SNPs.</code></pre>
<pre><code>## Warning: When method is an integer, must be &gt;0.</code></pre>
<pre><code>## 2 SNPs (7.7%) have FRQ values &gt; 0.5. Conventionally the FRQ column is intended to show the minor/effect allele frequency.
## The FRQ column was mapped from one of the following from the inputted  summary statistics file:
## FRQ, EAF, MAF, FRQ_U, F_U, FREQUENCY, FREQ, FREQ_TESTED_ALLELE, FRQ_TESTED_ALLELE, FREQ_EFFECT_ALLELE, FRQ_EFFECT_ALLELE, EFFECT_ALLELE_FREQUENCY, EFFECT_ALLELE_FREQ, EFFECT_ALLELE_FRQ, A1FREQ, A1FRQ, A2FREQ, A2FRQ, ALLELE_FREQUENCY, ALLELE_FREQ, ALLELE_FRQ, AF, MINOR_AF, EFFECT_AF, A2_AF, EFF_AF, ALT_AF, ALTERNATIVE_AF, INC_AF, A_2_AF, TESTED_AF</code></pre>
<pre><code>## As frq_is_maf=TRUE, the FRQ column will not be renamed. If the FRQ values were intended to represent major allele frequency,
## set frq_is_maf=FALSE to rename the column as MAJOR_ALLELE_FRQ and differentiate it from minor/effect allele frequency.</code></pre>
<pre><code>## Sorting coordinates.</code></pre>
<pre><code>## Writing in tabular format ==&gt; /var/folders/hd/jm8lzp7s4dl_wlkykzhz66x80000gn/T//RtmpG1GXnw/file68275b0ffd7d.tsv.gz</code></pre>
<pre><code>## Summary statistics report:
##    - 26 rows (25.7% of original 101 rows)
##    - 26 unique variants
##    - 0 genome-wide significant variants (P&lt;5e-8)
##    - 1 chromosomes</code></pre>
<pre><code>## Successfully finished preparing sumstats file, preview:</code></pre>
<pre><code>## Reading header.</code></pre>
<pre><code>##            SNP CHR    BP A1 A2   INFO    BETA     SE       LP    FRQ
## 1:  rs58108140   1 10583  G  A 0.1589  0.0312 0.0393 0.369267 0.1589
## 2:    rs806731   1 30923  G  T 0.7843 -0.0114 0.0353 0.126854 0.7843
## 3: rs116400033   1 51479  T  A 0.1829  0.0711 0.0370 1.262410 0.1829
## 4: rs141149254   1 54490  G  A 0.1353  0.0963 0.0423 1.641870 0.1353
##             ID          P IMPUTATION_SNP
## 1:  rs58108140 0.42730011             NA
## 2:    rs806731 0.74669974             NA
## 3: rs116400033 0.05464998             NA
## 4: rs141149254 0.02281025             NA</code></pre>
<pre><code>## Returning path to saved data.</code></pre>
<p>Note that using <code>log_folder_ind</code> returns a list from <code>format_sumstats</code> which includes the file locations of the differing classes of removed SNPs. Using <code>log_mungesumstats_msgs</code> saves the messages to the console to a file which is returned in the same list. Note that not all the messages will also print to screen anymore when you set <code>log_mungesumstats_msgs</code>:</p>
<div class="sourceCode" id="cb147"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/base/names.html">names</a></span><span class="op">(</span><span class="va">reformatted_vcf_2</span><span class="op">)</span></code></pre></div>
<pre><code>## [1] "sumstats"  "log_files"</code></pre>
<p>A user can load a file to view the excluded SNPs:</p>
<div class="sourceCode" id="cb149"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu">data.table</span><span class="fu">::</span><span class="fu"><a href="https://Rdatatable.gitlab.io/data.table/reference/fread.html">fread</a></span><span class="op">(</span><span class="va">reformatted_vcf_2</span><span class="op">$</span><span class="va">log_files</span><span class="op">$</span><span class="va">info_filter</span><span class="op">)</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">10</span><span class="op">]</span></code></pre></div>
<pre><code>##             SNP CHR    BP A1 A2   INFO    BETA     SE        LP    FRQ
##  1: rs146477069   1 54421  A  G 0.0352 -0.0240 0.0830 0.1121020 0.0352
##  2: rs143174675   1 54753  T  G 0.0188 -0.0766 0.1108 0.3103360 0.0188
##  3: rs114420996   1 58814  G  A 0.0693  0.0096 0.0574 0.0618807 0.0693
##  4:  rs62637815   1 59040  T  C 0.0637  0.0265 0.0593 0.1837590 0.0637
##  5:  rs56992750   1 61462  T  A 0.0141  0.1307 0.1183 0.5699250 0.0141
##  6: rs181431124   1 64649  A  C 0.0224  0.0284 0.1060 0.1030330 0.0224
##  7:  rs12401368   1 66507  T  A 0.0232  0.0549 0.1025 0.2275320 0.0232
##  8: rs188023513   1 77462  G  A 0.0458  0.1022 0.0714 0.8173000 0.0458
##  9:  rs62641297   1 77874  G  A 0.0231  0.0595 0.0994 0.2601110 0.0231
## 10:  rs78385339   1 77961  G  A 0.0347  0.1147 0.0852 0.7490920 0.0347
##              ID         P IMPUTATION_SNP
##  1: rs146477069 0.7724991             NA
##  2: rs143174675 0.4894000             NA
##  3: rs114420996 0.8672001             NA
##  4:  rs62637815 0.6549995             NA
##  5:  rs56992750 0.2692000             NA
##  6: rs181431124 0.7888002             NA
##  7:  rs12401368 0.5921995             NA
##  8:   rs2462497 0.1523000           TRUE
##  9:  rs62641297 0.5494004             NA
## 10:  rs78385339 0.1782001             NA</code></pre>
<p>And we can see all these SNPs had lower INFO values that our set threshold.</p>
<p>The different types of exclusion which lead to the names are explained below:</p>
<ul>
<li>
<strong>snp_multi_rs_one_row</strong> - Where the SNP (RS ID) contained more than one RS ID.</li>
<li>
<strong>snp_missing_rs</strong> - Where the SNP (RS ID) was missing the rs prefix. Note that these are only removed when other snps have an rs prefix.</li>
<li>
<strong>snp_multi_colon</strong> - Where the SNP ID has mutliple colons (“:”) in one SNP.</li>
<li>
<strong>snp_not_found_from_bp_chr</strong> - Where the RS ID was attempted to be imputed from the CHR and BP (Base-Pair) information, using the reference genome, but wasn’t successful.</li>
<li>
<strong>chr_bp_not_found_from_snp</strong> - Where the CHR and BP (Base-Pair) was attempted to be imputed from the SNP (RS ID), using the reference genome, but wasn’t successful.</li>
<li>
<strong>alleles_not_found_from_snp</strong> - Where the alleles (A1 and/or A2) was attempted to be imputed from the SNP (RS ID), using the reference genome, but wasn’t successful.</li>
<li>
<strong>alleles_dont_match_ref_gen</strong> - Where the alleles (A1 and/or A2) don’t match what’s on the reference genome.</li>
<li>
<strong>missing_data</strong> - Where there is data missing across the inputted columns.</li>
<li>
<strong>dup_snp_id</strong> - Where the SNP ID is duplicated in the input.</li>
<li>
<strong>dup_base_pair_position</strong> - Where the base-pair position is duplicated in the input.</li>
<li>
<strong>info_filter</strong> - SNP INFO value below the specified threshold.</li>
<li>
<strong>se_neg</strong> - SNPs SE (Standard Error) value is 0 or negative.</li>
<li>
<strong>effect_col_zero</strong> - SNPs effect column(s) value is zero e.g. BETA=0.</li>
<li>
<strong>n_large</strong> - SNPs N is N standard deviations greater than the mean.</li>
<li>
<strong>n_null</strong> - SNPs N is null.</li>
<li>
<strong>chr_excl</strong> - SNP is on a chromosome specified to be excluded.</li>
<li>
<strong>snp_strand_ambiguous</strong> - SNP is strand ambiguous.</li>
<li>
<strong>snp_bi_allelic</strong> - SNP is not bi-allelic.</li>
<li>
<strong>MungeSumstats_log_msg</strong> - Text file of all messages to the console created during MungeSumstats run.</li>
<li>
<strong>MungeSumstats_log_output</strong> - Text file of all errors to the console created during MungeSumstats run.</li>
</ul>
<p>Note to export to another type such as R native objects including data.table, GRanges, VRanges or save as a VCF file, set <code>return_format</code>:</p>
<div class="sourceCode" id="cb151"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co">#set</span>
<span class="va">reformatted_vcf_2</span> <span class="op">&lt;-</span> 
  <span class="fu">MungeSumstats</span><span class="fu">::</span><span class="fu"><a href="../reference/format_sumstats.html">format_sumstats</a></span><span class="op">(</span>path<span class="op">=</span><span class="va">ALSvcfPth</span>,
                                 ref_genome<span class="op">=</span><span class="st">"GRCh37"</span>,
                                 INFO_filter <span class="op">=</span> <span class="fl">0.1</span>,
                                 log_folder_ind<span class="op">=</span><span class="cn">TRUE</span>,
                                 imputation_ind<span class="op">=</span><span class="cn">TRUE</span>,
                                 log_mungesumstats_msgs<span class="op">=</span><span class="cn">TRUE</span>,
                                 return_format<span class="op">=</span><span class="st">"GRanges"</span><span class="op">)</span></code></pre></div>
<p>See our publication for further discussion of these checks and options:</p>
<p><a href="https://www.biorxiv.org/content/10.1101/2021.06.21.449239v1">Murphy and Skene. MungeSumstats: A Bioconductor package for the standardisation and quality control of many GWAS summary statistics</a>.</p>
</div>
<div id="extra-functionality" class="section level1">
<h1 class="hasAnchor">
<a href="#extra-functionality" class="anchor"></a>Extra Functionality</h1>
<p><em>MungeSumstats</em> also contains a function to quickly infer the genome build of multiple summary statistic files. This can be called separately to <code><a href="../reference/format_sumstats.html">format_sumstats()</a></code> which is useful if you want to just quickly check the genome build:</p>
<div class="sourceCode" id="cb152"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># Pass path to Educational Attainment Okbay sumstat file to a temp directory</span>
<span class="va">eduAttainOkbayPth</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html">system.file</a></span><span class="op">(</span><span class="st">"extdata"</span>, <span class="st">"eduAttainOkbay.txt"</span>,
                                  package <span class="op">=</span> <span class="st">"MungeSumstats"</span><span class="op">)</span>
<span class="va">sumstats_list</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html">list</a></span><span class="op">(</span>ss1 <span class="op">=</span> <span class="va">eduAttainOkbayPth</span>, ss2 <span class="op">=</span> <span class="va">eduAttainOkbayPth</span><span class="op">)</span>

<span class="va">ref_genomes</span> <span class="op">&lt;-</span> <span class="fu">MungeSumstats</span><span class="fu">::</span><span class="fu"><a href="../reference/get_genome_builds.html">get_genome_builds</a></span><span class="op">(</span>sumstats_list <span class="op">=</span> <span class="va">sumstats_list</span><span class="op">)</span></code></pre></div>
</div>
<div id="future-enhancements" class="section level1">
<h1 class="hasAnchor">
<a href="#future-enhancements" class="anchor"></a>Future Enhancements</h1>
<p>The <em>MungeSumstats</em> package aims to be able to handle the most common summary statistic file formats including VCF. If your file can not be formatted by <em>MungeSumstats</em> feel free to report the bug on github: <a href="https://github.com/neurogenomics/MungeSumstats" class="uri">https://github.com/neurogenomics/MungeSumstats</a> along with your summary statistic file header.</p>
<p>We also encourage people to edit the code to resolve their particular issues too and are happy to incorporate these through pull requests on github. If your summary statistic file headers are not recognised by <em>MungeSumstats</em> but correspond to one of:</p>
<pre><code>SNP, BP, CHR, A1, A2, P, Z, OR, BETA, LOG_ODDS,
SIGNED_SUMSTAT, N, N_CAS, N_CON, NSTUDY, INFO or FRQ </code></pre>
<p>feel free to update the <code><a href="../reference/sumstatsColHeaders.html">MungeSumstats::sumstatsColHeaders</a></code> following the approach in the data.R file and add your mapping. Then use a pull request on github and we will incorporate this change into the package.</p>
<p>A note on <code><a href="../reference/sumstatsColHeaders.html">MungeSumstats::sumstatsColHeaders</a></code> for summary statistic files with A0/A1. The mapping in <code><a href="../reference/sumstatsColHeaders.html">MungeSumstats::sumstatsColHeaders</a></code> converts A0 to A*, this is a special case so that the code knows to map A0/A1 to A1/A2 (ref/alt). The special case is needed since ordinarily A1 refers to the reference not the alternative allele.</p>
</div>
<div id="further-functionality" class="section level1">
<h1 class="hasAnchor">
<a href="#further-functionality" class="anchor"></a>Further functionality</h1>
<p>See the <a href="https://neurogenomics.github.io/MungeSumstats/articles/OpenGWAS.html">Open GWAS vignette</a> for how MungeSumstats can be used along with data from the MRC IEU Open GWAS Project and also Mungesumstats’ functionality to handle lists of summary statistics files.</p>
</div>
<div id="session-information" class="section level1">
<h1 class="hasAnchor">
<a href="#session-information" class="anchor"></a>Session Information</h1>
<details><pre><code>## R version 4.1.0 (2021-05-18)
## Platform: x86_64-apple-darwin17.0 (64-bit)
## Running under: macOS Big Sur 10.16
## 
## Matrix products: default
## BLAS:   /Library/Frameworks/R.framework/Versions/4.1/Resources/lib/libRblas.dylib
## LAPACK: /Library/Frameworks/R.framework/Versions/4.1/Resources/lib/libRlapack.dylib
## 
## locale:
## [1] en_IE.UTF-8/en_IE.UTF-8/en_IE.UTF-8/C/en_IE.UTF-8/en_IE.UTF-8
## 
## attached base packages:
## [1] stats     graphics  grDevices utils     datasets  methods   base     
## 
## other attached packages:
## [1] MungeSumstats_1.1.29 BiocStyle_2.20.2    
## 
## loaded via a namespace (and not attached):
##   [1] bitops_1.0-7                               
##   [2] matrixStats_0.61.0                         
##   [3] fs_1.5.0                                   
##   [4] bit64_4.0.5                                
##   [5] filelock_1.0.2                             
##   [6] progress_1.2.2                             
##   [7] httr_1.4.2                                 
##   [8] googleAuthR_1.4.0                          
##   [9] rprojroot_2.0.2                            
##  [10] GenomeInfoDb_1.28.4                        
##  [11] tools_4.1.0                                
##  [12] bslib_0.3.0                                
##  [13] utf8_1.2.2                                 
##  [14] R6_2.5.1                                   
##  [15] DBI_1.1.1                                  
##  [16] BiocGenerics_0.38.0                        
##  [17] tidyselect_1.1.1                           
##  [18] prettyunits_1.1.1                          
##  [19] curl_4.3.2                                 
##  [20] bit_4.0.4                                  
##  [21] compiler_4.1.0                             
##  [22] cli_3.0.1                                  
##  [23] textshaping_0.3.5                          
##  [24] Biobase_2.52.0                             
##  [25] xml2_1.3.2                                 
##  [26] desc_1.4.0                                 
##  [27] DelayedArray_0.18.0                        
##  [28] rtracklayer_1.52.1                         
##  [29] bookdown_0.24                              
##  [30] sass_0.4.0                                 
##  [31] rappdirs_0.3.3                             
##  [32] pkgdown_1.6.1                              
##  [33] systemfonts_1.0.2                          
##  [34] stringr_1.4.0                              
##  [35] digest_0.6.28                              
##  [36] Rsamtools_2.8.0                            
##  [37] rmarkdown_2.11                             
##  [38] R.utils_2.11.0                             
##  [39] XVector_0.32.0                             
##  [40] BSgenome.Hsapiens.1000genomes.hs37d5_0.99.1
##  [41] pkgconfig_2.0.3                            
##  [42] htmltools_0.5.2                            
##  [43] MatrixGenerics_1.4.3                       
##  [44] dbplyr_2.1.1                               
##  [45] fastmap_1.1.0                              
##  [46] BSgenome_1.60.0                            
##  [47] rlang_0.4.11                               
##  [48] rstudioapi_0.13                            
##  [49] RSQLite_2.2.8                              
##  [50] jquerylib_0.1.4                            
##  [51] BiocIO_1.2.0                               
##  [52] generics_0.1.0                             
##  [53] jsonlite_1.7.2                             
##  [54] BiocParallel_1.26.2                        
##  [55] dplyr_1.0.7                                
##  [56] R.oo_1.24.0                                
##  [57] VariantAnnotation_1.38.0                   
##  [58] RCurl_1.98-1.5                             
##  [59] magrittr_2.0.1                             
##  [60] GenomeInfoDbData_1.2.6                     
##  [61] Matrix_1.3-4                               
##  [62] Rcpp_1.0.7                                 
##  [63] S4Vectors_0.30.1                           
##  [64] fansi_0.5.0                                
##  [65] lifecycle_1.0.1                            
##  [66] R.methodsS3_1.8.1                          
##  [67] stringi_1.7.4                              
##  [68] yaml_2.2.1                                 
##  [69] SummarizedExperiment_1.22.0                
##  [70] zlibbioc_1.38.0                            
##  [71] BiocFileCache_2.0.0                        
##  [72] grid_4.1.0                                 
##  [73] blob_1.2.2                                 
##  [74] parallel_4.1.0                             
##  [75] crayon_1.4.1                               
##  [76] lattice_0.20-45                            
##  [77] Biostrings_2.60.2                          
##  [78] GenomicFeatures_1.44.2                     
##  [79] hms_1.1.1                                  
##  [80] KEGGREST_1.32.0                            
##  [81] knitr_1.34                                 
##  [82] pillar_1.6.3                               
##  [83] GenomicRanges_1.44.0                       
##  [84] rjson_0.2.20                               
##  [85] biomaRt_2.48.3                             
##  [86] stats4_4.1.0                               
##  [87] XML_3.99-0.8                               
##  [88] glue_1.4.2                                 
##  [89] evaluate_0.14                              
##  [90] SNPlocs.Hsapiens.dbSNP144.GRCh37_0.99.20   
##  [91] data.table_1.14.0                          
##  [92] BiocManager_1.30.16                        
##  [93] png_0.1-7                                  
##  [94] vctrs_0.3.8                                
##  [95] purrr_0.3.4                                
##  [96] assertthat_0.2.1                           
##  [97] cachem_1.0.6                               
##  [98] xfun_0.26                                  
##  [99] restfulr_0.0.13                            
## [100] ragg_1.1.3                                 
## [101] gargle_1.2.0                               
## [102] tibble_3.1.4                               
## [103] GenomicAlignments_1.28.0                   
## [104] AnnotationDbi_1.54.1                       
## [105] memoise_2.0.0                              
## [106] IRanges_2.26.0                             
## [107] ellipsis_0.3.2</code></pre>
</details>
</div>
<div id="references" class="section level1 unnumbered">
<h1 class="hasAnchor">
<a href="#references" class="anchor"></a>References</h1>
<div id="refs" class="references csl-bib-body" line-spacing="2">
<div id="ref-Skene2018" class="csl-entry">
<div class="csl-left-margin">1. </div>
<div class="csl-right-inline">Nathan G. Skene, T. E. B., Julien Bryois. Genetic identification of brain cell types underlying schizophrenia. <em>Nature Genetics</em> (2018). doi:<a href="https://doi.org/10.1038/s41588-018-0129-5">10.1038/s41588-018-0129-5</a>
</div>
</div>
</div>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

        <nav id="toc" data-toggle="toc"><h2 data-toc-skip>Contents</h2>
    </nav>
</div>

</div>



      <footer><div class="copyright">
  <p>Developed by Alan Murphy, Brian Schilder, Nathan Skene.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.6.1.</p>
</div>

      </footer>
</div>

  


  </body>
</html>
