<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>OpenGWAS • MungeSumstats</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link href="https://cdnjs.cloudflare.com/ajax/libs/bootswatch/3.4.0/flatly/bootstrap.min.css" rel="stylesheet" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="OpenGWAS">
<meta property="og:description" content="MungeSumstats">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--><!-- Global site tag (gtag.js) - Google Analytics --><script async src="https://www.googletagmanager.com/gtag/js?id=UA-175929032-1"></script><script>
  window.dataLayer = window.dataLayer || [];
  function gtag(){dataLayer.push(arguments);}
  gtag('js', new Date());

  gtag('config', 'UA-175929032-1');
</script>
</head>
<body data-spy="scroll" data-target="#toc">
    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">MungeSumstats</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">1.1.1</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../index.html">
    <span class="fas fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../articles/MungeSumstats.html">Get started</a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/OpenGWAS.html">OpenGWAS</a>
    </li>
  </ul>
</li>
<li>
  <a href="../news/index.html">Changelog</a>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/neurogenomics/MungeSumstats/">
    <span class="fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><script src="OpenGWAS_files/header-attrs-2.9/header-attrs.js"></script><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>OpenGWAS</h1>
                        <h4 class="author">Brian M. Schilder</h4>
            
            <h4 class="date">2021-07-14</h4>
      
      <small class="dont-index">Source: <a href="https://github.com/neurogenomics/MungeSumstats/blob/master/vignettes/OpenGWAS.Rmd"><code>vignettes/OpenGWAS.Rmd</code></a></small>
      <div class="hidden name"><code>OpenGWAS.Rmd</code></div>

    </div>

    
    
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu">knitr</span><span class="fu">::</span><span class="va"><a href="https://rdrr.io/pkg/knitr/man/opts_chunk.html">opts_chunk</a></span><span class="op">$</span><span class="fu">set</span><span class="op">(</span>echo <span class="op">=</span> <span class="cn">T</span>, root.dir<span class="op">=</span><span class="fu">here</span><span class="fu">::</span><span class="fu"><a href="https://here.r-lib.org//reference/here.html">here</a></span><span class="op">(</span><span class="op">)</span><span class="op">)</span>
<span class="fu">knitr</span><span class="fu">::</span><span class="va"><a href="https://rdrr.io/pkg/knitr/man/opts_knit.html">opts_knit</a></span><span class="op">$</span><span class="fu">set</span><span class="op">(</span>root.dir<span class="op">=</span><span class="fu">here</span><span class="fu">::</span><span class="fu"><a href="https://here.r-lib.org//reference/here.html">here</a></span><span class="op">(</span><span class="op">)</span><span class="op">)</span>

<span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/neurogenomics/MungeSumstats">MungeSumstats</a></span><span class="op">)</span></code></pre></div>
<p>Query and import data from the MRC IEU <a href="https://gwas.mrcieu.ac.uk/">Open GWAS Project</a>.</p>
<div id="find-gwas-datasets" class="section level1">
<h1 class="hasAnchor">
<a href="#find-gwas-datasets" class="anchor"></a>Find GWAS datasets</h1>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co">#### Search for datasets ####</span>
<span class="va">metagwas</span> <span class="op">&lt;-</span> <span class="fu">MungeSumstats</span><span class="fu">::</span><span class="fu"><a href="../reference/find_sumstats.html">find_sumstats</a></span><span class="op">(</span>traits <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"parkinson"</span>,<span class="st">"alzheimer"</span><span class="op">)</span>, 
                                         min_sample_size <span class="op">=</span> <span class="fl">5000</span><span class="op">)</span></code></pre></div>
<pre><code>## Collecting metadata from Open GWAS.</code></pre>
<pre><code>## Filtering metadata by substring criteria.</code></pre>
<pre><code>## Filtering metadata by sample/case/control/SNP size criteria.</code></pre>
<pre><code>## Excluding sample/case/control size with NAs.</code></pre>
<pre><code>## Found 16 GWAS datasets matching search criteria across:
##    - 12 trait(s)
##    - 1 population(s)
##    - 3 category(ies)
##    - 2 subcategory(ies)
##    - 5 publication(s)
##    - 6 consortia(ium)
##    - 1 genome build(s)</code></pre>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/utils/head.html">head</a></span><span class="op">(</span><span class="va">metagwas</span><span class="op">)</span></code></pre></div>
<pre><code>## # A tibble: 6 x 22
##   id     trait    group_name  year consortium      author sex   population unit 
##   &lt;chr&gt;  &lt;chr&gt;    &lt;chr&gt;      &lt;int&gt; &lt;chr&gt;           &lt;chr&gt;  &lt;chr&gt; &lt;chr&gt;      &lt;chr&gt;
## 1 ieu-a… Alzheim… public      2013 IGAP            Lambe… Male… European   log …
## 2 ieu-b… Alzheim… public      2019 Alzheimer Dise… Kunkl… Male… European   NA   
## 3 ieu-a… Alzheim… public      2013 IGAP            Lambe… Male… European   log …
## 4 ebi-a… Alzheim… public      2013 NA              Lambe… NA    European   logOR
## 5 ebi-a… Family … public      2018 NA              Mario… NA    European   NA   
## 6 ukb-b… Illness… public      2018 MRC-IEU         Ben E… Male… European   SD   
## # … with 13 more variables: sample_size &lt;int&gt;, nsnp &lt;int&gt;, build &lt;chr&gt;,
## #   category &lt;chr&gt;, subcategory &lt;chr&gt;, ontology &lt;chr&gt;, note &lt;chr&gt;, mr &lt;int&gt;,
## #   pmid &lt;int&gt;, priority &lt;int&gt;, ncase &lt;int&gt;, ncontrol &lt;int&gt;, sd &lt;dbl&gt;</code></pre>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">ids</span> <span class="op">&lt;-</span> <span class="op">(</span><span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/arrange.html">arrange</a></span><span class="op">(</span><span class="va">metagwas</span>, <span class="va">nsnp</span><span class="op">)</span><span class="op">)</span><span class="op">$</span><span class="va">id</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">2</span><span class="op">]</span>    </code></pre></div>
</div>
<div id="import-full-results" class="section level1">
<h1 class="hasAnchor">
<a href="#import-full-results" class="anchor"></a>Import full results</h1>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">datasets</span> <span class="op">&lt;-</span> <span class="fu">MungeSumstats</span><span class="fu">::</span><span class="fu"><a href="../reference/import_sumstats.html">import_sumstats</a></span><span class="op">(</span>ids <span class="op">=</span> <span class="va">ids</span><span class="op">)</span></code></pre></div>
<pre><code>## Processing 2 datasets from Open GWAS.</code></pre>
<pre><code>## 
## ========== Processing dataset : ieu-a-298 ==========</code></pre>
<pre><code>## Downloading VCF ==&gt; /var/folders/zq/h7mtybc533b1qzkys_ttgpth0000gn/T//RtmpA6fkpm/ieu-a-298.vcf.gz</code></pre>
<pre><code>## Downloading with download.file.
## Downloading with download.file.</code></pre>
<pre><code>## Warning in utils::download.file(input_url, out_file): URL https://
## gwas.mrcieu.ac.uk/files/ieu-a-298/ieu-a-298.vcf.gz.tbi: cannot open
## destfile '/var/folders/zq/h7mtybc533b1qzkys_ttgpth0000gn/T//RtmpA6fkpm/ieu-
## a-298.vcf.gz.tbi/ieu-a-298.vcf.gz.tbi', reason 'No such file or directory'</code></pre>
<pre><code>## Warning in utils::download.file(input_url, out_file): download had nonzero exit
## status</code></pre>
<pre><code>## 
## 
## *****::WARNING::*****
## - Formatted results will be saved to `tempdir()` by default.
## - This means all formatted summary stats will be deleted upon ending the R session.
## - To keep formatted summary stats, change `save_path` ( e.g. `save_path=file.path('./formatted',basename(path))` ),
##   or make sure to copy files elsewhere after processing ( e.g. `file.copy(save_path, './formatted/' )`.
## *****::******::*****</code></pre>
<pre><code>## Formatted summary statistics will be saved to ==&gt; /var/folders/zq/h7mtybc533b1qzkys_ttgpth0000gn/T//RtmpA6fkpm/ieu-a-298.tsv.gz</code></pre>
<pre><code>## Reading header.</code></pre>
<pre><code>## Importing.</code></pre>
<pre><code>## Reading VCF file.</code></pre>
<pre><code>## Removing non-standard columns: QUAL, FILTER, FORMAT</code></pre>
<pre><code>## Parsing 'ieu-a-298' data column.</code></pre>
<pre><code>## 1 empty column(s) detected.</code></pre>
<pre><code>## VCF file has -log10 P-values, these will be converted to unadjusted p-values in the 'P' column.</code></pre>
<pre><code>## Formatting INFO column.</code></pre>
<pre><code>## WARNING: All INFO scores are empty. Replacing all with 1.</code></pre>
<pre><code>## Standardising column headers.</code></pre>
<pre><code>## First line of summary statistics file:</code></pre>
<pre><code>## CHROM    POS ID  REF ALT INFO    ES  SE  LP  SS  P   </code></pre>
<pre><code>## Summary statistics report:
##    - 10,684 rows
##    - 10,684 unique variants
##    - 553 genome-wide significant variants (P&lt;5e-8)
##    - 22 chromosomes</code></pre>
<pre><code>## Checking for multi-GWAS.</code></pre>
<pre><code>## Inferring genome build.</code></pre>
<pre><code>## Inferred genome build: GRCH37</code></pre>
<pre><code>## Checking SNP RSIDs.</code></pre>
<pre><code>## Checking for merged allele column.</code></pre>
<pre><code>## Ensuring CHR and BP aren't missing.</code></pre>
<pre><code>## Checking SNP RSIDs.</code></pre>
<pre><code>## Ensuring all SNPs are on the reference genome.</code></pre>
<pre><code>## 273 SNPs are not on the reference genome. These will be corrected from the reference genome.</code></pre>
<pre><code>## Checking SNP RSIDs.</code></pre>
<pre><code>## Reordering so first three column headers are SNP, CHR and BP in this order.</code></pre>
<pre><code>## Checking for missing data.</code></pre>
<pre><code>## Checking for duplicate columns.</code></pre>
<pre><code>## Checking for p-vales &lt; 5e-324.</code></pre>
<pre><code>## Ensuring that the N column is all integers.</code></pre>
<pre><code>## The sumstats N column is not all integers, this could effect downstream analysis. These will be converted to integers.</code></pre>
<pre><code>## Ensuring all SNPs have N&lt;5 std dev below mean.</code></pre>
<pre><code>## Making all CHR uppercase.</code></pre>
<pre><code>## 326 SNPs are non-biallelic. These will be removed.</code></pre>
<pre><code>## Sorting coordinates.</code></pre>
<pre><code>## Writing in tabular format ==&gt; /var/folders/zq/h7mtybc533b1qzkys_ttgpth0000gn/T//RtmpA6fkpm/ieu-a-298.tsv.gz</code></pre>
<pre><code>## Summary statistics report:
##    - 10,085 rows (94.4% of original 10,684 rows)
##    - 10,085 unique variants
##    - 478 genome-wide significant variants (P&lt;5e-8)
##    - 22 chromosomes</code></pre>
<pre><code>## Succesfully finished preparing sumstats file, preview:</code></pre>
<pre><code>##           SNP CHR      BP A1 A2 INFO   BETA     SE      LP     N           P
## 1: rs76805690   1 2256288  C  A    1 0.0389 0.0175 1.58536 74046 0.025980051
## 2: rs75379543   1 2261983  C  A    1 0.0427 0.0167 1.96738 74046 0.010780031
## 3: rs75273719   1 2263666  G  A    1 0.0502 0.0171 2.47353 74046 0.003361012
## 4:   rs903904   1 2263888  C  T    1 0.0413 0.0169 1.82769 74046 0.014869967
## 5: rs80107522   1 2265070  T  C    1 0.0446 0.0168 2.11328 74046 0.007704066</code></pre>
<pre><code>## Returning path to saved data.</code></pre>
<pre><code>## 
## ieu-a-298 : Done in 1.28 minutes.</code></pre>
<pre><code>## 
## ========== Processing dataset : ieu-a-812 ==========</code></pre>
<pre><code>## Downloading VCF ==&gt; /var/folders/zq/h7mtybc533b1qzkys_ttgpth0000gn/T//RtmpA6fkpm/ieu-a-812.vcf.gz</code></pre>
<pre><code>## Downloading with download.file.
## Downloading with download.file.</code></pre>
<pre><code>## Warning in utils::download.file(input_url, out_file): URL https://
## gwas.mrcieu.ac.uk/files/ieu-a-812/ieu-a-812.vcf.gz.tbi: cannot open
## destfile '/var/folders/zq/h7mtybc533b1qzkys_ttgpth0000gn/T//RtmpA6fkpm/ieu-
## a-812.vcf.gz.tbi/ieu-a-812.vcf.gz.tbi', reason 'No such file or directory'

## Warning in utils::download.file(input_url, out_file): download had nonzero exit
## status</code></pre>
<pre><code>## 
## 
## *****::WARNING::*****
## - Formatted results will be saved to `tempdir()` by default.
## - This means all formatted summary stats will be deleted upon ending the R session.
## - To keep formatted summary stats, change `save_path` ( e.g. `save_path=file.path('./formatted',basename(path))` ),
##   or make sure to copy files elsewhere after processing ( e.g. `file.copy(save_path, './formatted/' )`.
## *****::******::*****</code></pre>
<pre><code>## Formatted summary statistics will be saved to ==&gt; /var/folders/zq/h7mtybc533b1qzkys_ttgpth0000gn/T//RtmpA6fkpm/ieu-a-812.tsv.gz</code></pre>
<pre><code>## Reading header.</code></pre>
<pre><code>## Importing.</code></pre>
<pre><code>## Reading VCF file.</code></pre>
<pre><code>## Removing non-standard columns: QUAL, FILTER, FORMAT</code></pre>
<pre><code>## Parsing 'ieu-a-812' data column.</code></pre>
<pre><code>## 2 empty column(s) detected.</code></pre>
<pre><code>## VCF file has -log10 P-values, these will be converted to unadjusted p-values in the 'P' column.</code></pre>
<pre><code>## Formatting INFO column.</code></pre>
<pre><code>## WARNING: All INFO scores are empty. Replacing all with 1.</code></pre>
<pre><code>## Standardising column headers.</code></pre>
<pre><code>## First line of summary statistics file:</code></pre>
<pre><code>## CHROM    POS ID  REF ALT INFO    ES  SE  LP  SS  P   </code></pre>
<pre><code>## Summary statistics report:
##    - 453,110 rows
##    - 453,110 unique variants
##    - 6 genome-wide significant variants (P&lt;5e-8)
##    - 22 chromosomes</code></pre>
<pre><code>## Checking for multi-GWAS.</code></pre>
<pre><code>## Inferring genome build.</code></pre>
<pre><code>## Inferred genome build: GRCH37</code></pre>
<pre><code>## Checking SNP RSIDs.</code></pre>
<pre><code>## Checking for merged allele column.</code></pre>
<pre><code>## Ensuring CHR and BP aren't missing.</code></pre>
<pre><code>## Checking SNP RSIDs.</code></pre>
<pre><code>## Ensuring all SNPs are on the reference genome.</code></pre>
<pre><code>## 1,456 SNPs are not on the reference genome. These will be corrected from the reference genome.</code></pre>
<pre><code>## Checking SNP RSIDs.</code></pre>
<pre><code>## Reordering so first three column headers are SNP, CHR and BP in this order.</code></pre>
<pre><code>## Checking for missing data.</code></pre>
<pre><code>## Checking for duplicate columns.</code></pre>
<pre><code>## Checking for p-vales &lt; 5e-324.</code></pre>
<pre><code>## Ensuring that the N column is all integers.</code></pre>
<pre><code>## The sumstats N column is not all integers, this could effect downstream analysis. These will be converted to integers.</code></pre>
<pre><code>## 514 base-pair positions are duplicated in the sumstats file. These duplicates will be removed</code></pre>
<pre><code>## Ensuring all SNPs have N&lt;5 std dev below mean.</code></pre>
<pre><code>## Making all CHR uppercase.</code></pre>
<pre><code>## 10,278 SNPs are non-biallelic. These will be removed.</code></pre>
<pre><code>## Sorting coordinates.</code></pre>
<pre><code>## Writing in tabular format ==&gt; /var/folders/zq/h7mtybc533b1qzkys_ttgpth0000gn/T//RtmpA6fkpm/ieu-a-812.tsv.gz</code></pre>
<pre><code>## Summary statistics report:
##    - 440,862 rows (97.3% of original 453,110 rows)
##    - 440,862 unique variants
##    - 6 genome-wide significant variants (P&lt;5e-8)
##    - 22 chromosomes</code></pre>
<pre><code>## Succesfully finished preparing sumstats file, preview:</code></pre>
<pre><code>##           SNP CHR      BP A1 A2 INFO       BETA        SE       LP    N
## 1: rs12562034   1  768448  G  A    1 -0.1789830 0.0632235 2.297570 5691
## 2:  rs3934834   1 1005806  C  T    1 -0.1070590 0.0555406 1.245730 5691
## 3:  rs9442372   1 1018704  A  G    1 -0.0246926 0.0411484 0.258770 5691
## 4:  rs3737728   1 1021415  A  G    1 -0.0580529 0.0456814 0.690157 5691
## 5:  rs6687776   1 1030565  C  T    1 -0.0459289 0.0540640 0.399354 5691
##              P
## 1: 0.005039994
## 2: 0.056789756
## 3: 0.551099478
## 4: 0.204099998
## 5: 0.398699784</code></pre>
<pre><code>## Returning path to saved data.</code></pre>
<pre><code>## 
## ieu-a-812 : Done in 1.26 minutes.</code></pre>
<div id="summarise-results" class="section level2">
<h2 class="hasAnchor">
<a href="#summarise-results" class="anchor"></a>Summarise results</h2>
<p>By default, <code>import_sumstats</code> results a named list where the names are the Open GWAS dataset IDs and the items are the respective paths to the formatted summary statistics.</p>
<div class="sourceCode" id="cb105"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/base/print.html">print</a></span><span class="op">(</span><span class="va">datasets</span><span class="op">)</span></code></pre></div>
<pre><code>## $`ieu-a-298`
## [1] "/var/folders/zq/h7mtybc533b1qzkys_ttgpth0000gn/T//RtmpA6fkpm/ieu-a-298.tsv.gz"
## 
## $`ieu-a-812`
## [1] "/var/folders/zq/h7mtybc533b1qzkys_ttgpth0000gn/T//RtmpA6fkpm/ieu-a-812.tsv.gz"</code></pre>
<p>You can easily turn this into a data.frame as well.</p>
<div class="sourceCode" id="cb107"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">results_df</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html">data.frame</a></span><span class="op">(</span>id<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/names.html">names</a></span><span class="op">(</span><span class="va">datasets</span><span class="op">)</span>, 
                         path<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/unlist.html">unlist</a></span><span class="op">(</span><span class="va">datasets</span><span class="op">)</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/base/print.html">print</a></span><span class="op">(</span><span class="va">results_df</span><span class="op">)</span></code></pre></div>
<pre><code>##                  id
## ieu-a-298 ieu-a-298
## ieu-a-812 ieu-a-812
##                                                                                    path
## ieu-a-298 /var/folders/zq/h7mtybc533b1qzkys_ttgpth0000gn/T//RtmpA6fkpm/ieu-a-298.tsv.gz
## ieu-a-812 /var/folders/zq/h7mtybc533b1qzkys_ttgpth0000gn/T//RtmpA6fkpm/ieu-a-812.tsv.gz</code></pre>
</div>
</div>
<div id="import-full-results-parallel" class="section level1">
<h1 class="hasAnchor">
<a href="#import-full-results-parallel" class="anchor"></a>Import full results (parallel)</h1>
<p><em>Optional</em>: Speed up with multi-threaded download via <a href="https://github.com/axel-download-accelerator/axel">axel</a>.</p>
<div class="sourceCode" id="cb109"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">datasets</span> <span class="op">&lt;-</span> <span class="fu">MungeSumstats</span><span class="fu">::</span><span class="fu"><a href="../reference/import_sumstats.html">import_sumstats</a></span><span class="op">(</span>ids <span class="op">=</span> <span class="va">ids</span>, 
                                           vcf_download <span class="op">=</span> <span class="cn">TRUE</span>, 
                                           download_method <span class="op">=</span> <span class="st">"axel"</span>, 
                                           nThread <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/Extremes.html">max</a></span><span class="op">(</span><span class="fl">2</span>,<span class="fu">future</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/future/man/re-exports.html">availableCores</a></span><span class="op">(</span><span class="op">)</span><span class="op">-</span><span class="fl">2</span><span class="op">)</span><span class="op">)</span></code></pre></div>
</div>
<div id="session-info" class="section level1">
<h1 class="hasAnchor">
<a href="#session-info" class="anchor"></a>Session Info</h1>
<details><div class="sourceCode" id="cb110"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu">utils</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/r/utils/sessionInfo.html">sessionInfo</a></span><span class="op">(</span><span class="op">)</span></code></pre></div>
<pre><code>## R version 4.1.0 (2021-05-18)
## Platform: x86_64-apple-darwin17.0 (64-bit)
## Running under: macOS Big Sur 10.16
## 
## Matrix products: default
## BLAS:   /Library/Frameworks/R.framework/Versions/4.1/Resources/lib/libRblas.dylib
## LAPACK: /Library/Frameworks/R.framework/Versions/4.1/Resources/lib/libRlapack.dylib
## 
## locale:
## [1] en_GB.UTF-8/en_GB.UTF-8/en_GB.UTF-8/C/en_GB.UTF-8/en_GB.UTF-8
## 
## attached base packages:
## [1] stats     graphics  grDevices utils     datasets  methods   base     
## 
## other attached packages:
## [1] MungeSumstats_1.1.1 BiocStyle_2.20.2   
## 
## loaded via a namespace (and not attached):
##   [1] bitops_1.0-7                               
##   [2] matrixStats_0.59.0                         
##   [3] fs_1.5.0                                   
##   [4] bit64_4.0.5                                
##   [5] filelock_1.0.2                             
##   [6] progress_1.2.2                             
##   [7] httr_1.4.2                                 
##   [8] rprojroot_2.0.2                            
##   [9] GenomeInfoDb_1.28.1                        
##  [10] tools_4.1.0                                
##  [11] bslib_0.2.5.1                              
##  [12] utf8_1.2.1                                 
##  [13] R6_2.5.0                                   
##  [14] DBI_1.1.1                                  
##  [15] BiocGenerics_0.38.0                        
##  [16] tidyselect_1.1.1                           
##  [17] prettyunits_1.1.1                          
##  [18] bit_4.0.4                                  
##  [19] curl_4.3.2                                 
##  [20] compiler_4.1.0                             
##  [21] cli_3.0.0                                  
##  [22] textshaping_0.3.5                          
##  [23] Biobase_2.52.0                             
##  [24] xml2_1.3.2                                 
##  [25] desc_1.3.0                                 
##  [26] DelayedArray_0.18.0                        
##  [27] rtracklayer_1.52.0                         
##  [28] bookdown_0.22                              
##  [29] sass_0.4.0                                 
##  [30] rappdirs_0.3.3                             
##  [31] pkgdown_1.6.1                              
##  [32] systemfonts_1.0.2                          
##  [33] stringr_1.4.0                              
##  [34] digest_0.6.27                              
##  [35] Rsamtools_2.8.0                            
##  [36] R.utils_2.10.1                             
##  [37] rmarkdown_2.9                              
##  [38] XVector_0.32.0                             
##  [39] BSgenome.Hsapiens.1000genomes.hs37d5_0.99.1
##  [40] pkgconfig_2.0.3                            
##  [41] htmltools_0.5.1.1                          
##  [42] MatrixGenerics_1.4.0                       
##  [43] dbplyr_2.1.1                               
##  [44] fastmap_1.1.0                              
##  [45] BSgenome_1.60.0                            
##  [46] rlang_0.4.11                               
##  [47] rstudioapi_0.13                            
##  [48] RSQLite_2.2.7                              
##  [49] jquerylib_0.1.4                            
##  [50] BiocIO_1.2.0                               
##  [51] generics_0.1.0                             
##  [52] jsonlite_1.7.2                             
##  [53] BiocParallel_1.26.1                        
##  [54] R.oo_1.24.0                                
##  [55] dplyr_1.0.7                                
##  [56] VariantAnnotation_1.38.0                   
##  [57] RCurl_1.98-1.3                             
##  [58] magrittr_2.0.1                             
##  [59] GenomeInfoDbData_1.2.6                     
##  [60] Matrix_1.3-3                               
##  [61] Rcpp_1.0.7                                 
##  [62] S4Vectors_0.30.0                           
##  [63] fansi_0.5.0                                
##  [64] R.methodsS3_1.8.1                          
##  [65] lifecycle_1.0.0                            
##  [66] stringi_1.6.2                              
##  [67] yaml_2.2.1                                 
##  [68] SummarizedExperiment_1.22.0                
##  [69] zlibbioc_1.38.0                            
##  [70] BiocFileCache_2.0.0                        
##  [71] grid_4.1.0                                 
##  [72] blob_1.2.1                                 
##  [73] parallel_4.1.0                             
##  [74] crayon_1.4.1                               
##  [75] lattice_0.20-44                            
##  [76] Biostrings_2.60.1                          
##  [77] GenomicFeatures_1.44.0                     
##  [78] hms_1.1.0                                  
##  [79] KEGGREST_1.32.0                            
##  [80] knitr_1.33                                 
##  [81] pillar_1.6.1                               
##  [82] GenomicRanges_1.44.0                       
##  [83] rjson_0.2.20                               
##  [84] BSgenome.Hsapiens.NCBI.GRCh38_1.3.1000     
##  [85] biomaRt_2.48.2                             
##  [86] stats4_4.1.0                               
##  [87] XML_3.99-0.6                               
##  [88] glue_1.4.2                                 
##  [89] evaluate_0.14                              
##  [90] SNPlocs.Hsapiens.dbSNP144.GRCh37_0.99.20   
##  [91] data.table_1.14.0                          
##  [92] BiocManager_1.30.16                        
##  [93] SNPlocs.Hsapiens.dbSNP144.GRCh38_0.99.20   
##  [94] png_0.1-7                                  
##  [95] vctrs_0.3.8                                
##  [96] purrr_0.3.4                                
##  [97] assertthat_0.2.1                           
##  [98] cachem_1.0.5                               
##  [99] xfun_0.24                                  
## [100] restfulr_0.0.13                            
## [101] ragg_1.1.3                                 
## [102] tibble_3.1.2                               
## [103] GenomicAlignments_1.28.0                   
## [104] AnnotationDbi_1.54.1                       
## [105] memoise_2.0.0                              
## [106] IRanges_2.26.0                             
## [107] ellipsis_0.3.2                             
## [108] here_1.0.1</code></pre>
</details>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

        <nav id="toc" data-toggle="toc"><h2 data-toc-skip>Contents</h2>
    </nav>
</div>

</div>



      <footer><div class="copyright">
  <p>Developed by Alan Murphy, Nathan Skene.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.6.1.</p>
</div>

      </footer>
</div>

  


  </body>
</html>
