<!-- Generated by pkgdown: do not edit by hand -->
<!DOCTYPE html>
<html lang="en">
  <head>
  <meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">

<title>Check that summary statistics from GWAS are in a homogeneous format — format_sumstats • MungeSumstats</title>


<!-- jquery -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script>
<!-- Bootstrap -->
<link href="https://cdnjs.cloudflare.com/ajax/libs/bootswatch/3.4.0/flatly/bootstrap.min.css" rel="stylesheet" crossorigin="anonymous" />


<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script>

<!-- bootstrap-toc -->
<link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script>

<!-- Font Awesome icons -->
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous" />
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous" />

<!-- clipboard.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script>

<!-- headroom.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script>

<!-- pkgdown -->
<link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script>




<meta property="og:title" content="Check that summary statistics from GWAS are in a homogeneous format — format_sumstats" />
<meta property="og:description" content="Check that summary statistics from GWAS are in a homogeneous format" />




<!-- mathjax -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script>

<!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->


<!-- Global site tag (gtag.js) - Google Analytics -->
<script async src="https://www.googletagmanager.com/gtag/js?id=UA-175929032-1"></script>
<script>
  window.dataLayer = window.dataLayer || [];
  function gtag(){dataLayer.push(arguments);}
  gtag('js', new Date());

  gtag('config', 'UA-175929032-1');
</script>


  </head>

  <body data-spy="scroll" data-target="#toc">
    <div class="container template-reference-topic">
      <header>
      <div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">MungeSumstats</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">0.99.29</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="../index.html">
    <span class="fas fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../articles/MungeSumstats.html">Get started</a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li>
  <a href="../news/index.html">Changelog</a>
</li>
      </ul>
      <ul class="nav navbar-nav navbar-right">
        <li>
  <a href="https://github.com/neurogenomics/MungeSumstats/">
    <span class="fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
      
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      

      </header>

<div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>Check that summary statistics from GWAS are in a homogeneous format</h1>
    <small class="dont-index">Source: <a href='https://github.com/neurogenomics/MungeSumstats/blob/master/R/format_sumstats.R'><code>R/format_sumstats.R</code></a></small>
    <div class="hidden name"><code>format_sumstats.Rd</code></div>
    </div>

    <div class="ref-description">
    <p>Check that summary statistics from GWAS are in a homogeneous format</p>
    </div>

    <pre class="usage"><span class='fu'>format_sumstats</span><span class='op'>(</span>
  <span class='va'>path</span>,
  ref_genome <span class='op'>=</span> <span class='st'>"GRCh37"</span>,
  convert_small_p <span class='op'>=</span> <span class='cn'>TRUE</span>,
  convert_n_int <span class='op'>=</span> <span class='cn'>TRUE</span>,
  analysis_trait <span class='op'>=</span> <span class='cn'>NULL</span>,
  INFO_filter <span class='op'>=</span> <span class='fl'>0.9</span>,
  N_std <span class='op'>=</span> <span class='fl'>5</span>,
  rmv_chr <span class='op'>=</span> <span class='fu'><a href='https://rdrr.io/r/base/c.html'>c</a></span><span class='op'>(</span><span class='st'>"X"</span>, <span class='st'>"Y"</span>, <span class='st'>"MT"</span><span class='op'>)</span>,
  on_ref_genome <span class='op'>=</span> <span class='cn'>TRUE</span>,
  strand_ambig_filter <span class='op'>=</span> <span class='cn'>FALSE</span>,
  allele_flip_check <span class='op'>=</span> <span class='cn'>TRUE</span>,
  bi_allelic_filter <span class='op'>=</span> <span class='cn'>TRUE</span>
<span class='op'>)</span></pre>

    <h2 class="hasAnchor" id="arguments"><a class="anchor" href="#arguments"></a>Arguments</h2>
    <table class="ref-arguments">
    <colgroup><col class="name" /><col class="desc" /></colgroup>
    <tr>
      <th>path</th>
      <td><p>Filepath for the summary statistics file to be formatted</p></td>
    </tr>
    <tr>
      <th>ref_genome</th>
      <td><p>name of the reference genome used for the GWAS (GRCh37 or GRCh38). Default is GRCh37.</p></td>
    </tr>
    <tr>
      <th>convert_small_p</th>
      <td><p>Binary, should p-values &lt; 5e-324 be converted to 0? Small p-values pass the R limit and can cause errors with LDSC/MAGMA and should be converted. Default is TRUE.</p></td>
    </tr>
    <tr>
      <th>convert_n_int</th>
      <td><p>Binary, if N (the number of samples) is not an integer, should this be rounded? Default is TRUE.</p></td>
    </tr>
    <tr>
      <th>analysis_trait</th>
      <td><p>If multiple traits were studied, name of the trait for analysis from the GWAS. Default is NULL</p></td>
    </tr>
    <tr>
      <th>INFO_filter</th>
      <td><p>numeric The minimum value permissible of the imputation information score (if present in sumstatsfile). Default 0.9</p></td>
    </tr>
    <tr>
      <th>N_std</th>
      <td><p>numeric The number of standard deviations above the mean a SNP's N is needed to be removed. Default is 5.</p></td>
    </tr>
    <tr>
      <th>rmv_chr</th>
      <td><p>vector or character The chromosomes on which the SNPs should be removed. Use NULL if no filtering necessary. Default is X, Y and mitochondrial.</p></td>
    </tr>
    <tr>
      <th>on_ref_genome</th>
      <td><p>Binary Should a check take place that all SNPs are on the reference genome by SNP ID. Default is TRUE</p></td>
    </tr>
    <tr>
      <th>strand_ambig_filter</th>
      <td><p>Binary Should SNPs with strand-ambiguous alleles be removed. Default is FALSE</p></td>
    </tr>
    <tr>
      <th>allele_flip_check</th>
      <td><p>Binary Should the allele columns be checked against reference genome to infer if flipping is necessary. Default is TRUE</p></td>
    </tr>
    <tr>
      <th>bi_allelic_filter</th>
      <td><p>Binary Should non-biallelic SNPs be removed. Default is TRUE</p></td>
    </tr>
    </table>

    <h2 class="hasAnchor" id="value"><a class="anchor" href="#value"></a>Value</h2>

    <p>The address for the modified sumstats file</p>

    <h2 class="hasAnchor" id="examples"><a class="anchor" href="#examples"></a>Examples</h2>
    <pre class="examples"><div class='input'><span class='co'>#Pass path to Educational Attainment Okbay sumstat file to a temp directory</span>
<span class='va'>eduAttainOkbayPth</span> <span class='op'>&lt;-</span> <span class='fu'><a href='https://rdrr.io/r/base/system.file.html'>system.file</a></span><span class='op'>(</span><span class='st'>"extdata"</span>,<span class='st'>"eduAttainOkbay.txt"</span>,
package<span class='op'>=</span><span class='st'>"MungeSumstats"</span><span class='op'>)</span>
<span class='co'>#pass path to format_sumstats</span>
<span class='co'>## Call uses reference genome as default with more than 2GB of memory,</span>
<span class='co'>## which is more than what 32-bit Windows can handle so remove certain checks</span>
<span class='va'>is_32bit_windows</span> <span class='op'>&lt;-</span> <span class='va'>.Platform</span><span class='op'>$</span><span class='va'>OS.type</span> <span class='op'>==</span> <span class='st'>"windows"</span> <span class='op'>&amp;&amp;</span> <span class='va'>.Platform</span><span class='op'>$</span><span class='va'>r_arch</span> <span class='op'>==</span> <span class='st'>"i386"</span>
<span class='kw'>if</span> <span class='op'>(</span><span class='op'>!</span><span class='va'>is_32bit_windows</span><span class='op'>)</span> <span class='op'>{</span>
<span class='va'>reformatted</span> <span class='op'>&lt;-</span> <span class='fu'>MungeSumstats</span><span class='fu'>::</span><span class='fu'>format_sumstats</span><span class='op'>(</span><span class='va'>eduAttainOkbayPth</span>,
ref_genome<span class='op'>=</span><span class='st'>"GRCh37"</span><span class='op'>)</span>
<span class='op'>}</span> <span class='kw'>else</span><span class='op'>{</span>
<span class='va'>reformatted</span> <span class='op'>&lt;-</span> <span class='fu'>MungeSumstats</span><span class='fu'>::</span><span class='fu'>format_sumstats</span><span class='op'>(</span><span class='va'>eduAttainOkbayPth</span>,
ref_genome<span class='op'>=</span><span class='st'>"GRCh37"</span>,on_ref_genome <span class='op'>=</span> <span class='cn'>FALSE</span>,strand_ambig_filter<span class='op'>=</span><span class='cn'>FALSE</span>,
bi_allelic_filter<span class='op'>=</span><span class='cn'>FALSE</span>,
allele_flip_check<span class='op'>=</span><span class='cn'>FALSE</span><span class='op'>)</span>
<span class='op'>}</span>
</div><div class='output co'>#&gt; <span class='message'>First line of summary statistics file: </span></div><div class='output co'>#&gt; <span class='message'>MarkerName	CHR	POS	A1	A2	EAF	Beta	SE	Pval	</span></div><div class='output co'>#&gt; <span class='message'>Succesfully finished preparing sumstats file, preview:</span></div><div class='output co'>#&gt; <span class='message'>SNP	CHR	BP	A1	A2	FRQ	BETA	SE	P</span></div><div class='output co'>#&gt; <span class='message'>rs12646808	4	3249828	T	C	0.6418	0.016	0.003	4.002e-08</span></div><div class='output co'>#&gt; <span class='message'>rs301800	1	8490603	T	C	0.1791	0.019	0.003	1.794e-08</span></div><div class='input'><span class='co'>#returned location has the updated summary statistics file</span>
</div></pre>
  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <nav id="toc" data-toggle="toc" class="sticky-top">
      <h2 data-toc-skip>Contents</h2>
    </nav>
  </div>
</div>


      <footer>
      <div class="copyright">
  <p>Developed by Alan Murphy, Nathan Skene.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.6.1.</p>
</div>

      </footer>
   </div>

  


  </body>
</html>


