<!-- Generated by pkgdown: do not edit by hand -->
<!DOCTYPE html>
<html lang="en">
  <head>
  <meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">

<title>Import full genome-wide GWAS summary statistics from Open GWAS — import_sumstats • MungeSumstats</title>


<!-- jquery -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script>
<!-- Bootstrap -->
<link href="https://cdnjs.cloudflare.com/ajax/libs/bootswatch/3.4.0/flatly/bootstrap.min.css" rel="stylesheet" crossorigin="anonymous" />


<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script>

<!-- bootstrap-toc -->
<link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script>

<!-- Font Awesome icons -->
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous" />
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous" />

<!-- clipboard.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script>

<!-- headroom.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script>

<!-- pkgdown -->
<link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script>




<meta property="og:title" content="Import full genome-wide GWAS summary statistics from Open GWAS — import_sumstats" />
<meta property="og:description" content="Requires internet access to run." />




<!-- mathjax -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script>

<!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->


<!-- Global site tag (gtag.js) - Google Analytics -->
<script async src="https://www.googletagmanager.com/gtag/js?id=UA-175929032-1"></script>
<script>
  window.dataLayer = window.dataLayer || [];
  function gtag(){dataLayer.push(arguments);}
  gtag('js', new Date());

  gtag('config', 'UA-175929032-1');
</script>


  </head>

  <body data-spy="scroll" data-target="#toc">
    <div class="container template-reference-topic">
      <header>
      <div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">MungeSumstats</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">1.1.21</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="../index.html">
    <span class="fas fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../articles/MungeSumstats.html">Get started</a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
    <li>
      <a href="../articles/OpenGWAS.html">Import GWAS summary statistics from Open GWAS</a>
    </li>
  </ul>
</li>
<li>
  <a href="../news/index.html">Changelog</a>
</li>
      </ul>
      <ul class="nav navbar-nav navbar-right">
        <li>
  <a href="https://github.com/neurogenomics/MungeSumstats/">
    <span class="fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
      
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      

      </header>

<div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>Import full genome-wide GWAS summary statistics from Open GWAS</h1>
    <small class="dont-index">Source: <a href='https://github.com/neurogenomics/MungeSumstats/blob/master/R/import_sumstats.R'><code>R/import_sumstats.R</code></a></small>
    <div class="hidden name"><code>import_sumstats.Rd</code></div>
    </div>

    <div class="ref-description">
    <p>Requires internet access to run.</p>
    </div>

    <pre class="usage"><span class='fu'>import_sumstats</span><span class='op'>(</span>
  <span class='va'>ids</span>,
  vcf_dir <span class='op'>=</span> <span class='fu'><a href='https://rdrr.io/r/base/tempfile.html'>tempdir</a></span><span class='op'>(</span><span class='op'>)</span>,
  vcf_download <span class='op'>=</span> <span class='cn'>TRUE</span>,
  save_dir <span class='op'>=</span> <span class='fu'><a href='https://rdrr.io/r/base/tempfile.html'>tempdir</a></span><span class='op'>(</span><span class='op'>)</span>,
  write_vcf <span class='op'>=</span> <span class='cn'>FALSE</span>,
  download_method <span class='op'>=</span> <span class='st'>"download.file"</span>,
  quiet <span class='op'>=</span> <span class='cn'>TRUE</span>,
  force_new_vcf <span class='op'>=</span> <span class='cn'>FALSE</span>,
  nThread <span class='op'>=</span> <span class='fl'>1</span>,
  parallel_across_ids <span class='op'>=</span> <span class='cn'>FALSE</span>,
  <span class='va'>...</span>
<span class='op'>)</span></pre>

    <h2 class="hasAnchor" id="arguments"><a class="anchor" href="#arguments"></a>Arguments</h2>
    <table class="ref-arguments">
    <colgroup><col class="name" /><col class="desc" /></colgroup>
    <tr>
      <th>ids</th>
      <td><p>List of Open GWAS study IDs (e.g. <code><a href='https://rdrr.io/r/base/c.html'>c("prot-a-664", "ieu-b-4760")</a></code>).</p></td>
    </tr>
    <tr>
      <th>vcf_dir</th>
      <td><p>Where to download the original VCF from Open GWAS.
<em>WARNING:</em> This is set to <code><a href='https://rdrr.io/r/base/tempfile.html'>tempdir()</a></code> by default.
This means the raw (pre-formatted) VCFs be deleted upon ending the R session.
Change this to keep the raw VCF file on disk (e.g. <code>vcf_dir="./raw_vcf"</code>).</p></td>
    </tr>
    <tr>
      <th>vcf_download</th>
      <td><p>Download the original VCF from Open GWAS.</p></td>
    </tr>
    <tr>
      <th>save_dir</th>
      <td><p>Directory to save formatted summary statistics in.</p></td>
    </tr>
    <tr>
      <th>write_vcf</th>
      <td><p>Whether to write as VCF (TRUE) or tabular file (FALSE).</p></td>
    </tr>
    <tr>
      <th>download_method</th>
      <td><p><code>"axel"</code> (multi-threaded) or
<code>"download.file"</code> (single-threaded) .</p></td>
    </tr>
    <tr>
      <th>quiet</th>
      <td><p>Run quietly.</p></td>
    </tr>
    <tr>
      <th>force_new_vcf</th>
      <td><p>Overwrite a previously downloaded VCF with the same path name.</p></td>
    </tr>
    <tr>
      <th>nThread</th>
      <td><p>Number of threads to use for parallel processes.</p></td>
    </tr>
    <tr>
      <th>parallel_across_ids</th>
      <td><p>If <code>parallel_across_ids=TRUE</code> and <code>nThread&gt;1</code>,
then each ID in <code>ids</code> will be processed in parallel.</p></td>
    </tr>
    <tr>
      <th>...</th>
      <td><p>Additional arguments to be passed to <code><a href='format_sumstats.html'>MungeSumstats::format_sumstats</a></code>.</p></td>
    </tr>
    </table>

    <h2 class="hasAnchor" id="value"><a class="anchor" href="#value"></a>Value</h2>

    <p>Either a named list of data objects or paths,
depending on the arguments passed to <code>format_sumstats</code>.</p>

    <h2 class="hasAnchor" id="examples"><a class="anchor" href="#examples"></a>Examples</h2>
    <pre class="examples"><div class='input'><span class='co'>### Search by criteria</span>
<span class='va'>metagwas</span> <span class='op'>&lt;-</span> <span class='fu'><a href='find_sumstats.html'>find_sumstats</a></span><span class='op'>(</span>
    traits <span class='op'>=</span> <span class='fu'><a href='https://rdrr.io/r/base/c.html'>c</a></span><span class='op'>(</span><span class='st'>"parkinson"</span>, <span class='st'>"alzheimer"</span><span class='op'>)</span>,
    min_sample_size <span class='op'>=</span> <span class='fl'>5000</span>
<span class='op'>)</span>
</div><div class='output co'>#&gt; <span class='message'>Collecting metadata from Open GWAS.</span></div><div class='output co'>#&gt; <span class='message'>Filtering metadata by substring criteria.</span></div><div class='output co'>#&gt; <span class='message'>Filtering metadata by sample/case/control/SNP size criteria.</span></div><div class='output co'>#&gt; <span class='message'>Excluding sample/case/control size with NAs.</span></div><div class='output co'>#&gt; <span class='message'>Found 16 GWAS datasets matching search criteria across:</span>
#&gt; <span class='message'>   - 12 trait(s)</span>
#&gt; <span class='message'>   - 1 population(s)</span>
#&gt; <span class='message'>   - 3 category(ies)</span>
#&gt; <span class='message'>   - 2 subcategory(ies)</span>
#&gt; <span class='message'>   - 5 publication(s)</span>
#&gt; <span class='message'>   - 6 consortia(ium)</span>
#&gt; <span class='message'>   - 1 genome build(s)</span></div><div class='input'><span class='co'>### Only use a subset for testing purposes</span>
<span class='va'>ids</span> <span class='op'>&lt;-</span> <span class='op'>(</span><span class='fu'>dplyr</span><span class='fu'>::</span><span class='fu'><a href='https://dplyr.tidyverse.org/reference/arrange.html'>arrange</a></span><span class='op'>(</span><span class='va'>metagwas</span>, <span class='va'>nsnp</span><span class='op'>)</span><span class='op'>)</span><span class='op'>$</span><span class='va'>id</span>

<span class='co'>### Default usage</span>
<span class='co'>## You can supply \code{import_sumstats()}</span>
<span class='co'>## with a list of as many OpenGWAS IDs as you want,</span>
<span class='co'>## but we'll just give one to save time.</span>

<span class='co'>## Call uses reference genome as default with more than 2GB of memory,</span>
<span class='co'>## which is more than what 32-bit Windows can handle so remove certain checks</span>
<span class='co'>## commented out down to runtime</span>
<span class='co'># datasets &lt;- import_sumstats(ids = ids[1])</span>
</div></pre>
  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <nav id="toc" data-toggle="toc" class="sticky-top">
      <h2 data-toc-skip>Contents</h2>
    </nav>
  </div>
</div>


      <footer>
      <div class="copyright">
  <p>Developed by Alan Murphy, Nathan Skene.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.6.1.</p>
</div>

      </footer>
   </div>

  


  </body>
</html>


