<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1.0"><title>Infer if effect relates to a1 or A2 if ambiguously named — infer_effect_column • MungeSumstats</title><!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.7.1/jquery.min.js" integrity="sha512-v2CJ7UaYy4JwqLDIrZUI/4hqeoQieOmAZNXBeQyjo21dadnwR+8ZaIJVT8EE2iyI61OV8e6M8PP2/4hpQINQ/g==" crossorigin="anonymous" referrerpolicy="no-referrer"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous"><script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css"><script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous"><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous"><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet"><script src="../pkgdown.js"></script><meta property="og:title" content="Infer if effect relates to a1 or A2 if ambiguously named — infer_effect_column"><meta property="og:description" content="Three checks are made to infer which allele the effect/frequency information
relates to if they are ambiguous (named A1 and A2 or equivalent):
Check if ambiguous naming conventions are used (i.e. allele 1 and 2 or
equivalent). If not exit, otherwise continue to next checks. This can be
checked by using the mapping file and splitting A1/A2 mappings by those that
contain 1 or 2 (ambiguous) or doesn't contain 1 or 2 e.g. effect,
tested (unambiguous so fine for MSS to handle as is).
Look for effect column/frequency column where the A1/A2 explicitly
mentioned, if found then  we know the direction and should update A1/A2
naming so A2 is the effect column. We can look for such columns by getting
every combination of A1/A2 naming and effect/frq naming.
If not found in 2, a final check should be against the reference genome,
whichever of A1 and A2 has more of a match with the reference genome should
be taken as not the effect allele. There is an assumption in this but is
still better than guessing the ambiguous allele naming.

"><!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--></head><body data-spy="scroll" data-target="#toc">
    

    <div class="container template-reference-topic">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">MungeSumstats</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">1.13.4</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav"><li>
  <a href="../articles/MungeSumstats.html">Get started</a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu"><li>
      <a href="../articles/docker.html">Docker/Singularity Containers</a>
    </li>
    <li>
      <a href="../articles/OpenGWAS.html">Import GWAS summary statistics from Open GWAS</a>
    </li>
  </ul></li>
<li>
  <a href="../news/index.html">Changelog</a>
</li>
      </ul><ul class="nav navbar-nav navbar-right"><li>
  <a href="https://github.com/neurogenomics/MungeSumstats/" class="external-link">
    <span class="fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul></div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>Infer if effect relates to a1 or A2 if ambiguously named</h1>
    <small class="dont-index">Source: <a href="https://github.com/neurogenomics/MungeSumstats/blob/HEAD/R/infer_effect_column.R" class="external-link"><code>R/infer_effect_column.R</code></a></small>
    <div class="hidden name"><code>infer_effect_column.Rd</code></div>
    </div>

    <div class="ref-description">
    <p>Three checks are made to infer which allele the effect/frequency information
relates to if they are ambiguous (named A1 and A2 or equivalent):</p><ol><li><p>Check if ambiguous naming conventions are used (i.e. allele 1 and 2 or
equivalent). If not exit, otherwise continue to next checks. This can be
checked by using the mapping file and splitting A1/A2 mappings by those that
contain 1 or 2 (ambiguous) or doesn't contain 1 or 2 e.g. effect,
tested (unambiguous so fine for MSS to handle as is).</p></li>
<li><p>Look for effect column/frequency column where the A1/A2 explicitly
mentioned, if found then  we know the direction and should update A1/A2
naming so A2 is the effect column. We can look for such columns by getting
every combination of A1/A2 naming and effect/frq naming.</p></li>
<li><p>If not found in 2, a final check should be against the reference genome,
whichever of A1 and A2 has more of a match with the reference genome should
be taken as <strong>not</strong> the effect allele. There is an assumption in this but is
still better than guessing the ambiguous allele naming.</p></li>
</ol></div>

    <div id="ref-usage">
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">infer_effect_column</span><span class="op">(</span></span>
<span>  <span class="va">sumstats_dt</span>,</span>
<span>  dbSNP <span class="op">=</span> <span class="fl">155</span>,</span>
<span>  sampled_snps <span class="op">=</span> <span class="fl">10000</span>,</span>
<span>  mapping_file <span class="op">=</span> <span class="va">sumstatsColHeaders</span>,</span>
<span>  nThread <span class="op">=</span> <span class="va">nThread</span>,</span>
<span>  ref_genome <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  on_ref_genome <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  infer_eff_direction <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  return_list <span class="op">=</span> <span class="cn">TRUE</span></span>
<span><span class="op">)</span></span></code></pre></div>
    </div>

    <div id="arguments">
    <h2>Arguments</h2>
    <dl><dt>sumstats_dt</dt>
<dd><p>data table obj of the summary statistics file for the
GWAS.</p></dd>


<dt>dbSNP</dt>
<dd><p>version of dbSNP to be used for imputation (144 or 155).</p></dd>


<dt>sampled_snps</dt>
<dd><p>Downsample the number of SNPs used when inferring genome
build to save time.</p></dd>


<dt>mapping_file</dt>
<dd><p>MungeSumstats has a pre-defined column-name mapping file
which should cover the most common column headers and their interpretations.
However, if a column header that is in youf file is missing of the mapping we
give is incorrect you can supply your own mapping file. Must be a 2 column
dataframe with column names "Uncorrected" and "Corrected". See
data(sumstatsColHeaders) for default mapping and necessary format.</p></dd>


<dt>nThread</dt>
<dd><p>Number of threads to use for parallel processes.</p></dd>


<dt>ref_genome</dt>
<dd><p>name of the reference genome used for the GWAS ("GRCh37" or
"GRCh38"). Argument is case-insensitive. Default is NULL which infers the
reference genome from the data.</p></dd>


<dt>on_ref_genome</dt>
<dd><p>Binary Should a check take place that all SNPs are on
the reference genome by SNP ID. Default is TRUE.</p></dd>


<dt>infer_eff_direction</dt>
<dd><p>Binary Should a check take place to ensure the
alleles match the effect direction? Default is TRUE.</p></dd>


<dt>return_list</dt>
<dd><p>Return the <code>sumstats_dt</code> within a named list
(default: <code>TRUE</code>).</p></dd>

</dl></div>
    <div id="value">
    <h2>Value</h2>
    

<p>list containing sumstats_dt, the modified summary statistics data
table object</p>
    </div>

    <div id="ref-examples">
    <h2>Examples</h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="r-in"><span><span class="va">sumstats</span> <span class="op">&lt;-</span> <span class="fu">MungeSumstats</span><span class="fu">::</span><span class="fu"><a href="formatted_example.html">formatted_example</a></span><span class="op">(</span><span class="op">)</span></span></span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Standardising column headers.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> First line of summary statistics file: </span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> MarkerName	CHR	POS	A1	A2	EAF	Beta	SE	Pval	</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Sorting coordinates with 'data.table'.</span>
<span class="r-in"><span><span class="co">#for speed, don't run on_ref_genome part of check (on_ref_genome = FALSE)</span></span></span>
<span class="r-in"><span><span class="va">sumstats_dt2</span><span class="op">&lt;-</span><span class="fu">infer_effect_column</span><span class="op">(</span>sumstats_dt<span class="op">=</span><span class="va">sumstats</span>,on_ref_genome <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span></span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Infer Effect Column</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> First line of summary statistics file: </span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> SNP	CHR	BP	A1	A2	FRQ	BETA	SE	P	</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Allele columns are ambiguous, attempting to infer direction</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Can't infer allele columns from sumstats</span>
</code></pre></div>
    </div>
  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <nav id="toc" data-toggle="toc" class="sticky-top"><h2 data-toc-skip>Contents</h2>
    </nav></div>
</div>


      <footer><div class="copyright">
  <p></p><p>Developed by Alan Murphy, Brian Schilder, Nathan Skene.</p>
</div>

<div class="pkgdown">
  <p></p><p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.9.</p>
</div>

      </footer></div>

  


  

  </body></html>

