<!-- Generated by pkgdown: do not edit by hand -->
<!DOCTYPE html>
<html lang="en">
  <head>
  <meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">

<title>Ensure that the input parameters are logical — validate_parameters • MungeSumstats</title>


<!-- jquery -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script>
<!-- Bootstrap -->
<link href="https://cdnjs.cloudflare.com/ajax/libs/bootswatch/3.4.0/flatly/bootstrap.min.css" rel="stylesheet" crossorigin="anonymous" />


<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script>

<!-- bootstrap-toc -->
<link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script>

<!-- Font Awesome icons -->
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous" />
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous" />

<!-- clipboard.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script>

<!-- headroom.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script>

<!-- pkgdown -->
<link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script>




<meta property="og:title" content="Ensure that the input parameters are logical — validate_parameters" />
<meta property="og:description" content="Ensure that the input parameters are logical" />




<!-- mathjax -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script>

<!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->


<!-- Global site tag (gtag.js) - Google Analytics -->
<script async src="https://www.googletagmanager.com/gtag/js?id=UA-175929032-1"></script>
<script>
  window.dataLayer = window.dataLayer || [];
  function gtag(){dataLayer.push(arguments);}
  gtag('js', new Date());

  gtag('config', 'UA-175929032-1');
</script>


  </head>

  <body data-spy="scroll" data-target="#toc">
    <div class="container template-reference-topic">
      <header>
      <div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">MungeSumstats</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">1.1.2</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="../index.html">
    <span class="fas fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../articles/MungeSumstats.html">Get started</a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
    <li>
      <a href="../articles/OpenGWAS.html">Import GWAS summary statistics from Open GWAS</a>
    </li>
  </ul>
</li>
<li>
  <a href="../news/index.html">Changelog</a>
</li>
      </ul>
      <ul class="nav navbar-nav navbar-right">
        <li>
  <a href="https://github.com/neurogenomics/MungeSumstats/">
    <span class="fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
      
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      

      </header>

<div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>Ensure that the input parameters are logical</h1>
    <small class="dont-index">Source: <a href='https://github.com/neurogenomics/MungeSumstats/blob/master/R/validate_parameters.R'><code>R/validate_parameters.R</code></a></small>
    <div class="hidden name"><code>validate_parameters.Rd</code></div>
    </div>

    <div class="ref-description">
    <p>Ensure that the input parameters are logical</p>
    </div>

    <pre class="usage"><span class='fu'>validate_parameters</span><span class='op'>(</span>
  <span class='va'>path</span>,
  <span class='va'>ref_genome</span>,
  <span class='va'>convert_small_p</span>,
  <span class='va'>compute_z</span>,
  <span class='va'>compute_n</span>,
  <span class='va'>convert_n_int</span>,
  <span class='va'>analysis_trait</span>,
  <span class='va'>INFO_filter</span>,
  <span class='va'>pos_se</span>,
  <span class='va'>effect_columns_nonzero</span>,
  <span class='va'>N_std</span>,
  <span class='va'>N_dropNA</span>,
  <span class='va'>rmv_chr</span>,
  <span class='va'>on_ref_genome</span>,
  <span class='va'>strand_ambig_filter</span>,
  <span class='va'>allele_flip_check</span>,
  <span class='va'>allele_flip_drop</span>,
  <span class='va'>allele_flip_z</span>,
  <span class='va'>bi_allelic_filter</span>,
  <span class='va'>snp_ids_are_rs_ids</span>,
  <span class='va'>write_vcf</span>,
  <span class='va'>return_format</span>,
  <span class='va'>ldsc_format</span>,
  <span class='va'>imputation_ind</span>,
  <span class='va'>log_folder_ind</span>,
  <span class='va'>log_mungesumstats_msgs</span>
<span class='op'>)</span></pre>

    <h2 class="hasAnchor" id="arguments"><a class="anchor" href="#arguments"></a>Arguments</h2>
    <table class="ref-arguments">
    <colgroup><col class="name" /><col class="desc" /></colgroup>
    <tr>
      <th>path</th>
      <td><p>Filepath for the summary statistics file to be formatted. Default
is a temporary file.</p></td>
    </tr>
    <tr>
      <th>ref_genome</th>
      <td><p>name of the reference genome used for the GWAS ("GRCh37" or
"GRCh38"). Argument is case-insensitive. Default is NULL which infers the
reference genome from the data.</p></td>
    </tr>
    <tr>
      <th>convert_small_p</th>
      <td><p>Binary, should p-values &lt; 5e-324 be converted to 0?
Small p-values pass the R limit and can cause errors with LDSC/MAGMA and
should be converted. Default is TRUE.</p></td>
    </tr>
    <tr>
      <th>compute_z</th>
      <td><p>Whether to compute Z-score column from P. Default is FALSE.
<strong>Note</strong> that imputing the Z-score for every SNP will not correct be
perfectly correct and may result in a loss of power. This should only be done
as a last resort.</p></td>
    </tr>
    <tr>
      <th>compute_n</th>
      <td><p>Whether to impute N. Default of 0 won't impute, any other
integer will be imputed as the N (sample size) for every SNP in the dataset.
<strong>Note</strong> that imputing the sample size for every SNP is not correct and
should only be done as a last resort.</p></td>
    </tr>
    <tr>
      <th>convert_n_int</th>
      <td><p>Binary, if N (the number of samples) is not an integer,
should this be rounded? Default is TRUE.</p></td>
    </tr>
    <tr>
      <th>analysis_trait</th>
      <td><p>If multiple traits were studied, name of the trait for
analysis from the GWAS. Default is NULL.</p></td>
    </tr>
    <tr>
      <th>INFO_filter</th>
      <td><p>numeric The minimum value permissible of the imputation
information score (if present in sumstats file). Default 0.9.</p></td>
    </tr>
    <tr>
      <th>pos_se</th>
      <td><p>Binary Should the standard Error (SE) column be checked to
ensure it is greater than 0? Those that are, are removed (if present in
sumstats file). Default TRUE.</p></td>
    </tr>
    <tr>
      <th>effect_columns_nonzero</th>
      <td><p>Binary should the effect columns in the data
BETA,OR (odds ratio),LOG_ODDS,SIGNED_SUMSTAT be checked to ensure no SNP=0.
Those that do are removed(if present in sumstats file). Default TRUE.</p></td>
    </tr>
    <tr>
      <th>N_std</th>
      <td><p>numeric The number of standard deviations above the mean a SNP's
N is needed to be removed. Default is 5.</p></td>
    </tr>
    <tr>
      <th>N_dropNA</th>
      <td><p>Drop rows where N is missing.Default is TRUE.</p></td>
    </tr>
    <tr>
      <th>rmv_chr</th>
      <td><p>vector or character The chromosomes on which the SNPs should
be removed. Use NULL if no filtering necessary. Default is X, Y and
mitochondrial.</p></td>
    </tr>
    <tr>
      <th>on_ref_genome</th>
      <td><p>Binary Should a check take place that all SNPs are on
the reference genome by SNP ID. Default is TRUE.</p></td>
    </tr>
    <tr>
      <th>strand_ambig_filter</th>
      <td><p>Binary Should SNPs with strand-ambiguous alleles
be removed. Default is FALSE.</p></td>
    </tr>
    <tr>
      <th>allele_flip_check</th>
      <td><p>Binary Should the allele columns be checked against
reference genome to infer if flipping is necessary. Default is TRUE.</p></td>
    </tr>
    <tr>
      <th>allele_flip_drop</th>
      <td><p>Binary Should the SNPs for which neither their A1 or
A2 base pair values match a reference genome be dropped. Default is TRUE.</p></td>
    </tr>
    <tr>
      <th>allele_flip_z</th>
      <td><p>Binary should the Z-score be flipped along with effect
columns like Beta? It is assumed to be calculated off the effect size not
the P-value and so will be flipped i.e. default TRUE.</p></td>
    </tr>
    <tr>
      <th>bi_allelic_filter</th>
      <td><p>Binary Should non-biallelic SNPs be removed. Default
is TRUE.</p></td>
    </tr>
    <tr>
      <th>snp_ids_are_rs_ids</th>
      <td><p>Binary Should the supplied SNP ID's be assumed to
be RS IDs. If not, imputation using the SNP ID for other columns like
base-pair position or chromosome will not be possible. If set to FALSE, the
SNP RS ID will be imputed from the reference genome if possible. Default is
TRUE.</p></td>
    </tr>
    <tr>
      <th>write_vcf</th>
      <td><p>Whether to write as VCF (TRUE) or tabular file (FALSE).</p></td>
    </tr>
    <tr>
      <th>return_format</th>
      <td><p>If return_data is TRUE. Object type to be returned
("data.table","vranges","granges").</p></td>
    </tr>
    <tr>
      <th>ldsc_format</th>
      <td><p>Binary Ensure that output format meets all requirements
to be fed directly into LDSC without the need for additional munging. Default
is FALSE</p></td>
    </tr>
    <tr>
      <th>imputation_ind</th>
      <td><p>Binary Should a column be added for each imputation
step to show what SNPs have imputed values for differing fields. This
includes a field denoting SNP allele flipping (flipped). On the flipped
value, this denoted whether the alelles where switched based on
MungeSumstats initial choice of A1, A2 from the input column headers and thus
may not align with what the creator intended.<strong>Note</strong> these columns will be
in the formatted summary statistics returned. Default is FALSE.</p></td>
    </tr>
    <tr>
      <th>log_folder_ind</th>
      <td><p>Binary Should log files be stored containing all
filtered out SNPs (separate file per filter). The data is outputted in the
same format specified for the resulting sumstats file. The only exception to
this rule is if output is vcf, then log file saved as .tsv.gz. Default is
FALSE.</p></td>
    </tr>
    <tr>
      <th>log_mungesumstats_msgs</th>
      <td><p>Binary Should a log be stored containing all
messages and errors printed by MungeSumstats in a run. Default is FALSE</p></td>
    </tr>
    </table>

    <h2 class="hasAnchor" id="value"><a class="anchor" href="#value"></a>Value</h2>

    <p>No return</p>

  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <nav id="toc" data-toggle="toc" class="sticky-top">
      <h2 data-toc-skip>Contents</h2>
    </nav>
  </div>
</div>


      <footer>
      <div class="copyright">
  <p>Developed by Alan Murphy, Nathan Skene.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.6.1.</p>
</div>

      </footer>
   </div>

  


  </body>
</html>


