<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Import GWAS summary statistics from Open GWAS • MungeSumstats</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="Import GWAS summary statistics from Open GWAS">
<meta property="og:description" content="MungeSumstats">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    

    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">MungeSumstats</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">1.7.7</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../articles/MungeSumstats.html">Get started</a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/OpenGWAS.html">Import GWAS summary statistics from Open GWAS</a>
    </li>
    <li>
      <a href="../articles/docker.html">Docker/Singularity Containers</a>
    </li>
  </ul>
</li>
<li>
  <a href="../news/index.html">Changelog</a>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/neurogenomics/MungeSumstats/" class="external-link">
    <span class="fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>Import GWAS summary statistics from Open
GWAS</h1>
                        <h4 data-toc-skip class="author"><h5> ¶
<i>Authors</i>: Alan Murphy, Brian Schilder and Nathan Skene ¶
</h5></h4>
            
            <h4 data-toc-skip class="date"><h5> ¶ <i>Updated</i>:
Dec-02-2022 ¶ </h5></h4>
      
      <small class="dont-index">Source: <a href="https://github.com/neurogenomics/MungeSumstats/blob/HEAD/vignettes/OpenGWAS.Rmd" class="external-link"><code>vignettes/OpenGWAS.Rmd</code></a></small>
      <div class="hidden name"><code>OpenGWAS.Rmd</code></div>

    </div>

    
    
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/neurogenomics/MungeSumstats" class="external-link">MungeSumstats</a></span><span class="op">)</span></span></code></pre></div>
<p>MungeSumstats now offers high throughput query and import
functionality to data from the MRC IEU <a href="https://gwas.mrcieu.ac.uk/" class="external-link">Open GWAS Project</a>.</p>
<div class="section level2">
<h2 id="find-gwas-datasets">Find GWAS datasets<a class="anchor" aria-label="anchor" href="#find-gwas-datasets"></a>
</h2>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">#### Search for datasets ####</span></span>
<span><span class="va">metagwas</span> <span class="op">&lt;-</span> <span class="fu">MungeSumstats</span><span class="fu">::</span><span class="fu"><a href="../reference/find_sumstats.html">find_sumstats</a></span><span class="op">(</span>traits <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"parkinson"</span>,<span class="st">"alzheimer"</span><span class="op">)</span>, </span>
<span>                                         min_sample_size <span class="op">=</span> <span class="fl">1000</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">metagwas</span>,<span class="fl">3</span><span class="op">)</span></span>
<span><span class="va">ids</span> <span class="op">&lt;-</span> <span class="op">(</span><span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/arrange.html" class="external-link">arrange</a></span><span class="op">(</span><span class="va">metagwas</span>, <span class="va">nsnp</span><span class="op">)</span><span class="op">)</span><span class="op">$</span><span class="va">id</span>  </span></code></pre></div>
<pre><code><span><span class="co">##          id               trait group_name year    author</span></span>
<span><span class="co">## 1 ieu-a-298 Alzheimer's disease     public 2013   Lambert</span></span>
<span><span class="co">## 2   ieu-b-2 Alzheimer's disease     public 2019 Kunkle BW</span></span>
<span><span class="co">## 3 ieu-a-297 Alzheimer's disease     public 2013   Lambert</span></span>
<span><span class="co">##                                                                                                                                                                                                                                                                                                                    consortium</span></span>
<span><span class="co">## 1                                                                                                                                                                                                                                                                                                                        IGAP</span></span>
<span><span class="co">## 2 Alzheimer Disease Genetics Consortium (ADGC), European Alzheimer's Disease Initiative (EADI), Cohorts for Heart and Aging Research in Genomic Epidemiology Consortium (CHARGE), Genetic and Environmental Risk in AD/Defining Genetic, Polygenic and Environmental Risk for Alzheimer's Disease Consortium (GERAD/PERADES),</span></span>
<span><span class="co">## 3                                                                                                                                                                                                                                                                                                                        IGAP</span></span>
<span><span class="co">##                 sex population     unit     nsnp sample_size       build</span></span>
<span><span class="co">## 1 Males and Females   European log odds    11633       74046 HG19/GRCh37</span></span>
<span><span class="co">## 2 Males and Females   European       NA 10528610       63926 HG19/GRCh37</span></span>
<span><span class="co">## 3 Males and Females   European log odds  7055882       54162 HG19/GRCh37</span></span>
<span><span class="co">##   category                subcategory ontology mr priority     pmid sd</span></span>
<span><span class="co">## 1  Disease Psychiatric / neurological       NA  1        1 24162737 NA</span></span>
<span><span class="co">## 2   Binary Psychiatric / neurological       NA  1        0 30820047 NA</span></span>
<span><span class="co">## 3  Disease Psychiatric / neurological       NA  1        2 24162737 NA</span></span>
<span><span class="co">##                                                                      note ncase</span></span>
<span><span class="co">## 1 Exposure only; Effect allele frequencies are missing; forward(+) strand 25580</span></span>
<span><span class="co">## 2                                                                      NA 21982</span></span>
<span><span class="co">## 3                Effect allele frequencies are missing; forward(+) strand 17008</span></span>
<span><span class="co">##   ncontrol     N</span></span>
<span><span class="co">## 1    48466 74046</span></span>
<span><span class="co">## 2    41944 63926</span></span>
<span><span class="co">## 3    37154 54162</span></span></code></pre>
</div>
<div class="section level2">
<h2 id="import-full-results">Import full results<a class="anchor" aria-label="anchor" href="#import-full-results"></a>
</h2>
<p>You can supply <code><a href="../reference/import_sumstats.html">import_sumstats()</a></code> with a list of as many
OpenGWAS IDs as you want, but we’ll just give one to save time.</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">datasets</span> <span class="op">&lt;-</span> <span class="fu">MungeSumstats</span><span class="fu">::</span><span class="fu"><a href="../reference/import_sumstats.html">import_sumstats</a></span><span class="op">(</span>ids <span class="op">=</span> <span class="st">"ieu-a-298"</span>,</span>
<span>                                           ref_genome <span class="op">=</span> <span class="st">"GRCH37"</span><span class="op">)</span></span></code></pre></div>
<div class="section level3">
<h3 id="summarise-results">Summarise results<a class="anchor" aria-label="anchor" href="#summarise-results"></a>
</h3>
<p>By default, <code>import_sumstats</code> results a named list where
the names are the Open GWAS dataset IDs and the items are the respective
paths to the formatted summary statistics.</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/print.html" class="external-link">print</a></span><span class="op">(</span><span class="va">datasets</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## $`ieu-a-298`</span></span>
<span><span class="co">## [1] "/tmp/RtmpjX8IcG/ieu-a-298.tsv.gz"</span></span></code></pre>
<p>You can easily turn this into a data.frame as well.</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">results_df</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html" class="external-link">data.frame</a></span><span class="op">(</span>id<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">datasets</span><span class="op">)</span>, </span>
<span>                         path<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/unlist.html" class="external-link">unlist</a></span><span class="op">(</span><span class="va">datasets</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/print.html" class="external-link">print</a></span><span class="op">(</span><span class="va">results_df</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">##                  id                             path</span></span>
<span><span class="co">## ieu-a-298 ieu-a-298 /tmp/RtmpjX8IcG/ieu-a-298.tsv.gz</span></span></code></pre>
</div>
</div>
<div class="section level2">
<h2 id="import-full-results-parallel">Import full results (parallel)<a class="anchor" aria-label="anchor" href="#import-full-results-parallel"></a>
</h2>
<p><em>Optional</em>: Speed up with multi-threaded download via <a href="https://github.com/axel-download-accelerator/axel" class="external-link">axel</a>.</p>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">datasets</span> <span class="op">&lt;-</span> <span class="fu">MungeSumstats</span><span class="fu">::</span><span class="fu"><a href="../reference/import_sumstats.html">import_sumstats</a></span><span class="op">(</span>ids <span class="op">=</span> <span class="va">ids</span>, </span>
<span>                                           vcf_download <span class="op">=</span> <span class="cn">TRUE</span>, </span>
<span>                                           download_method <span class="op">=</span> <span class="st">"axel"</span>, </span>
<span>                                           nThread <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/Extremes.html" class="external-link">max</a></span><span class="op">(</span><span class="fl">2</span>,<span class="fu">future</span><span class="fu">::</span><span class="fu">availableCores</span><span class="op">(</span><span class="op">)</span><span class="op">-</span><span class="fl">2</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="further-functionality">Further functionality<a class="anchor" aria-label="anchor" href="#further-functionality"></a>
</h2>
<p>See the <a href="https://neurogenomics.github.io/MungeSumstats/articles/MungeSumstats.html" class="external-link">Getting
started vignette</a> for more information on how to use MungeSumstats
and its functionality.</p>
</div>
<div class="section level2">
<h2 id="session-info">Session Info<a class="anchor" aria-label="anchor" href="#session-info"></a>
</h2>
<details><div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">utils</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/r/utils/sessionInfo.html" class="external-link">sessionInfo</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## R Under development (unstable) (2022-11-23 r83380)</span></span>
<span><span class="co">## Platform: x86_64-pc-linux-gnu (64-bit)</span></span>
<span><span class="co">## Running under: Ubuntu 22.04.1 LTS</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## Matrix products: default</span></span>
<span><span class="co">## BLAS:   /usr/lib/x86_64-linux-gnu/openblas-pthread/libblas.so.3</span></span>
<span><span class="co">## LAPACK: /usr/lib/x86_64-linux-gnu/openblas-pthread/libopenblasp-r0.3.20.so</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## locale:</span></span>
<span><span class="co">##  [1] LC_CTYPE=en_US.UTF-8       LC_NUMERIC=C              </span></span>
<span><span class="co">##  [3] LC_TIME=en_US.UTF-8        LC_COLLATE=en_US.UTF-8    </span></span>
<span><span class="co">##  [5] LC_MONETARY=en_US.UTF-8    LC_MESSAGES=en_US.UTF-8   </span></span>
<span><span class="co">##  [7] LC_PAPER=en_US.UTF-8       LC_NAME=C                 </span></span>
<span><span class="co">##  [9] LC_ADDRESS=C               LC_TELEPHONE=C            </span></span>
<span><span class="co">## [11] LC_MEASUREMENT=en_US.UTF-8 LC_IDENTIFICATION=C       </span></span>
<span><span class="co">## </span></span>
<span><span class="co">## time zone: UTC</span></span>
<span><span class="co">## tzcode source: system (glibc)</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## attached base packages:</span></span>
<span><span class="co">## [1] stats     graphics  grDevices utils     datasets  methods   base     </span></span>
<span><span class="co">## </span></span>
<span><span class="co">## other attached packages:</span></span>
<span><span class="co">## [1] MungeSumstats_1.7.7 BiocStyle_2.27.0   </span></span>
<span><span class="co">## </span></span>
<span><span class="co">## loaded via a namespace (and not attached):</span></span>
<span><span class="co">##   [1] tidyselect_1.2.0            dplyr_1.0.10               </span></span>
<span><span class="co">##   [3] blob_1.2.3                  filelock_1.0.2             </span></span>
<span><span class="co">##   [5] R.utils_2.12.2              Biostrings_2.67.0          </span></span>
<span><span class="co">##   [7] bitops_1.0-7                fastmap_1.1.0              </span></span>
<span><span class="co">##   [9] RCurl_1.98-1.9              BiocFileCache_2.7.0        </span></span>
<span><span class="co">##  [11] VariantAnnotation_1.45.0    GenomicAlignments_1.35.0   </span></span>
<span><span class="co">##  [13] XML_3.99-0.12               digest_0.6.30              </span></span>
<span><span class="co">##  [15] lifecycle_1.0.3             ellipsis_0.3.2             </span></span>
<span><span class="co">##  [17] KEGGREST_1.39.0             RSQLite_2.2.19             </span></span>
<span><span class="co">##  [19] googleAuthR_2.0.0           magrittr_2.0.3             </span></span>
<span><span class="co">##  [21] compiler_4.3.0              rlang_1.0.6                </span></span>
<span><span class="co">##  [23] sass_0.4.4                  progress_1.2.2             </span></span>
<span><span class="co">##  [25] tools_4.3.0                 utf8_1.2.2                 </span></span>
<span><span class="co">##  [27] yaml_2.3.6                  data.table_1.14.6          </span></span>
<span><span class="co">##  [29] rtracklayer_1.59.0          knitr_1.41                 </span></span>
<span><span class="co">##  [31] prettyunits_1.1.1           curl_4.3.3                 </span></span>
<span><span class="co">##  [33] bit_4.0.5                   DelayedArray_0.25.0        </span></span>
<span><span class="co">##  [35] xml2_1.3.3                  BiocParallel_1.33.6        </span></span>
<span><span class="co">##  [37] purrr_0.3.5                 BiocGenerics_0.45.0        </span></span>
<span><span class="co">##  [39] desc_1.4.2                  R.oo_1.25.0                </span></span>
<span><span class="co">##  [41] grid_4.3.0                  stats4_4.3.0               </span></span>
<span><span class="co">##  [43] fansi_1.0.3                 biomaRt_2.55.0             </span></span>
<span><span class="co">##  [45] SummarizedExperiment_1.29.1 cli_3.4.1                  </span></span>
<span><span class="co">##  [47] rmarkdown_2.18              crayon_1.5.2               </span></span>
<span><span class="co">##  [49] generics_0.1.3              ragg_1.2.4                 </span></span>
<span><span class="co">##  [51] httr_1.4.4                  rjson_0.2.21               </span></span>
<span><span class="co">##  [53] DBI_1.1.3                   cachem_1.0.6               </span></span>
<span><span class="co">##  [55] stringr_1.4.1               zlibbioc_1.45.0            </span></span>
<span><span class="co">##  [57] assertthat_0.2.1            parallel_4.3.0             </span></span>
<span><span class="co">##  [59] AnnotationDbi_1.61.0        BiocManager_1.30.19        </span></span>
<span><span class="co">##  [61] XVector_0.39.0              restfulr_0.0.15            </span></span>
<span><span class="co">##  [63] matrixStats_0.63.0          vctrs_0.5.1                </span></span>
<span><span class="co">##  [65] Matrix_1.5-3                jsonlite_1.8.3             </span></span>
<span><span class="co">##  [67] bookdown_0.30               IRanges_2.33.0             </span></span>
<span><span class="co">##  [69] hms_1.1.2                   S4Vectors_0.37.0           </span></span>
<span><span class="co">##  [71] bit64_4.0.5                 systemfonts_1.0.4          </span></span>
<span><span class="co">##  [73] GenomicFeatures_1.51.2      jquerylib_0.1.4            </span></span>
<span><span class="co">##  [75] glue_1.6.2                  pkgdown_2.0.6.9000         </span></span>
<span><span class="co">##  [77] codetools_0.2-18            stringi_1.7.8              </span></span>
<span><span class="co">##  [79] GenomeInfoDb_1.35.5         BiocIO_1.9.1               </span></span>
<span><span class="co">##  [81] GenomicRanges_1.51.1        tibble_3.1.8               </span></span>
<span><span class="co">##  [83] pillar_1.8.1                rappdirs_0.3.3             </span></span>
<span><span class="co">##  [85] htmltools_0.5.3             GenomeInfoDbData_1.2.9     </span></span>
<span><span class="co">##  [87] BSgenome_1.67.1             R6_2.5.1                   </span></span>
<span><span class="co">##  [89] dbplyr_2.2.1                textshaping_0.3.6          </span></span>
<span><span class="co">##  [91] rprojroot_2.0.3             evaluate_0.18              </span></span>
<span><span class="co">##  [93] lattice_0.20-45             Biobase_2.59.0             </span></span>
<span><span class="co">##  [95] R.methodsS3_1.8.2           png_0.1-8                  </span></span>
<span><span class="co">##  [97] Rsamtools_2.15.0            gargle_1.2.1               </span></span>
<span><span class="co">##  [99] memoise_2.0.1               bslib_0.4.1                </span></span>
<span><span class="co">## [101] Rcpp_1.0.9                  xfun_0.35                  </span></span>
<span><span class="co">## [103] fs_1.5.2                    MatrixGenerics_1.11.0      </span></span>
<span><span class="co">## [105] pkgconfig_2.0.3</span></span></code></pre>
</details>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

        <nav id="toc" data-toggle="toc"><h2 data-toc-skip>Contents</h2>
    </nav>
</div>

</div>



      <footer><div class="copyright">
  <p></p>
<p>Developed by Alan Murphy, Brian Schilder, Nathan Skene.</p>
</div>

<div class="pkgdown">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.6.9000.</p>
</div>

      </footer>
</div>

  


  

  </body>
</html>
